
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pypath.inputs.biomart &#8212; pypath 0.14.30 documentation</title>
    <link rel="stylesheet" href="../../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/insipid.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/pypathdocs.css" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script defer src="../../../_static/insipid.js"></script>
    <script defer src="../../../_static/insipid-sidebar.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-capable" content="yes">
  </head><body>
    <script type="text/javascript">
        document.body.classList.add('js');
    </script>
    <input type="checkbox" id="sidebar-checkbox" style="display: none;">
    <label id="overlay" for="sidebar-checkbox"></label>
    <div class="sidebar-resize-handle"></div>
    <script type="text/javascript">
        try {
            let sidebar = localStorage.getItem('sphinx-sidebar');
            const sidebar_width = localStorage.getItem('sphinx-sidebar-width');
            if (sidebar_width) {
                document.documentElement.style.setProperty('--sidebar-width', sidebar_width);
            }
            // show sidebar on wide screen
            if (!sidebar && window.matchMedia('(min-width: 100rem)').matches) {
                sidebar = 'visible';
                // NB: We don't store the value in localStorage!
            }
            if (sidebar === 'visible') {
                document.getElementById('sidebar-checkbox').checked = true;
            }
        } catch(e) {
            console.info(e);
        }
    </script>
    <header id="topbar-placeholder">
      <div id="topbar">
        <div id="titlebar">
          <div class="buttons">
            <label for="sidebar-checkbox" id="sidebar-button" role="button" tabindex="0" aria-controls="sphinxsidebar" accesskey="M">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"/></svg>
            </label>
          </div>
          <div class="title">
            <a class="parent" href="../inputs.html" accesskey="U">pypath.inputs</a>
            <a class="top" href="#">pypath.inputs.biomart</a>
          </div>
          <div class="buttons">
            <button id="fullscreen-button" type="button" aria-hidden="true">
              <span class="enable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M212.686 315.314L120 408l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C10.697 480 0 469.255 0 456V344c0-21.382 25.803-32.09 40.922-16.971L72 360l92.686-92.686c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.249 6.248 6.249 16.378 0 22.627zm22.628-118.628L328 104l-32.922-31.029C279.958 57.851 290.666 32 312.048 32h112C437.303 32 448 42.745 448 56v112c0 21.382-25.803 32.09-40.922 16.971L376 152l-92.686 92.686c-6.248 6.248-16.379 6.248-22.627 0l-25.373-25.373c-6.249-6.248-6.249-16.378 0-22.627z"/></svg>
              </span>
              <span class="disable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"/></svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </header>
    <nav>
    </nav>

    <nav class="relbar">
    </nav>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
<h1>Source code for pypath.inputs.biomart</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">#</span>
<span class="c1">#  This file is part of the `pypath` python module</span>
<span class="c1">#  Helps to translate from the mouse data to human data</span>
<span class="c1">#</span>
<span class="c1">#  Copyright</span>
<span class="c1">#  2014-2022</span>
<span class="c1">#  EMBL, EMBL-EBI, Uniklinik RWTH Aachen, Heidelberg University</span>
<span class="c1">#</span>
<span class="c1">#  Authors: Dénes Türei (turei.denes@gmail.com)</span>
<span class="c1">#           Nicolàs Palacio</span>
<span class="c1">#           Sebastian Lobentanzer</span>
<span class="c1">#           Erva Ulusoy</span>
<span class="c1">#           Olga Ivanova</span>
<span class="c1">#           Ahmet Rifaioglu</span>
<span class="c1">#</span>
<span class="c1">#  Distributed under the GPLv3 License.</span>
<span class="c1">#  See accompanying file LICENSE.txt or copy at</span>
<span class="c1">#      http://www.gnu.org/licenses/gpl-3.0.html</span>
<span class="c1">#</span>
<span class="c1">#  Website: http://pypath.omnipathdb.org/</span>
<span class="c1">#</span>

<span class="kn">from</span> <span class="nn">future.utils</span> <span class="kn">import</span> <span class="n">iteritems</span>

<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">collections</span>

<span class="kn">import</span> <span class="nn">pypath.share.session</span> <span class="k">as</span> <span class="nn">session_mod</span>
<span class="kn">import</span> <span class="nn">pypath.share.common</span> <span class="k">as</span> <span class="nn">common</span>
<span class="kn">import</span> <span class="nn">pypath.share.curl</span> <span class="k">as</span> <span class="nn">curl</span>
<span class="kn">import</span> <span class="nn">pypath.resources.urls</span> <span class="k">as</span> <span class="nn">urls</span>
<span class="kn">import</span> <span class="nn">pypath.utils.taxonomy</span> <span class="k">as</span> <span class="nn">taxonomy</span>

<span class="n">_logger</span> <span class="o">=</span> <span class="n">session_mod</span><span class="o">.</span><span class="n">Logger</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;biomart_input&#39;</span><span class="p">)</span>


<span class="c1"># for mouse homologues: Filter name = &quot;with_mmusculus_homolog&quot;</span>
<span class="n">_filter_xml_template</span> <span class="o">=</span> <span class="s1">&#39;&lt;Filter name=&quot;</span><span class="si">%s</span><span class="s1">&quot; excluded=&quot;0&quot;/&gt;&#39;</span>
<span class="n">_attr_xml_template</span> <span class="o">=</span> <span class="s1">&#39;&lt;Attribute name=&quot;</span><span class="si">%s</span><span class="s1">&quot; /&gt;&#39;</span>


<div class="viewcode-block" id="biomart_query"><a class="viewcode-back" href="../../../_autosummary/pypath.inputs.biomart.biomart_query.html#pypath.inputs.biomart.biomart_query">[docs]</a><span class="k">def</span> <span class="nf">biomart_query</span><span class="p">(</span>
        <span class="n">attrs</span><span class="p">,</span>
        <span class="n">filters</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;hsapiens_gene_ensembl&#39;</span><span class="p">,</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Query the Ensembl Biomart web service.</span>
<span class="sd">    Use https://www.ensembl.org/biomart/martview/ to check for attribute</span>
<span class="sd">    and dataset names.</span>

<span class="sd">    Args</span>
<span class="sd">        attrs (str,list): One or more Ensembl attribute names.</span>
<span class="sd">        filters (str,list): One or more Ensembl filter names.</span>
<span class="sd">        transcript (bool): Include Ensembl transcript IDs in the result.</span>
<span class="sd">        peptide (bool): Include Ensembl peptide IDs in the result.</span>
<span class="sd">        gene (bool): Include Ensembl gene IDs in the result.</span>
<span class="sd">        dataset (str): An Ensembl dataset name.</span>

<span class="sd">    Yields:</span>
<span class="sd">        Named tuples with the requested attributes for each record returned</span>
<span class="sd">        by Ensembl Biomart.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_attrs</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">if</span> <span class="n">gene</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_gene_id&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">transcript</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_transcript_id&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">peptide</span><span class="p">:</span>

        <span class="n">_attrs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;ensembl_peptide_id&#39;</span><span class="p">)</span>

    <span class="n">_attrs</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">attrs</span><span class="p">))</span>
    <span class="n">filters</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">filters</span><span class="p">)</span>

    <span class="n">record</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;EnsemblRecord&#39;</span><span class="p">,</span>
        <span class="n">_attrs</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
        <span class="s1">&#39;Downloading data from Ensembl Biomart: &#39;</span>
        <span class="s1">&#39;dataset=`</span><span class="si">%s</span><span class="s1">`, &#39;</span>
        <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">&#39;</span>
        <span class="s1">&#39;attributes=`</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">dataset</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="s1">&#39;filters=`</span><span class="si">%s</span><span class="s1">`, &#39;</span> <span class="o">%</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filters</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">filters</span> <span class="k">else</span>
                <span class="s1">&#39;&#39;</span>
            <span class="p">),</span>
            <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">_attrs</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">rewsp</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;\n\s+&#39;</span><span class="p">)</span>

    <span class="n">xml_template_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">common</span><span class="o">.</span><span class="n">DATA</span><span class="p">,</span> <span class="s1">&#39;ensembl_biomart_query.xml&#39;</span><span class="p">)</span>

    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">xml_template_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fp</span><span class="p">:</span>

        <span class="n">xml_template</span> <span class="o">=</span> <span class="n">fp</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>

    <span class="n">filter_part</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">_filter_xml_template</span> <span class="o">%</span> <span class="n">_filter</span>
        <span class="k">for</span> <span class="n">_filter</span> <span class="ow">in</span> <span class="n">filters</span>
    <span class="p">)</span>
    <span class="n">attr_part</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">_attr_xml_template</span> <span class="o">%</span> <span class="n">_attr</span>
        <span class="k">for</span> <span class="n">_attr</span> <span class="ow">in</span> <span class="n">_attrs</span>
    <span class="p">)</span>

    <span class="n">xml_query</span> <span class="o">=</span> <span class="n">xml_template</span> <span class="o">%</span> <span class="p">(</span>
        <span class="n">dataset</span><span class="p">,</span>
        <span class="n">filter_part</span><span class="p">,</span>
        <span class="n">attr_part</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">xml_query</span> <span class="o">=</span> <span class="n">rewsp</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">xml_query</span><span class="p">)</span>

    <span class="n">biomart_url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;ensembl&#39;</span><span class="p">][</span><span class="s1">&#39;biomart_url&#39;</span><span class="p">]</span> <span class="o">%</span> <span class="n">xml_query</span>

    <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">biomart_url</span><span class="p">,</span> <span class="n">large</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">silent</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

    <span class="n">success</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">c</span><span class="o">.</span><span class="n">result</span><span class="p">:</span>

        <span class="n">_line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n\r</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

        <span class="n">success</span> <span class="o">=</span> <span class="n">success</span> <span class="ow">or</span> <span class="n">_line</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;[success]&#39;</span>

        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">_line</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">_fields</span><span class="p">):</span>

            <span class="k">yield</span> <span class="n">record</span><span class="p">(</span><span class="o">*</span><span class="n">_line</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">success</span><span class="p">:</span>

        <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Error: Interrupted transfer while downlading data &#39;</span>
            <span class="s1">&#39;from Ensembl Biomart (missing `success` tag).&#39;</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="biomart_homology"><a class="viewcode-back" href="../../../_autosummary/pypath.inputs.biomart.biomart_homology.html#pypath.inputs.biomart.biomart_homology">[docs]</a><span class="k">def</span> <span class="nf">biomart_homology</span><span class="p">(</span>
        <span class="n">source_organism</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">target_organism</span> <span class="o">=</span> <span class="mi">10090</span><span class="p">,</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves homology data from Ensembl Biomart.</span>

<span class="sd">    Returns</span>
<span class="sd">        List of named tuples filtered to genes of the source organism having</span>
<span class="sd">        orthologues in the target organism, with homology related fields.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">ensure_organism</span><span class="p">(</span><span class="n">organism</span><span class="p">):</span>

        <span class="n">organism_ensembl</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">organism_ensembl</span><span class="p">:</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Could not find Ensembl taxon ID for `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">organism_ensembl</span>


    <span class="n">source_organism</span> <span class="o">=</span> <span class="n">ensure_organism</span><span class="p">(</span><span class="n">source_organism</span><span class="p">)</span>
    <span class="n">target_organism</span> <span class="o">=</span> <span class="n">ensure_organism</span><span class="p">(</span><span class="n">target_organism</span><span class="p">)</span>

    <span class="n">homolog_attrs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;homolog_ensembl_peptide&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_ensembl_gene&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_orthology_type&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_orthology_confidence&#39;</span><span class="p">,</span>
        <span class="s1">&#39;homolog_canonical_transcript_protein&#39;</span><span class="p">,</span>
    <span class="p">]</span>

    <span class="n">homolog_attrs</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">target_organism</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="n">homolog_attrs</span>
    <span class="p">]</span>

    <span class="n">filters</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;with_</span><span class="si">%s</span><span class="s1">_homolog&#39;</span> <span class="o">%</span> <span class="n">target_organism</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="nb">list</span><span class="p">(</span>
        <span class="n">biomart_query</span><span class="p">(</span>
            <span class="n">attrs</span> <span class="o">=</span> <span class="n">homolog_attrs</span><span class="p">,</span>
            <span class="n">filters</span> <span class="o">=</span> <span class="n">filters</span><span class="p">,</span>
            <span class="n">transcript</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">gene</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">peptide</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_gene_ensembl&#39;</span> <span class="o">%</span> <span class="n">source_organism</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span></div>



<div class="viewcode-block" id="biomart_microarray_types"><a class="viewcode-back" href="../../../_autosummary/pypath.inputs.biomart.biomart_microarray_types.html#pypath.inputs.biomart.biomart_microarray_types">[docs]</a><span class="k">def</span> <span class="nf">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Retrieves a list of available microarray types for an organism.</span>

<span class="sd">    Args</span>
<span class="sd">        organism (int,str): Name or ID of an organism.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>

    <span class="n">url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;ensembl&#39;</span><span class="p">][</span><span class="s1">&#39;arraytypes&#39;</span><span class="p">]</span> <span class="o">%</span> <span class="n">organism</span>
    <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">c</span><span class="o">.</span><span class="n">result</span><span class="p">)</span>

    <span class="n">_</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">r</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">label</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1"> </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span>
                <span class="n">r</span><span class="p">[</span><span class="s1">&#39;vendor&#39;</span><span class="p">],</span>
                <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;[-_]&#39;</span><span class="p">,</span> <span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">[</span><span class="s1">&#39;array&#39;</span><span class="p">]),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">result</span>
    <span class="p">]</span>

    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="biomart_microarray"><a class="viewcode-back" href="../../../_autosummary/pypath.inputs.biomart.biomart_microarray.html#pypath.inputs.biomart.biomart_microarray">[docs]</a><span class="k">def</span> <span class="nf">biomart_microarray</span><span class="p">(</span>
        <span class="n">array_type</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">organism</span> <span class="o">=</span> <span class="mi">9606</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Microarray probe identifier mappings.</span>

<span class="sd">    Args</span>
<span class="sd">        array_type (str): The microarray model, as shown on the BioMart</span>
<span class="sd">            webpage, or the corresponding code. For a full list of available</span>
<span class="sd">            identifiers see the ``biomart_microarray_types``.</span>
<span class="sd">        gene (bool): Include the mapping to Ensembl gene IDs.</span>
<span class="sd">        transcript (bool): Include the mapping to Ensembl transcript IDs.</span>
<span class="sd">        peptide (bool): Include the mapping to Ensembl peptide IDs.</span>
<span class="sd">        organism (int,str): Name or ID of an organism.</span>

<span class="sd">    Returns</span>
<span class="sd">        A dictionary with Ensembl gene, transcript and peptide IDs as keys</span>
<span class="sd">        and sets of microarray probe IDs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">organism</span><span class="p">)</span>
    <span class="n">array_types</span> <span class="o">=</span> <span class="n">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">)</span>
    <span class="n">array_types</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">at</span> <span class="ow">in</span> <span class="n">array_types</span>
    <span class="p">}</span>

    <span class="n">_array_type</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">array_type</span><span class="o">.</span>
        <span class="n">lower</span><span class="p">()</span><span class="o">.</span>
        <span class="n">replace</span><span class="p">(</span><span class="s1">&#39;probe&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">)</span><span class="o">.</span>
        <span class="n">strip</span><span class="p">()</span><span class="o">.</span>
        <span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">)</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">_array_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">array_types</span><span class="p">:</span>

        <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;No such array type in Ensembl BioMart: `</span><span class="si">%s</span><span class="s1">` (</span><span class="si">%s</span><span class="s1">).&#39;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">array_type</span><span class="p">,</span>
            <span class="n">_array_type</span>
        <span class="p">)</span>
        <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

    <span class="n">attrs</span> <span class="o">=</span> <span class="p">[</span><span class="n">_array_type</span><span class="p">]</span>
    <span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_gene_ensembl&#39;</span> <span class="o">%</span> <span class="n">organism</span>

    <span class="n">biomart_result</span> <span class="o">=</span> <span class="n">biomart_query</span><span class="p">(</span>
        <span class="n">attrs</span> <span class="o">=</span> <span class="n">attrs</span><span class="p">,</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">,</span>
        <span class="n">dataset</span> <span class="o">=</span> <span class="n">dataset</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>
    <span class="n">_locals</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span>
    <span class="n">ensembl_attrs</span> <span class="o">=</span> <span class="nb">tuple</span><span class="p">(</span>
        <span class="n">attr</span>
        <span class="k">for</span> <span class="n">attr</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;gene&#39;</span><span class="p">,</span> <span class="s1">&#39;transcript&#39;</span><span class="p">,</span> <span class="s1">&#39;peptide&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">_locals</span><span class="p">[</span><span class="n">attr</span><span class="p">]</span>
    <span class="p">)</span>

    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">biomart_result</span><span class="p">:</span>

        <span class="n">array_probe_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">_array_type</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">array_probe_id</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">ensembl_attr</span> <span class="ow">in</span> <span class="n">ensembl_attrs</span><span class="p">:</span>

                <span class="n">ensembl_id</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="s1">&#39;ensembl_</span><span class="si">%s</span><span class="s1">_id&#39;</span> <span class="o">%</span> <span class="n">ensembl_attr</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">ensembl_id</span><span class="p">:</span>

                    <span class="n">result</span><span class="p">[</span><span class="n">ensembl_id</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">array_probe_id</span><span class="p">)</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>


<div class="viewcode-block" id="biomart_microarrays"><a class="viewcode-back" href="../../../_autosummary/pypath.inputs.biomart.biomart_microarrays.html#pypath.inputs.biomart.biomart_microarrays">[docs]</a><span class="k">def</span> <span class="nf">biomart_microarrays</span><span class="p">(</span>
        <span class="n">organism</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">vendor</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">transcript</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">peptide</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Microarray probe identifier mappings for multiple microarrays.</span>
<span class="sd">    Retrieves probe mappings for all array types for one organism,</span>
<span class="sd">    optionally limited to one or more array vendors. Note: depending</span>
<span class="sd">    on the number of array models, it can take minutes to download</span>
<span class="sd">    the data.</span>

<span class="sd">    Args</span>
<span class="sd">        organism (int,str): Name or ID of an organism.</span>
<span class="sd">        vendor (str,set): One or more vendors. None means all vendors. For</span>
<span class="sd">            human, possible values are AFFY, ILLUMINA, AGILENT, CODELINK</span>
<span class="sd">            and PHALANX.</span>
<span class="sd">        gene (bool): Include the mapping to Ensembl gene IDs.</span>
<span class="sd">        transcript (bool): Include the mapping to Ensembl transcript IDs.</span>
<span class="sd">        peptide (bool): Include the mapping to Ensembl peptide IDs.</span>

<span class="sd">    Returns</span>
<span class="sd">        A dictionary with Ensembl gene, transcript and peptide IDs as keys</span>
<span class="sd">        and sets of tuples with microarray types and probe IDs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">record</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
        <span class="s1">&#39;Probe&#39;</span><span class="p">,</span>
        <span class="p">(</span>
            <span class="s1">&#39;array&#39;</span><span class="p">,</span>
            <span class="s1">&#39;probe&#39;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">array_types</span> <span class="o">=</span> <span class="n">biomart_microarray_types</span><span class="p">(</span><span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">)</span>
    <span class="n">vendor</span> <span class="o">=</span> <span class="p">{</span><span class="n">v</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="n">vendor</span><span class="p">)}</span>

    <span class="n">result</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">at</span> <span class="ow">in</span> <span class="n">array_types</span><span class="p">:</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">vendor</span> <span class="ow">or</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;vendor&#39;</span><span class="p">]</span> <span class="ow">in</span> <span class="n">vendor</span><span class="p">:</span>

            <span class="n">probe_map</span> <span class="o">=</span> <span class="n">biomart_microarray</span><span class="p">(</span>
                <span class="n">array_type</span> <span class="o">=</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">],</span>
                <span class="n">organism</span> <span class="o">=</span> <span class="n">organism</span><span class="p">,</span>
                <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
                <span class="n">transcript</span> <span class="o">=</span> <span class="n">transcript</span><span class="p">,</span>
                <span class="n">peptide</span> <span class="o">=</span> <span class="n">peptide</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">gene_id</span><span class="p">,</span> <span class="n">probe_ids</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="n">probe_map</span><span class="p">):</span>

                <span class="k">for</span> <span class="n">probe_id</span> <span class="ow">in</span> <span class="n">probe_ids</span><span class="p">:</span>

                    <span class="n">result</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span>
                        <span class="n">record</span><span class="p">(</span>
                            <span class="n">array</span> <span class="o">=</span> <span class="n">at</span><span class="p">[</span><span class="s1">&#39;label&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">),</span>
                            <span class="n">probe</span> <span class="o">=</span> <span class="n">probe_id</span><span class="p">,</span>
                        <span class="p">)</span>
                    <span class="p">)</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span><span class="n">result</span><span class="p">)</span></div>
</pre></div>
            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

          <div class="sidebar-resize-handle"></div>

  <div>
    <h3><a href="../../../index.html">Table of Contents</a></h3>
    <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html"><em>pypath:</em> A Python module for molecular signaling prior knowledge processing</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../webservice.html">Webservice</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../releasehistory.html">Release history</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../_autosummary/pypath.html">pypath</a></li>
</ul>

  </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
<div id="ethical-ad-placement"></div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <nav class="relbar">
    </nav>

    <footer role="contentinfo">
      &#169; Copyright 2022, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Erva Ulusoy, Ahmet Rifaioglu, Charles Tapley Hoyt.
      Created using <a class="reference external" href="https://www.sphinx-doc.org/">Sphinx</a> 5.3.0.
      <a class="reference external" href="https://insipid-sphinx-theme.readthedocs.io/">Insipid Theme</a>.
    </footer>
  </body>
</html>