
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pypath.utils.homology &#8212; pypath 0.14.31 documentation</title>
    <link rel="stylesheet" href="../../../_static/basic.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/insipid.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/css/pypathdocs.css" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/sphinx_highlight.js"></script>
    <script defer src="../../../_static/insipid.js"></script>
    <script defer src="../../../_static/insipid-sidebar.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-capable" content="yes">
  </head><body>
    <script type="text/javascript">
        document.body.classList.add('js');
    </script>
    <input type="checkbox" id="sidebar-checkbox" style="display: none;">
    <label id="overlay" for="sidebar-checkbox"></label>
    <div class="sidebar-resize-handle"></div>
    <script type="text/javascript">
        try {
            let sidebar = localStorage.getItem('sphinx-sidebar');
            const sidebar_width = localStorage.getItem('sphinx-sidebar-width');
            if (sidebar_width) {
                document.documentElement.style.setProperty('--sidebar-width', sidebar_width);
            }
            // show sidebar on wide screen
            if (!sidebar && window.matchMedia('(min-width: 100rem)').matches) {
                sidebar = 'visible';
                // NB: We don't store the value in localStorage!
            }
            if (sidebar === 'visible') {
                document.getElementById('sidebar-checkbox').checked = true;
            }
        } catch(e) {
            console.info(e);
        }
    </script>
    <header id="topbar-placeholder">
      <div id="topbar">
        <div id="titlebar">
          <div class="buttons">
            <label for="sidebar-checkbox" id="sidebar-button" role="button" tabindex="0" aria-controls="sphinxsidebar" accesskey="M">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"/></svg>
            </label>
          </div>
          <div class="title">
            <a class="parent" href="../../pypath.html" accesskey="U">pypath</a>
            <a class="top" href="#">pypath.utils.homology</a>
          </div>
          <div class="buttons">
            <button id="fullscreen-button" type="button" aria-hidden="true">
              <span class="enable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M212.686 315.314L120 408l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C10.697 480 0 469.255 0 456V344c0-21.382 25.803-32.09 40.922-16.971L72 360l92.686-92.686c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.249 6.248 6.249 16.378 0 22.627zm22.628-118.628L328 104l-32.922-31.029C279.958 57.851 290.666 32 312.048 32h112C437.303 32 448 42.745 448 56v112c0 21.382-25.803 32.09-40.922 16.971L376 152l-92.686 92.686c-6.248 6.248-16.379 6.248-22.627 0l-25.373-25.373c-6.249-6.248-6.249-16.378 0-22.627z"/></svg>
              </span>
              <span class="disable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"/></svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </header>
    <nav>
    </nav>

    <nav class="relbar">
    </nav>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
<h1>Source code for pypath.utils.homology</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>

<span class="c1">#</span>
<span class="c1">#  This file is part of the `pypath` python module</span>
<span class="c1">#  Helps to translate from the mouse data to human data</span>
<span class="c1">#</span>
<span class="c1">#  Copyright</span>
<span class="c1">#  2014-2022</span>
<span class="c1">#  EMBL, EMBL-EBI, Uniklinik RWTH Aachen, Heidelberg University</span>
<span class="c1">#</span>
<span class="c1">#  Authors: Dénes Türei (turei.denes@gmail.com)</span>
<span class="c1">#           Nicolàs Palacio</span>
<span class="c1">#           Sebastian Lobentanzer</span>
<span class="c1">#           Erva Ulusoy</span>
<span class="c1">#           Olga Ivanova</span>
<span class="c1">#           Ahmet Rifaioglu</span>
<span class="c1">#</span>
<span class="c1">#  Distributed under the GPLv3 License.</span>
<span class="c1">#  See accompanying file LICENSE.txt or copy at</span>
<span class="c1">#      http://www.gnu.org/licenses/gpl-3.0.html</span>
<span class="c1">#</span>
<span class="c1">#  Website: http://pypath.omnipathdb.org/</span>
<span class="c1">#</span>

<span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">annotations</span>

<span class="kn">from</span> <span class="nn">future.utils</span> <span class="kn">import</span> <span class="n">iteritems</span>
<span class="kn">from</span> <span class="nn">past.builtins</span> <span class="kn">import</span> <span class="n">xrange</span><span class="p">,</span> <span class="nb">range</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Literal</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">importlib</span> <span class="k">as</span> <span class="nn">imp</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">copy</span>

<span class="kn">import</span> <span class="nn">timeloop</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">import</span> <span class="nn">pypath.utils.mapping</span> <span class="k">as</span> <span class="nn">mapping</span>
<span class="kn">import</span> <span class="nn">pypath.share.common</span> <span class="k">as</span> <span class="nn">common</span>
<span class="kn">import</span> <span class="nn">pypath.internals.intera</span> <span class="k">as</span> <span class="nn">intera</span>
<span class="kn">import</span> <span class="nn">pypath.resources.urls</span> <span class="k">as</span> <span class="nn">urls</span>
<span class="kn">import</span> <span class="nn">pypath.share.curl</span> <span class="k">as</span> <span class="nn">curl</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.uniprot</span> <span class="k">as</span> <span class="nn">uniprot_input</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.homologene</span> <span class="k">as</span> <span class="nn">homologene_input</span>
<span class="kn">import</span> <span class="nn">pypath.inputs.biomart</span> <span class="k">as</span> <span class="nn">biomart</span>
<span class="kn">import</span> <span class="nn">pypath.utils.seq</span> <span class="k">as</span> <span class="nn">_se</span>
<span class="kn">import</span> <span class="nn">pypath.share.session</span> <span class="k">as</span> <span class="nn">session</span>
<span class="kn">import</span> <span class="nn">pypath.share.settings</span> <span class="k">as</span> <span class="nn">settings</span>
<span class="kn">import</span> <span class="nn">pypath.utils.taxonomy</span> <span class="k">as</span> <span class="nn">taxonomy</span>
<span class="kn">import</span> <span class="nn">pypath.share.cache</span> <span class="k">as</span> <span class="nn">cache_mod</span>

<span class="n">_homology_cleanup_timeloop</span> <span class="o">=</span> <span class="n">timeloop</span><span class="o">.</span><span class="n">Timeloop</span><span class="p">()</span>
<span class="n">_homology_cleanup_timeloop</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="mi">9999</span><span class="p">)</span>

<span class="n">_logger</span> <span class="o">=</span> <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">(</span><span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;homology&#39;</span><span class="p">)</span>
<span class="n">_log</span> <span class="o">=</span> <span class="n">_logger</span><span class="o">.</span><span class="n">_log</span>


<div class="viewcode-block" id="Ortholog"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.Ortholog.html#pypath.utils.homology.Ortholog">[docs]</a><span class="k">class</span> <span class="nc">Ortholog</span><span class="p">(</span>
        <span class="n">collections</span><span class="o">.</span><span class="n">namedtuple</span><span class="p">(</span>
            <span class="s1">&#39;OrthologBase&#39;</span><span class="p">,</span>
            <span class="p">(</span>
                <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                <span class="s1">&#39;resource&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensembl_hc&#39;</span><span class="p">,</span>
                <span class="s1">&#39;ensembl_type&#39;</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
    <span class="p">):</span>

    <span class="k">def</span> <span class="fm">__new__</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span>
            <span class="n">uniprot</span><span class="p">,</span>
            <span class="n">resource</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_type</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="p">):</span>

        <span class="k">return</span> <span class="nb">super</span><span class="p">(</span><span class="n">Ortholog</span><span class="p">,</span> <span class="bp">cls</span><span class="p">)</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span>
            <span class="bp">cls</span><span class="p">,</span>
            <span class="n">uniprot</span><span class="p">,</span>
            <span class="n">resource</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="n">ensembl_hc</span><span class="p">,</span>
            <span class="n">ensembl_type</span> <span class="o">=</span> <span class="n">ensembl_type</span><span class="p">,</span>
        <span class="p">)</span>


    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">uniprot</span>


    <span class="k">def</span> <span class="fm">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="s1">&#39;&lt;Ortholog </span><span class="si">%s</span><span class="s1"> (</span><span class="si">%s</span><span class="s1">)&gt;&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">uniprot</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">resource</span><span class="p">)</span>


    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="fm">__str__</span><span class="p">()</span>


    <span class="k">def</span> <span class="fm">__hash__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__hash__</span><span class="p">()</span></div>


<div class="viewcode-block" id="HomologyManager"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager">[docs]</a><span class="k">class</span> <span class="nc">HomologyManager</span><span class="p">(</span><span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">):</span>

    <span class="n">TRANSLATION_PARAM</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s1">&#39;homologene&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl_hc&#39;</span><span class="p">,</span>
        <span class="s1">&#39;ensembl_types&#39;</span><span class="p">,</span>
    <span class="p">)</span>

<div class="viewcode-block" id="HomologyManager.__init__"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">cleanup_period</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
            <span class="n">lifetime</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>

        <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;homology&#39;</span><span class="p">)</span>


        <span class="nd">@_homology_cleanup_timeloop</span><span class="o">.</span><span class="n">job</span><span class="p">(</span>
            <span class="n">interval</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">timedelta</span><span class="p">(</span>
                <span class="n">seconds</span> <span class="o">=</span> <span class="n">cleanup_period</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="k">def</span> <span class="nf">_cleanup</span><span class="p">():</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_remove_expired</span><span class="p">()</span>


        <span class="n">_homology_cleanup_timeloop</span><span class="o">.</span><span class="n">start</span><span class="p">(</span><span class="n">block</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">lifetime</span> <span class="o">=</span> <span class="n">lifetime</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">tables</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cachedir</span> <span class="o">=</span> <span class="n">cache_mod</span><span class="o">.</span><span class="n">get_cachedir</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;homologene&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl_hc&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl_types&#39;</span><span class="p">):</span>

            <span class="nb">setattr</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">param</span><span class="p">,</span>
                <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;homology_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">param</span><span class="p">)</span>
                    <span class="k">if</span> <span class="nb">locals</span><span class="p">()[</span><span class="n">param</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                <span class="nb">locals</span><span class="p">()[</span><span class="n">param</span><span class="p">]</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ortholog_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;ortholog_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span><span class="s1">&#39;HomologyManager has been created.&#39;</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">which_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span> <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

        <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="n">only_swissprot</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>


    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>

        <span class="n">cachefile</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">-homology-</span><span class="si">%u</span><span class="s1">-</span><span class="si">%u</span><span class="s1">.pickle&#39;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">common</span><span class="o">.</span><span class="n">md5</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">key</span><span class="p">)),</span>
            <span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
        <span class="p">)</span>
        <span class="n">cachefile</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">cachedir</span><span class="p">,</span> <span class="n">cachefile</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">cachefile</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="n">cachefile</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">homologene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">homologene</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">ensembl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">ensembl_hc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_hc</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;Homology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> (only SwissProt: </span><span class="si">%s</span><span class="s1">) &#39;</span>
                <span class="s1">&#39;has been loaded from `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">key</span> <span class="o">+</span> <span class="p">(</span><span class="n">cachefile</span><span class="p">,))</span>
            <span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_load</span><span class="p">(</span><span class="n">key</span><span class="p">)</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> <span class="nb">open</span><span class="p">(</span><span class="n">cachefile</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">))</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;Homology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> (only SwissProt: </span><span class="si">%s</span><span class="s1">) &#39;</span>
                <span class="s1">&#39;has been saved to `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">key</span> <span class="o">+</span> <span class="p">(</span><span class="n">cachefile</span><span class="p">,))</span>
            <span class="p">)</span>


    <span class="k">def</span> <span class="nf">_load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span>

        <span class="k">return</span> <span class="n">ProteinHomology</span><span class="p">(</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">key</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">homologene</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_hc</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span><span class="p">,</span>
        <span class="p">)</span>


<div class="viewcode-block" id="HomologyManager.translate"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager.translate">[docs]</a>    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">source_id</span><span class="p">,</span>
            <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For one or more UniProt ID of the source organism returns all</span>
<span class="sd">        orthologs from the target organism.</span>

<span class="sd">        Args</span>
<span class="sd">            source_id (str,list): UniProt ID of one or more protein in the</span>
<span class="sd">                source organism.</span>
<span class="sd">            target (int,str): The target organism.</span>
<span class="sd">            source (int,str): The source organism.</span>
<span class="sd">            homologene (bool): Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl (bool): Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc (bool): Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types (list): The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type (str): Identifier type to translate from and to.</span>

<span class="sd">        Returns</span>
<span class="sd">            Set of UniProt IDs of homologous proteins in the target taxon.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>

        <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>
        <span class="n">homologene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">homologene</span> <span class="k">if</span> <span class="n">homologene</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">homologene</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span> <span class="k">if</span> <span class="n">ensembl</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_hc</span> <span class="k">if</span> <span class="n">ensembl_hc</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl_hc</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="k">if</span> <span class="n">ensembl_types</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl_types</span>
        <span class="p">)</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="n">ensembl_types</span><span class="p">)</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ortholog_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;ortholog_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ensembl_types</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">table</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="n">source_id</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="o">**</span><span class="n">param</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="HomologyManager.get_dict"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager.get_dict">[docs]</a>    <span class="k">def</span> <span class="nf">get_dict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">]]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a dictionary for one source organism and ID type.</span>

<span class="sd">        Args</span>
<span class="sd">            target:</span>
<span class="sd">                Tha organism to translate to.</span>
<span class="sd">            source:</span>
<span class="sd">                The organism to translate from.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type:</span>
<span class="sd">                Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">                first element will be used for keys (source organism) and</span>
<span class="sd">                the second for the values (target organism). By default</span>
<span class="sd">                UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">        Return</span>
<span class="sd">            A dict with identifiers of the source organism as keys, and</span>
<span class="sd">            sets of their orthologs as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>

        <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">table</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">param</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="HomologyManager.get_df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager.get_df">[docs]</a>    <span class="k">def</span> <span class="nf">get_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a data frame for one source organism and ID type.</span>

<span class="sd">        Args</span>
<span class="sd">            target:</span>
<span class="sd">                The target organism.</span>
<span class="sd">            source:</span>
<span class="sd">                The source organism.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type:</span>
<span class="sd">                Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">                first element will be used for keys (source organism) and</span>
<span class="sd">                the second for the values (target organism). By default</span>
<span class="sd">                UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">        Return</span>
<span class="sd">            A data frame with pairs of orthologous identifiers,</span>
<span class="sd">            in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>

        <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">table</span><span class="o">.</span><span class="n">df</span><span class="p">(</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
            <span class="o">**</span><span class="n">param</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="HomologyManager.translate_df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.HomologyManager.html#pypath.utils.homology.HomologyManager.translate_df">[docs]</a>    <span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">df</span><span class="p">,</span>
            <span class="n">target</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">cols</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate columns in a data frame.</span>

<span class="sd">        Args</span>
<span class="sd">            df:</span>
<span class="sd">                A data frame.</span>
<span class="sd">            cols:</span>
<span class="sd">                One or more columns to be translated. It can be a single</span>
<span class="sd">                column name, an iterable of column names or a dict where</span>
<span class="sd">                keys are column names and values are ID types. Except this</span>
<span class="sd">                last case, identifiers are assumed to be UniProt.</span>
<span class="sd">            source:</span>
<span class="sd">                The organism to translate from.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Same as providing a dict to ``cols``, but beware, keys</span>
<span class="sd">                (column names) can not match existing argument names of</span>
<span class="sd">                this function.</span>

<span class="sd">        Return</span>
<span class="sd">            A data frame with the same column layout as the input, and the</span>
<span class="sd">            identifiers translated as demanded. Rows that could not be</span>
<span class="sd">            translated are omitted.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>

        <span class="n">table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">which_table</span><span class="p">(</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">param</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="nb">locals</span><span class="p">())</span>

        <span class="k">return</span> <span class="n">table</span><span class="o">.</span><span class="n">translate_df</span><span class="p">(</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">,</span>
            <span class="n">cols</span> <span class="o">=</span> <span class="n">cols</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="o">**</span><span class="n">param</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span>
        <span class="p">)</span></div>



    <span class="k">def</span> <span class="nf">_translation_param</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>

        <span class="n">param</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">name</span><span class="p">,</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">)</span> <span class="k">if</span> <span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">loc</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">TRANSLATION_PARAM</span>
        <span class="p">)</span>

        <span class="n">param</span><span class="p">[</span><span class="s1">&#39;ensembl_types&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="n">param</span><span class="p">[</span><span class="s1">&#39;ensembl_types&#39;</span><span class="p">])</span>
        <span class="n">param</span><span class="p">[</span><span class="s1">&#39;ensembl_types&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ortholog_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;ortholog_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">param</span><span class="p">[</span><span class="s1">&#39;ensembl_types&#39;</span><span class="p">]</span>
        <span class="p">}</span>

        <span class="k">return</span> <span class="n">param</span>


    <span class="k">def</span> <span class="nf">_remove_expired</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">last_used</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>

            <span class="k">if</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">last_used</span> <span class="o">&gt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">lifetime</span> <span class="ow">and</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                    <span class="s1">&#39;Removing homology table from taxon </span><span class="si">%u</span><span class="s1"> to </span><span class="si">%u</span><span class="s1"> &#39;</span>
                    <span class="s1">&#39;(only SwissProt: </span><span class="si">%s</span><span class="s1">)&#39;</span> <span class="o">%</span> <span class="n">key</span>
                <span class="p">)</span>

                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">tables</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
                <span class="k">del</span> <span class="bp">self</span><span class="o">.</span><span class="n">expiry</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>


    <span class="k">def</span> <span class="fm">__del__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">_homology_cleanup_timeloop</span><span class="p">,</span> <span class="s1">&#39;stop&#39;</span><span class="p">):</span>

            <span class="n">_homology_cleanup_timeloop</span><span class="o">.</span><span class="n">stop</span><span class="p">()</span></div>


<div class="viewcode-block" id="SequenceContainer"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.SequenceContainer.html#pypath.utils.homology.SequenceContainer">[docs]</a><span class="k">class</span> <span class="nc">SequenceContainer</span><span class="p">(</span><span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="p">):</span>


<div class="viewcode-block" id="SequenceContainer.__init__"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.SequenceContainer.html#pypath.utils.homology.SequenceContainer.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">preload_seq</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">isoforms</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is an object to store sequences of multiple</span>
<span class="sd">        organisms and select the appropriate one.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_logger&#39;</span><span class="p">):</span>

            <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;homology&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">seq_isoforms</span> <span class="o">=</span> <span class="n">isoforms</span>

        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">preload_seq</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">load_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;seq&#39;</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">seq</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">taxon</span> <span class="o">=</span> <span class="n">taxon</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">ncbi_tax_id</span>

        <span class="k">if</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">]</span> <span class="o">=</span> <span class="n">_se</span><span class="o">.</span><span class="n">swissprot_seq</span><span class="p">(</span>
                <span class="n">organism</span> <span class="o">=</span> <span class="n">taxon</span><span class="p">,</span>
                <span class="n">isoforms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_isoforms</span>
            <span class="p">)</span>


    <span class="k">def</span> <span class="nf">get_seq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">,</span> <span class="n">taxon</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">taxon</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">taxon</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">]:</span>

                <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="n">taxon</span><span class="p">][</span><span class="n">protein</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">for</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">seq</span><span class="p">):</span>

                <span class="k">if</span> <span class="n">protein</span> <span class="ow">in</span> <span class="n">seq</span><span class="p">:</span>

                    <span class="k">return</span> <span class="n">seq</span><span class="p">[</span><span class="n">protein</span><span class="p">]</span></div>


<div class="viewcode-block" id="Proteomes"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.Proteomes.html#pypath.utils.homology.Proteomes">[docs]</a><span class="k">class</span> <span class="nc">Proteomes</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>


<div class="viewcode-block" id="Proteomes.__init__"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.Proteomes.html#pypath.utils.homology.Proteomes.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">preload_prot</span> <span class="o">=</span> <span class="p">[],</span> <span class="n">swissprot_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_taxonomy&#39;</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">taxon</span> <span class="ow">in</span> <span class="n">preload_prot</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">swissprot_only</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">load_proteome</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">taxon</span><span class="p">,</span> <span class="n">swissprot_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

        <span class="n">key</span> <span class="o">=</span> <span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="n">swissprot_only</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">uniprot_input</span><span class="o">.</span><span class="n">all_uniprots</span><span class="p">(</span><span class="o">*</span><span class="n">key</span><span class="p">))</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[</span><span class="n">key</span><span class="p">]:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span><span class="p">[</span><span class="n">protein</span><span class="p">]</span> <span class="o">=</span> <span class="n">key</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">swissprot_only</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="n">taxon</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">get_taxon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">,</span> <span class="n">swissprot_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">swissprot_only</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">is_swissprot</span><span class="p">(</span><span class="n">protein</span><span class="p">):</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span><span class="p">[</span><span class="n">protein</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>


    <span class="k">def</span> <span class="nf">get_taxon_trembl</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_protein</span><span class="p">(</span><span class="n">protein</span><span class="p">):</span>

            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span><span class="p">[</span><span class="n">protein</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>


    <span class="k">def</span> <span class="nf">has_protein</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">return</span> <span class="n">protein</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span>


    <span class="k">def</span> <span class="nf">is_swissprot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein</span><span class="p">):</span>

        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">has_protein</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">_taxonomy</span><span class="p">[</span><span class="n">protein</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span></div>


<div class="viewcode-block" id="ProteinHomology"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology">[docs]</a><span class="k">class</span> <span class="nc">ProteinHomology</span><span class="p">(</span><span class="n">Proteomes</span><span class="p">):</span>


<div class="viewcode-block" id="ProteinHomology.__init__"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">homologene</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
            <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This class translates between homologous UniProt IDs of two organisms</span>
<span class="sd">        based on NCBI HomoloGene and Ensembl data. In case of HomoloGene,</span>
<span class="sd">        the UniProt-UniProt translation table is created by translating the</span>
<span class="sd">        source organism UniProts to RefSeq and Entrez IDs, finding the</span>
<span class="sd">        homologues (orthologues) for these IDs, and then translating them</span>
<span class="sd">        to the target organism UniProt IDs. In case of Ensembl, we obtain</span>
<span class="sd">        data with Ensembl protein identifiers and translate those to UniProt.</span>

<span class="sd">        Args</span>
<span class="sd">            target (int): NCBI Taxonomy ID of the target organism.</span>
<span class="sd">            source (int): NCBI Taxonomy ID of the default source organism.</span>
<span class="sd">                Multiple source organisms can be used on the same instance.</span>
<span class="sd">            only_swissprot (bool): Use only SwissProt IDs.</span>
<span class="sd">            homologene (bool): Use homology information from NCBI HomoloGene.</span>
<span class="sd">            ensembl (bool): Use homology information from Ensembl.</span>
<span class="sd">            ensembl_hc (bool): Use only the high confidence</span>
<span class="sd">                orthology relations from Ensembl.</span>
<span class="sd">            ensembl_types (list): Ensembl orthology relation types to use.</span>
<span class="sd">                Possible values are `one2one`, `one2many` and `many2many`.</span>
<span class="sd">                By default only `one2one` is used.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">target</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="n">source</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_default_source</span> <span class="o">=</span> <span class="mi">9606</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_default_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">param</span> <span class="ow">in</span> <span class="p">(</span><span class="s1">&#39;homologene&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl_hc&#39;</span><span class="p">,</span> <span class="s1">&#39;ensembl_types&#39;</span><span class="p">):</span>

            <span class="nb">setattr</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">param</span><span class="p">,</span>
                <span class="n">settings</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;homology_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">param</span><span class="p">)</span>
                    <span class="k">if</span> <span class="nb">locals</span><span class="p">()[</span><span class="n">param</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span>
                <span class="nb">locals</span><span class="p">()[</span><span class="n">param</span><span class="p">]</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ortholog_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;ortholog_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span>
        <span class="p">}</span>

        <span class="n">Proteomes</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">source</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">source</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">reload</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="n">modname</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__module__</span>
        <span class="n">mod</span> <span class="o">=</span> <span class="nb">__import__</span><span class="p">(</span><span class="n">modname</span><span class="p">,</span> <span class="n">fromlist</span><span class="o">=</span><span class="p">[</span><span class="n">modname</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;.&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">imp</span><span class="o">.</span><span class="n">reload</span><span class="p">(</span><span class="n">mod</span><span class="p">)</span>
        <span class="n">new</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">mod</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span>
        <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;__class__&#39;</span><span class="p">,</span> <span class="n">new</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">load</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">homologene</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_homologene</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load_ensembl</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">ensure_source_taxon</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">):</span>

        <span class="k">if</span> <span class="n">source</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">set_default_source</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>


    <span class="k">def</span> <span class="nf">get_source</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>

        <span class="n">source</span> <span class="o">=</span> <span class="n">source</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">source</span> <span class="ow">or</span> <span class="bp">self</span><span class="o">.</span><span class="n">_default_source</span>
        <span class="n">_source</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ncbi_tax_id</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">_source</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="s1">&#39;Unknown organism: `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="nb">str</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="k">return</span> <span class="n">_source</span>


<div class="viewcode-block" id="ProteinHomology.translate"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.translate">[docs]</a>    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">protein</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">list</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        For one UniProt ID of the source organism returns all orthologues</span>
<span class="sd">        from the target organism.</span>

<span class="sd">        Args</span>
<span class="sd">            protein:</span>
<span class="sd">                UniProt ID of one or more protein in the source organism.</span>
<span class="sd">            source:</span>
<span class="sd">                The source organism.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type:</span>
<span class="sd">                Identifier type to translate from and to.</span>

<span class="sd">        Returns</span>
<span class="sd">            Set of UniProt IDs of homologous proteins in the target taxon.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">protein</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">(</span><span class="n">protein</span><span class="p">,)</span>
                <span class="k">if</span> <span class="nb">hasattr</span><span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="s1">&#39;components&#39;</span><span class="p">)</span> <span class="k">else</span>
            <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">protein</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">source</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s1">&#39;Can not translate without knowing the organism of the input.&#39;</span>
            <span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">homologene</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">homologene</span> <span class="k">if</span> <span class="n">homologene</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">homologene</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span> <span class="k">if</span> <span class="n">ensembl</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_hc</span> <span class="k">if</span> <span class="n">ensembl_hc</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl_hc</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl_types</span> <span class="k">if</span> <span class="n">ensembl_types</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">ensembl_types</span>
        <span class="p">)</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="n">common</span><span class="o">.</span><span class="n">to_set</span><span class="p">(</span><span class="n">ensembl_types</span><span class="p">)</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">x</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;ortholog_&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="p">(</span><span class="s1">&#39;ortholog_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ensembl_types</span>
        <span class="p">}</span>

        <span class="k">if</span> <span class="n">id_type</span> <span class="o">!=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">:</span>

            <span class="n">protein</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                <span class="n">protein</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                <span class="n">target_id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">p</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">protein</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxon</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span>
        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ensure_source_taxon</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">protein</span><span class="p">:</span>

            <span class="n">result</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="n">o</span><span class="o">.</span><span class="n">uniprot</span>
                    <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">][</span><span class="n">p</span><span class="p">]</span>
                    <span class="k">if</span> <span class="p">(</span>
                        <span class="p">(</span>
                            <span class="n">homologene</span> <span class="ow">and</span>
                            <span class="n">o</span><span class="o">.</span><span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;HomoloGene&#39;</span>
                        <span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>
                            <span class="n">ensembl</span> <span class="ow">and</span>
                            <span class="n">o</span><span class="o">.</span><span class="n">resource</span> <span class="o">==</span> <span class="s1">&#39;Ensembl&#39;</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="ow">not</span> <span class="n">ensembl_hc</span> <span class="ow">or</span>
                                <span class="n">o</span><span class="o">.</span><span class="n">ensembl_hc</span>
                            <span class="p">)</span> <span class="ow">and</span>
                            <span class="n">o</span><span class="o">.</span><span class="n">ensembl_type</span> <span class="ow">in</span> <span class="n">ensembl_types</span>
                        <span class="p">)</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">uniprot</span> <span class="ow">in</span> <span class="n">o</span><span class="o">.</span><span class="n">uniprot</span>
                <span class="p">}</span>
            <span class="p">)</span>

        <span class="k">if</span> <span class="n">id_type</span> <span class="o">!=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                <span class="n">result</span><span class="p">,</span>
                <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                <span class="n">target_id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="ProteinHomology.asdict"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.asdict">[docs]</a>    <span class="k">def</span> <span class="nf">asdict</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a dictionary for one source organism and ID type.</span>

<span class="sd">        Args</span>
<span class="sd">            source:</span>
<span class="sd">                The source organism.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type:</span>
<span class="sd">                Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">                first element will be used for keys (source organism) and</span>
<span class="sd">                the second for the values (target organism). By default</span>
<span class="sd">                UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">        Return</span>
<span class="sd">            A dict with identifiers of the source organism as keys, and</span>
<span class="sd">            sets of their orthologs as values.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ensure_source_taxon</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="n">id_type</span> <span class="o">=</span> <span class="p">(</span><span class="n">id_type</span><span class="p">,)</span> <span class="o">*</span> <span class="mi">2</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">id_type</span><span class="p">,</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">id_type</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span>
                <span class="nb">isinstance</span><span class="p">(</span><span class="n">id_type</span><span class="p">,</span> <span class="nb">tuple</span><span class="p">)</span> <span class="ow">and</span>
                <span class="nb">len</span><span class="p">(</span><span class="n">id_type</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="mi">2</span> <span class="ow">and</span>
                <span class="nb">isinstance</span><span class="p">(</span><span class="n">id_type</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="ow">and</span>
                <span class="nb">isinstance</span><span class="p">(</span><span class="n">id_type</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span>
            <span class="p">):</span>

            <span class="n">msg</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s1">&#39;`ProteinHomology.asdict`: &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;wrong format for `id_type`: `</span><span class="si">{</span><span class="n">id_type</span><span class="si">}</span><span class="s1">`.&#39;</span>
            <span class="p">)</span>
            <span class="n">_log</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Creating dict of orthologous genes between organisms &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;`</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s1">` (ID type `</span><span class="si">{</span><span class="n">id_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">`) and &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;`</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">` (ID type `</span><span class="si">{</span><span class="n">id_type</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">`).&#39;</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">id_src</span><span class="p">,</span>
                <span class="n">mapping</span><span class="o">.</span><span class="n">map_names</span><span class="p">(</span>
                    <span class="n">names</span> <span class="o">=</span> <span class="p">{</span><span class="n">o</span><span class="o">.</span><span class="n">uniprot</span> <span class="k">for</span> <span class="n">o</span> <span class="ow">in</span> <span class="n">orthologs</span><span class="p">},</span>
                    <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="n">target_id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                    <span class="n">uniprot_cleanup</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">uniprot_src</span><span class="p">,</span> <span class="n">orthologs</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">]</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">id_src</span> <span class="ow">in</span> <span class="p">(</span>
                <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                    <span class="n">name</span> <span class="o">=</span> <span class="n">uniprot_src</span><span class="p">,</span>
                    <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                    <span class="n">target_id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                    <span class="n">uniprot_cleanup</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span></div>


<div class="viewcode-block" id="ProteinHomology.df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.df">[docs]</a>    <span class="k">def</span> <span class="nf">df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">id_type</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a data frame for one source organism and ID type.</span>

<span class="sd">        Args</span>
<span class="sd">            source:</span>
<span class="sd">                The source organism.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            id_type:</span>
<span class="sd">                Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">                first element will be used for keys (source organism) and</span>
<span class="sd">                the second for the values (target organism). By default</span>
<span class="sd">                UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">                ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">        Return</span>
<span class="sd">            A data frame with pairs of orthologous identifiers,</span>
<span class="sd">            in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Creating translation data frame between &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;organisms `</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s1">` and `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;self&#39;</span><span class="p">)</span>

        <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">asdict</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">(),</span>
                <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;source&#39;</span><span class="p">,</span> <span class="s1">&#39;target&#39;</span><span class="p">],</span>
            <span class="p">)</span><span class="o">.</span>
            <span class="n">explode</span><span class="p">(</span><span class="s1">&#39;target&#39;</span><span class="p">,</span> <span class="n">ignore_index</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span>
            <span class="n">dropna</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span></div>


<div class="viewcode-block" id="ProteinHomology.translate_df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.translate_df">[docs]</a>    <span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
            <span class="n">cols</span><span class="p">:</span> <span class="p">(</span>
                <span class="nb">str</span> <span class="o">|</span>
                <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span>
                <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">|</span>
                <span class="kc">None</span>
            <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
                <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
            <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translate columns in a data frame.</span>

<span class="sd">        Args</span>
<span class="sd">            df:</span>
<span class="sd">                A data frame.</span>
<span class="sd">            cols:</span>
<span class="sd">                One or more columns to be translated. It can be a single</span>
<span class="sd">                column name, an iterable of column names or a dict where</span>
<span class="sd">                keys are column names and values are ID types. Except this</span>
<span class="sd">                last case, identifiers are assumed to be UniProt.</span>
<span class="sd">            source:</span>
<span class="sd">                The source organism.</span>
<span class="sd">            homologene:</span>
<span class="sd">                Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">            ensembl:</span>
<span class="sd">                Use Ensembl data for ortholog lookup.</span>
<span class="sd">            ensembl_hc:</span>
<span class="sd">                Use only high confidence orthology relations</span>
<span class="sd">                from Ensembl. By default it is True. You can also set it</span>
<span class="sd">                by the `ensembl_hc` attribute.</span>
<span class="sd">            ensembl_types:</span>
<span class="sd">                The Ensembl orthology relationship types</span>
<span class="sd">                to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">                `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">                also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">            kwargs:</span>
<span class="sd">                Same as providing a dict to ``cols``, but beware, keys</span>
<span class="sd">                (column names) can not match existing argument names of</span>
<span class="sd">                this function.</span>

<span class="sd">        Return</span>
<span class="sd">            A data frame with the same column layout as the input, and the</span>
<span class="sd">            identifiers translated as demanded. Rows that could not be</span>
<span class="sd">            translated are omitted.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="sa">f</span><span class="s1">&#39;Translating data frame column(s) from &#39;</span>
            <span class="sa">f</span><span class="s1">&#39;organism `</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s1">` to `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
        <span class="p">)</span>

        <span class="n">loc</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span>

        <span class="n">col_order</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">cols</span><span class="p">,</span> <span class="nb">dict</span><span class="p">):</span>

            <span class="n">cols</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">col</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">to_list</span><span class="p">(</span><span class="n">cols</span><span class="p">))</span>

        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span>
        <span class="n">id_types</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>

        <span class="n">transdfs</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span>
                <span class="n">id_type</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="p">(</span>
                    <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
                    <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
                    <span class="o">**</span><span class="bp">self</span><span class="o">.</span><span class="n">_translation_param</span><span class="p">(</span><span class="n">loc</span><span class="p">)</span>
                <span class="p">)</span><span class="o">.</span>
                <span class="n">rename</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="s1">&#39;source&#39;</span><span class="p">:</span> <span class="s1">&#39;pypath_internal_source&#39;</span><span class="p">,</span>
                        <span class="s1">&#39;target&#39;</span><span class="p">:</span> <span class="s1">&#39;pypath_internal_target&#39;</span><span class="p">,</span>
                    <span class="p">},</span>
                    <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">id_type</span> <span class="ow">in</span> <span class="n">id_types</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">col</span><span class="p">,</span> <span class="n">id_type</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

            <span class="n">_log</span><span class="p">(</span>
                <span class="sa">f</span><span class="s1">&#39;Translating `</span><span class="si">{</span><span class="n">id_type</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">` IDs of organism `</span><span class="si">{</span><span class="n">source</span><span class="si">}</span><span class="s1">` &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;in column `</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s1">` to `</span><span class="si">{</span><span class="n">id_type</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">` IDs of &#39;</span>
                <span class="sa">f</span><span class="s1">&#39;organism `</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="si">}</span><span class="s1">`.&#39;</span>
            <span class="p">)</span>

            <span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span>
                    <span class="n">transdfs</span><span class="p">[</span><span class="n">id_type</span><span class="p">]</span><span class="o">.</span>
                        <span class="n">rename</span><span class="p">({</span><span class="s1">&#39;pypath_internal_source&#39;</span><span class="p">:</span> <span class="n">col</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">),</span>
                    <span class="n">on</span> <span class="o">=</span> <span class="n">col</span><span class="p">,</span>
                    <span class="n">how</span> <span class="o">=</span> <span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                <span class="p">)</span><span class="o">.</span>
                <span class="n">drop</span><span class="p">(</span><span class="n">col</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span>
                <span class="n">rename</span><span class="p">({</span><span class="s1">&#39;pypath_internal_target&#39;</span><span class="p">:</span> <span class="n">col</span><span class="p">},</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
            <span class="p">)</span>

        <span class="k">return</span> <span class="n">df</span><span class="p">[</span><span class="n">col_order</span><span class="p">]</span></div>


    <span class="k">def</span> <span class="nf">_translation_param</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">loc</span><span class="p">:</span> <span class="nb">dict</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">loc</span><span class="p">[</span><span class="n">p</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">HomologyManager</span><span class="o">.</span><span class="n">TRANSLATION_PARAM</span>
        <span class="p">)</span>

<div class="viewcode-block" id="ProteinHomology.load_homologene"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.ProteinHomology.html#pypath.utils.homology.ProteinHomology.load_homologene">[docs]</a>    <span class="k">def</span> <span class="nf">load_homologene</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Builds orthology translation table as dict from UniProt to Uniprot,</span>
<span class="sd">        obtained from NCBI HomoloGene data. Uses RefSeq and Entrez IDs for</span>
<span class="sd">        translation.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">source</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="n">hge</span> <span class="o">=</span> <span class="n">homologene_input</span><span class="o">.</span><span class="n">homologene_dict</span><span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">)</span>
        <span class="n">hgr</span> <span class="o">=</span> <span class="n">homologene_input</span><span class="o">.</span><span class="n">homologene_dict</span><span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span> <span class="s1">&#39;refseq&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">load_proteome</span><span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">)</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Loading homology data from NCBI HomoloGene &#39;</span>
            <span class="s1">&#39;between organisms `</span><span class="si">%u</span><span class="s1">` and `</span><span class="si">%u</span><span class="s1">`&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_proteomes</span><span class="p">[(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">only_swissprot</span><span class="p">)]:</span>

            <span class="n">source_e</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">,</span> <span class="n">source</span><span class="p">)</span>
            <span class="n">source_r</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;refseqp&#39;</span><span class="p">,</span> <span class="n">source</span><span class="p">)</span>
            <span class="n">target_u</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">target_r</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="n">target_e</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">source_e</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">hge</span><span class="p">:</span>

                    <span class="n">target_e</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hge</span><span class="p">[</span><span class="n">e</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">source_r</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">hgr</span><span class="p">:</span>

                    <span class="n">target_r</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">hgr</span><span class="p">[</span><span class="n">r</span><span class="p">])</span>

            <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">target_e</span><span class="p">:</span>

                <span class="n">target_u</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">e</span><span class="p">,</span> <span class="s1">&#39;entrez&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">target_r</span><span class="p">:</span>

                <span class="n">target_u</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="nb">set</span><span class="p">(</span>
                        <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span><span class="n">e</span><span class="p">,</span> <span class="s1">&#39;refseqp&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>

            <span class="n">target_u</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">itertools</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span>
                    <span class="o">*</span><span class="nb">map</span><span class="p">(</span>
                        <span class="k">lambda</span> <span class="n">tu</span><span class="p">:</span>
                            <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                                <span class="n">tu</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span>
                            <span class="p">),</span>
                        <span class="n">target_u</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">][</span><span class="n">u</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="n">Ortholog</span><span class="p">(</span><span class="n">u</span><span class="p">,</span> <span class="s1">&#39;HomoloGene&#39;</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">target_u</span>
                <span class="p">}</span>
            <span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">load_ensembl</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">source</span><span class="p">):</span>

        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>
        <span class="n">target_organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
        <span class="n">source_organism</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">ensure_ensembl_name</span><span class="p">(</span><span class="n">source</span><span class="p">)</span>

        <span class="n">_log</span><span class="p">(</span>
            <span class="s1">&#39;Loading homology data from Ensembl &#39;</span>
            <span class="s1">&#39;between organisms `</span><span class="si">%u</span><span class="s1">` and `</span><span class="si">%u</span><span class="s1">`&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
        <span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">target_organism</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">source_organism</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">ensembl</span><span class="p">[</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;No Ensembl homology data available between &#39;</span>
                <span class="s1">&#39;organisms `</span><span class="si">%s</span><span class="s1">` and `</span><span class="si">%s</span><span class="s1">`.&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">source</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="k">return</span>

        <span class="k">if</span> <span class="n">source</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">]</span> <span class="o">=</span> <span class="n">collections</span><span class="o">.</span><span class="n">defaultdict</span><span class="p">(</span><span class="nb">set</span><span class="p">)</span>

        <span class="n">attr_target_ensp</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_homolog_ensembl_peptide&#39;</span> <span class="o">%</span> <span class="n">target_organism</span>
        <span class="n">attr_conf</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_homolog_orthology_confidence&#39;</span> <span class="o">%</span> <span class="n">target_organism</span>
        <span class="n">attr_type</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_homolog_orthology_type&#39;</span> <span class="o">%</span> <span class="n">target_organism</span>

        <span class="n">ensembl_data</span> <span class="o">=</span> <span class="n">biomart</span><span class="o">.</span><span class="n">biomart_homology</span><span class="p">(</span>
            <span class="n">source_organism</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">target_organism</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">ensembl_data</span><span class="p">:</span>

            <span class="n">source_uniprots</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                <span class="n">r</span><span class="o">.</span><span class="n">ensembl_peptide_id</span><span class="p">,</span>
                <span class="s1">&#39;ensp&#39;</span><span class="p">,</span>
                <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                <span class="n">source</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">target_uniprots</span> <span class="o">=</span> <span class="n">mapping</span><span class="o">.</span><span class="n">map_name</span><span class="p">(</span>
                <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_target_ensp</span><span class="p">),</span>
                <span class="s1">&#39;ensp&#39;</span><span class="p">,</span>
                <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">source_uniprots</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">orthologs</span><span class="p">[</span><span class="n">source</span><span class="p">][</span><span class="n">u</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
                    <span class="p">{</span>
                        <span class="n">Ortholog</span><span class="p">(</span>
                            <span class="n">u</span><span class="p">,</span>
                            <span class="s1">&#39;Ensembl&#39;</span><span class="p">,</span>
                            <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_conf</span><span class="p">)</span> <span class="o">==</span> <span class="s1">&#39;1&#39;</span><span class="p">,</span>
                            <span class="n">ensembl_type</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">r</span><span class="p">,</span> <span class="n">attr_type</span><span class="p">),</span>
                        <span class="p">)</span>
                        <span class="k">for</span> <span class="n">u</span> <span class="ow">in</span> <span class="n">target_uniprots</span>
                    <span class="p">}</span>
                <span class="p">)</span></div>


<div class="viewcode-block" id="PtmHomology"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology">[docs]</a><span class="k">class</span> <span class="nc">PtmHomology</span><span class="p">(</span><span class="n">ProteinHomology</span><span class="p">,</span> <span class="n">SequenceContainer</span><span class="p">):</span>


<div class="viewcode-block" id="PtmHomology.__init__"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">target</span><span class="p">,</span>
        <span class="n">source</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">strict</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="p">):</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s1">&#39;_logger&#39;</span><span class="p">):</span>

            <span class="n">session</span><span class="o">.</span><span class="n">Logger</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;homology&#39;</span><span class="p">)</span>

        <span class="n">ProteinHomology</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
            <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">SequenceContainer</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">load_seq</span><span class="p">(</span><span class="n">taxon</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reptm</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;([A-Z\d]{6,10})_([A-Z])(\d*)&#39;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">strict</span> <span class="o">=</span> <span class="n">strict</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ptm_orthology</span><span class="p">()</span></div>


<div class="viewcode-block" id="PtmHomology.translate_site"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology.translate_site">[docs]</a>    <span class="k">def</span> <span class="nf">translate_site</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">protein</span><span class="p">,</span>
            <span class="n">res</span><span class="p">,</span>
            <span class="n">offset</span><span class="p">,</span>
            <span class="n">isoform</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
            <span class="n">typ</span> <span class="o">=</span> <span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span>
            <span class="n">source_taxon</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translates one PTM site.</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">set_default_source</span><span class="p">(</span><span class="n">source_taxon</span><span class="p">)</span>

        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">(</span><span class="n">source_taxon</span><span class="p">)</span>

        <span class="n">sourceptm</span> <span class="o">=</span> <span class="p">(</span><span class="n">protein</span><span class="p">,</span> <span class="n">isoform</span><span class="p">,</span> <span class="n">res</span><span class="p">,</span> <span class="n">offset</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">typ</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_taxon</span><span class="p">(</span><span class="n">protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">)</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">:</span>
            <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">sourceptm</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">result</span>

        <span class="k">if</span> <span class="n">sourceptm</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">:</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">sourceptm</span><span class="p">]:</span>

                <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">sourceptm</span><span class="p">]</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">result</span> <span class="ow">and</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">strict</span><span class="p">:</span>

            <span class="n">tsubs</span> <span class="o">=</span> <span class="n">ProteinHomology</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
                <span class="bp">self</span><span class="p">,</span>
                <span class="n">protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span>
                <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">tsub</span> <span class="ow">in</span> <span class="n">tsubs</span><span class="p">:</span>

                <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="n">tsub</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">se</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">for</span> <span class="n">toffset</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="n">offset</span><span class="p">,</span> <span class="n">offset</span> <span class="o">+</span> <span class="mi">3</span><span class="p">):</span>

                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isoforms</span><span class="p">():</span>

                        <span class="n">tres</span> <span class="o">=</span> <span class="n">se</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">toffset</span><span class="p">,</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">i</span><span class="p">)</span>

                        <span class="k">if</span> <span class="n">tres</span> <span class="o">==</span> <span class="n">res</span><span class="p">:</span>

                            <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">((</span>
                                <span class="n">tsub</span><span class="p">,</span>
                                <span class="n">i</span><span class="p">,</span>
                                <span class="n">tres</span><span class="p">,</span>
                                <span class="n">toffset</span><span class="p">,</span>
                                <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                                <span class="n">typ</span>
                            <span class="p">))</span>

                    <span class="k">if</span> <span class="n">result</span><span class="p">:</span>
                        <span class="k">break</span>

        <span class="k">return</span> <span class="n">result</span></div>


    <span class="k">def</span> <span class="nf">translate_domain</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">domain</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span>
                <span class="nb">map</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                        <span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">(</span>
                            <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">,</span>
                            <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
                        <span class="p">),</span>
                    <span class="n">ProteinHomology</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
                        <span class="bp">self</span><span class="p">,</span>
                        <span class="n">domain</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">identifier</span><span class="p">,</span>
                        <span class="n">source</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_source</span><span class="p">()</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>


    <span class="k">def</span> <span class="nf">translate_ptm</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">ptm</span><span class="p">):</span>

        <span class="n">tptms</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">number</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">residue</span><span class="o">.</span><span class="n">isoform</span><span class="p">,</span>
            <span class="n">ptm</span><span class="o">.</span><span class="n">typ</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">tptms</span><span class="p">:</span>

            <span class="n">se</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get_seq</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

            <span class="k">if</span> <span class="p">(</span><span class="n">se</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isof</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">strict</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="n">res</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">(</span>
                <span class="n">number</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span>
                <span class="n">name</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">region</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">se</span><span class="o">.</span><span class="n">get_region</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="n">se</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">se</span><span class="o">.</span><span class="n">isof</span>
                <span class="k">else</span> <span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="n">mot</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Motif</span><span class="p">(</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">start</span> <span class="o">=</span> <span class="n">start</span><span class="p">,</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">end</span><span class="p">,</span>
                <span class="n">instance</span> <span class="o">=</span> <span class="n">region</span><span class="p">,</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">ptm</span> <span class="o">=</span> <span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">(</span>
                <span class="n">protein</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">motif</span> <span class="o">=</span> <span class="n">mot</span><span class="p">,</span>
                <span class="n">residue</span> <span class="o">=</span> <span class="n">res</span><span class="p">,</span>
                <span class="n">typ</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">5</span><span class="p">],</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">evidences</span> <span class="o">=</span> <span class="n">ptm</span><span class="o">.</span><span class="n">evidences</span><span class="p">,</span>
                <span class="n">ncbi_tax_id</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">target</span><span class="p">,</span>
            <span class="p">)</span>

            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ptm</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span>


    <span class="k">def</span> <span class="nf">translate_domain_motif</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">dmotif</span><span class="p">):</span>

        <span class="n">ds</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain</span><span class="p">(</span><span class="n">dmotif</span><span class="o">.</span><span class="n">domain</span><span class="p">)</span>
        <span class="n">ps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_ptm</span><span class="p">(</span><span class="n">dmotif</span><span class="o">.</span><span class="n">ptm</span><span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span>
                <span class="nb">map</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                        <span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">(</span>
                            <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                            <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                            <span class="n">evidences</span> <span class="o">=</span> <span class="n">dmotif</span><span class="o">.</span><span class="n">evidences</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">itertools</span><span class="o">.</span><span class="n">product</span><span class="p">(</span><span class="n">ds</span><span class="p">,</span> <span class="n">ps</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>


    <span class="k">def</span> <span class="nf">translate_residue</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">residue</span><span class="p">):</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="nb">list</span><span class="p">(</span>
                <span class="nb">map</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">r</span><span class="p">:</span>
                        <span class="n">intera</span><span class="o">.</span><span class="n">Residue</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">isoform</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span>
                        <span class="n">residue</span><span class="o">.</span><span class="n">protein</span><span class="p">,</span>
                        <span class="n">residue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                        <span class="n">residue</span><span class="o">.</span><span class="n">number</span><span class="p">,</span>
                        <span class="n">residue</span><span class="o">.</span><span class="n">isoform</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>


<div class="viewcode-block" id="PtmHomology.translate"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology.translate">[docs]</a>    <span class="k">def</span> <span class="nf">translate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">return_strings</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Translates anything:</span>

<span class="sd">        - one UniProt ID</span>
<span class="sd">        - one PTM provided as tuple of (UniProt, amino acid, offest)</span>
<span class="sd">        - one PTM provided as string (e.g. `P00533_S231`)</span>
<span class="sd">        - instance from pypath.intera: DomainMotif, Domain or Ptm</span>

<span class="sd">        Additional arguments can be isoform and typ (modification type).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">is</span> <span class="nb">tuple</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span><span class="o">*</span><span class="n">x</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">char_types</span><span class="p">:</span>

            <span class="n">ptm</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reptm</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">ptm</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

                <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_site</span><span class="p">(</span><span class="n">ptm</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ptm</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                                             <span class="nb">int</span><span class="p">(</span><span class="n">ptm</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">return_strings</span><span class="p">:</span>

            <span class="n">result</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">r</span><span class="p">:</span>
                              <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">_</span><span class="si">%s%u</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="n">r</span><span class="p">[</span><span class="mi">3</span><span class="p">]),</span>
                              <span class="n">result</span><span class="p">))</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">Ptm</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_ptm</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">Domain</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">intera</span><span class="o">.</span><span class="n">DomainMotif</span><span class="p">):</span>

            <span class="n">result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_domain_motif</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="PtmHomology.ptm_orthology"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.PtmHomology.html#pypath.utils.homology.PtmHomology.ptm_orthology">[docs]</a>    <span class="k">def</span> <span class="nf">ptm_orthology</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Creates an orthology translation dict of phosphosites</span>
<span class="sd">        based on phosphorylation sites table from PhosphoSitePlus.</span>
<span class="sd">        In the result all PTMs represented by a tuple of the following</span>
<span class="sd">        6 elements: UniProt ID, isoform (int), residue one letter code,</span>
<span class="sd">        residue number (int), NCBI Taxonomy ID (int), modification type.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="n">nondigit</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;[^\d]+&#39;</span><span class="p">)</span>

        <span class="n">unknown_taxa</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

        <span class="k">for</span> <span class="n">typ</span> <span class="ow">in</span> <span class="n">common</span><span class="o">.</span><span class="n">psite_mod_types</span><span class="p">:</span>

            <span class="n">groups</span> <span class="o">=</span> <span class="p">{}</span>

            <span class="n">url</span> <span class="o">=</span> <span class="n">urls</span><span class="o">.</span><span class="n">urls</span><span class="p">[</span><span class="s1">&#39;psite_</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">typ</span><span class="p">[</span><span class="mi">0</span><span class="p">]][</span><span class="s1">&#39;url&#39;</span><span class="p">]</span>
            <span class="n">c</span> <span class="o">=</span> <span class="n">curl</span><span class="o">.</span><span class="n">Curl</span><span class="p">(</span><span class="n">url</span><span class="p">,</span> <span class="n">silent</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">large</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="n">data</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">result</span>

            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">xrange</span><span class="p">(</span><span class="mi">4</span><span class="p">):</span>
                <span class="n">null</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">data</span><span class="p">:</span>

                <span class="n">r</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>

                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">r</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">10</span><span class="p">:</span>

                    <span class="k">continue</span>

                <span class="n">uniprot</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
                <span class="n">isoform</span> <span class="o">=</span> <span class="p">(</span>
                    <span class="mi">1</span>
                        <span class="k">if</span> <span class="s1">&#39;-&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">uniprot</span> <span class="k">else</span>
                    <span class="nb">int</span><span class="p">(</span><span class="n">uniprot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
                <span class="p">)</span>
                <span class="n">uniprot</span> <span class="o">=</span> <span class="n">uniprot</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">aa</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">4</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">nondigit</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">r</span><span class="p">[</span><span class="mi">4</span><span class="p">]))</span>

                <span class="k">if</span> <span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">taxa</span><span class="p">:</span>

                    <span class="n">unknown_taxa</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
                    <span class="k">continue</span>

                <span class="n">tax</span> <span class="o">=</span> <span class="n">taxonomy</span><span class="o">.</span><span class="n">taxa</span><span class="p">[</span><span class="n">r</span><span class="p">[</span><span class="mi">6</span><span class="p">]]</span>
                <span class="n">group</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">r</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span>

                <span class="n">this_site</span> <span class="o">=</span> <span class="p">(</span><span class="n">uniprot</span><span class="p">,</span> <span class="n">isoform</span><span class="p">,</span> <span class="n">aa</span><span class="p">,</span> <span class="n">num</span><span class="p">,</span> <span class="n">tax</span><span class="p">,</span> <span class="n">typ</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>

                <span class="k">if</span> <span class="n">group</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">groups</span><span class="p">:</span>
                    <span class="n">groups</span><span class="p">[</span><span class="n">group</span><span class="p">]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

                <span class="n">groups</span><span class="p">[</span><span class="n">group</span><span class="p">]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">this_site</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">group</span><span class="p">,</span> <span class="n">sites</span> <span class="ow">in</span> <span class="n">iteritems</span><span class="p">(</span><span class="n">groups</span><span class="p">):</span>

                <span class="k">for</span> <span class="n">site1</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>

                    <span class="k">for</span> <span class="n">site2</span> <span class="ow">in</span> <span class="n">sites</span><span class="p">:</span>

                        <span class="k">if</span> <span class="n">site1</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]:</span>

                            <span class="k">continue</span>

                        <span class="k">if</span> <span class="n">site1</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">:</span>

                            <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">site1</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>

                        <span class="k">if</span> <span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">site1</span><span class="p">]:</span>

                            <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">site1</span><span class="p">][</span><span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([])</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">ptmhomo</span><span class="p">[</span><span class="n">site1</span><span class="p">][</span><span class="n">site2</span><span class="p">[</span><span class="mi">4</span><span class="p">]]</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">site2</span><span class="p">)</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">unknown_taxa</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">_log</span><span class="p">(</span>
                <span class="s1">&#39;Unknown taxa encountered: </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="p">(</span>
                    <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">unknown_taxa</span><span class="p">))</span>
                <span class="p">)</span>
            <span class="p">)</span></div></div>


<div class="viewcode-block" id="init"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.init.html#pypath.utils.homology.init">[docs]</a><span class="k">def</span> <span class="nf">init</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates an instance of the homology manager. Stores it in the module</span>
<span class="sd">    namespace.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">globals</span><span class="p">()[</span><span class="s1">&#39;manager&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">HomologyManager</span><span class="p">()</span></div>


<div class="viewcode-block" id="get_manager"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.get_manager.html#pypath.utils.homology.get_manager">[docs]</a><span class="k">def</span> <span class="nf">get_manager</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns the homology manager, an object which loads and unloads the</span>
<span class="sd">    homology lookup tables as necessary, and provides the interface for</span>
<span class="sd">    querying the homology data. Normally an instance of the manager</span>
<span class="sd">    belongs to the module, and if it does not exist yet, will be created</span>
<span class="sd">    automatically.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="s1">&#39;manager&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="nb">globals</span><span class="p">():</span>

        <span class="n">init</span><span class="p">()</span>

    <span class="k">return</span> <span class="nb">globals</span><span class="p">()[</span><span class="s1">&#39;manager&#39;</span><span class="p">]</span></div>


<div class="viewcode-block" id="translate"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.translate.html#pypath.utils.homology.translate">[docs]</a><span class="k">def</span> <span class="nf">translate</span><span class="p">(</span>
        <span class="n">source_id</span><span class="p">,</span>
        <span class="n">target</span><span class="p">,</span>
        <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Homology translation. For a UniProt ID, finds the corresponding</span>
<span class="sd">    homologous (orthologous) genes in another organism.</span>

<span class="sd">    Args</span>
<span class="sd">        source_id (str,list): UniProt ID of one or more protein in the</span>
<span class="sd">                source organism.</span>
<span class="sd">        target (int,str): The target organism.</span>
<span class="sd">        source (int,str): The source organism.</span>
<span class="sd">        only_swissprot (bool): Use only SwissProt IDs. For human and some</span>
<span class="sd">            popular model organisms this is advisible, as almost all proteins</span>
<span class="sd">            have reviewed record in UniProt.</span>
<span class="sd">        homologene (bool): Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">        ensembl (bool): Use Ensembl data for ortholog lookup.</span>
<span class="sd">        ensembl_hc (bool): Use only high confidence orthology relations</span>
<span class="sd">            from Ensembl. By default it is True. You can also set it</span>
<span class="sd">            by the `ensembl_hc` attribute.</span>
<span class="sd">        ensembl_types (list): The Ensembl orthology relationship types</span>
<span class="sd">            to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">            `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">            also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">        id_type (str): Type of the identifiers to translate from and to.</span>

<span class="sd">    Returns</span>
<span class="sd">        Set of UniProt IDs of orthologous gene products in the target</span>
<span class="sd">        organism.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">translate</span><span class="p">(</span>
        <span class="n">source_id</span>  <span class="o">=</span> <span class="n">source_id</span><span class="p">,</span>
        <span class="n">target</span> <span class="o">=</span> <span class="n">target</span><span class="p">,</span>
        <span class="n">source</span> <span class="o">=</span> <span class="n">source</span><span class="p">,</span>
        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="n">only_swissprot</span><span class="p">,</span>
        <span class="n">homologene</span> <span class="o">=</span> <span class="n">homologene</span><span class="p">,</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="n">ensembl</span><span class="p">,</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="n">ensembl_hc</span><span class="p">,</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="n">ensembl_types</span><span class="p">,</span>
        <span class="n">id_type</span> <span class="o">=</span> <span class="n">id_type</span><span class="p">,</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="get_dict"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.get_dict.html#pypath.utils.homology.get_dict">[docs]</a><span class="k">def</span> <span class="nf">get_dict</span><span class="p">(</span>
        <span class="n">target</span><span class="p">,</span>
        <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a dictionary for one source organism and ID type.</span>

<span class="sd">    Args</span>
<span class="sd">        target:</span>
<span class="sd">            Tha organism to translate to.</span>
<span class="sd">        source:</span>
<span class="sd">            The organism to translate from.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use Ensembl data for ortholog lookup.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only high confidence orthology relations</span>
<span class="sd">            from Ensembl. By default it is True. You can also set it</span>
<span class="sd">            by the `ensembl_hc` attribute.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            The Ensembl orthology relationship types</span>
<span class="sd">            to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">            `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">            also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">        id_type:</span>
<span class="sd">            Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">            first element will be used for keys (source organism) and</span>
<span class="sd">            the second for the values (target organism). By default</span>
<span class="sd">            UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">            ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">            ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">    Return</span>
<span class="sd">        A dict with identifiers of the source organism as keys, and</span>
<span class="sd">        sets of their orthologs as values.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">get_dict</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.get_df.html#pypath.utils.homology.get_df">[docs]</a><span class="k">def</span> <span class="nf">get_df</span><span class="p">(</span>
        <span class="n">target</span><span class="p">,</span>
        <span class="n">source</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">only_swissprot</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">homologene</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_types</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">id_type</span> <span class="o">=</span> <span class="s1">&#39;uniprot&#39;</span><span class="p">,</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create a data frame for one source organism and ID type.</span>

<span class="sd">    Args</span>
<span class="sd">        target:</span>
<span class="sd">            The organism to translate to.</span>
<span class="sd">        source:</span>
<span class="sd">            The organism to translate from.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use Ensembl data for ortholog lookup.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only high confidence orthology relations</span>
<span class="sd">            from Ensembl. By default it is True. You can also set it</span>
<span class="sd">            by the `ensembl_hc` attribute.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            The Ensembl orthology relationship types</span>
<span class="sd">            to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">            `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">            also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">        id_type:</span>
<span class="sd">            Identifier type to translate from and to. If a tuple, the</span>
<span class="sd">            first element will be used for keys (source organism) and</span>
<span class="sd">            the second for the values (target organism). By default</span>
<span class="sd">            UniProt IDs are assumed. For a complete list of ID types see</span>
<span class="sd">            ``pypath.utils.mapping.get_mapper().id_types()`` and</span>
<span class="sd">            ``pypath.utils.mapping.get_mapper().mapping_tables()``.</span>

<span class="sd">    Return</span>
<span class="sd">        A data frame with pairs of orthologous identifiers,</span>
<span class="sd">        in two columns: &quot;source&quot; and &quot;target&quot;.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">get_df</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">)</span></div>


<div class="viewcode-block" id="translate_df"><a class="viewcode-back" href="../../../_autosummary/pypath.utils.homology.translate_df.html#pypath.utils.homology.translate_df">[docs]</a><span class="k">def</span> <span class="nf">translate_df</span><span class="p">(</span>
        <span class="n">df</span><span class="p">,</span>
        <span class="n">target</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">cols</span><span class="p">:</span> <span class="p">(</span>
            <span class="nb">str</span> <span class="o">|</span>
            <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">|</span>
            <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">|</span>
            <span class="kc">None</span>
        <span class="p">)</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">source</span><span class="p">:</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">str</span> <span class="o">=</span> <span class="mi">9606</span><span class="p">,</span>
        <span class="n">homologene</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_hc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ensembl_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Literal</span><span class="p">[</span>
            <span class="s1">&#39;one2one&#39;</span><span class="p">,</span> <span class="s1">&#39;one2many&#39;</span><span class="p">,</span> <span class="s1">&#39;many2many&#39;</span>
        <span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="o">**</span><span class="n">kwargs</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">]</span>
    <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Translate columns in a data frame.</span>

<span class="sd">    Args</span>
<span class="sd">        df:</span>
<span class="sd">            A data frame.</span>
<span class="sd">        target:</span>
<span class="sd">            The organism to translate to.</span>
<span class="sd">        cols:</span>
<span class="sd">            One or more columns to be translated. It can be a single</span>
<span class="sd">            column name, an iterable of column names or a dict where</span>
<span class="sd">            keys are column names and values are ID types. Except this</span>
<span class="sd">            last case, identifiers are assumed to be UniProt.</span>
<span class="sd">        source:</span>
<span class="sd">            The organism to translate from.</span>
<span class="sd">        homologene:</span>
<span class="sd">            Use NCBI HomoloGene data for ortholog lookup.</span>
<span class="sd">        ensembl:</span>
<span class="sd">            Use Ensembl data for ortholog lookup.</span>
<span class="sd">        ensembl_hc:</span>
<span class="sd">            Use only high confidence orthology relations</span>
<span class="sd">            from Ensembl. By default it is True. You can also set it</span>
<span class="sd">            by the `ensembl_hc` attribute.</span>
<span class="sd">        ensembl_types:</span>
<span class="sd">            The Ensembl orthology relationship types</span>
<span class="sd">            to use. Possible values are `one2one`, `one2many` and</span>
<span class="sd">            `many2many`. By default only `one2one` is used. You can</span>
<span class="sd">            also set this parameter by the `ensembl_types` attribute.</span>
<span class="sd">        kwargs:</span>
<span class="sd">            Same as providing a dict to ``cols``, but beware, keys</span>
<span class="sd">            (column names) can not match existing argument names of</span>
<span class="sd">            this function.</span>

<span class="sd">    Return</span>
<span class="sd">        A data frame with the same column layout as the input, and the</span>
<span class="sd">        identifiers translated as demanded. Rows that could not be</span>
<span class="sd">        translated are omitted.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">manager</span> <span class="o">=</span> <span class="n">get_manager</span><span class="p">()</span>

    <span class="n">args</span> <span class="o">=</span> <span class="nb">locals</span><span class="p">()</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;manager&#39;</span><span class="p">)</span>
    <span class="n">kwargs</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;kwargs&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">manager</span><span class="o">.</span><span class="n">translate_df</span><span class="p">(</span><span class="o">**</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></div>
</pre></div>
            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

          <div class="sidebar-resize-handle"></div>

  <div>
    <h3><a href="../../../index.html">Table of Contents</a></h3>
    <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html"><em>pypath:</em> A Python module for molecular signaling prior knowledge processing</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../webservice.html">Webservice</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../releasehistory.html">Release history</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../_autosummary/pypath.html">pypath</a></li>
</ul>

  </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>
<div id="ethical-ad-placement"></div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <nav class="relbar">
    </nav>

    <footer role="contentinfo">
      &#169; Copyright 2023, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Erva Ulusoy, Ahmet Rifaioglu, Charles Tapley Hoyt.
      Created using <a class="reference external" href="https://www.sphinx-doc.org/">Sphinx</a> 5.3.0.
      <a class="reference external" href="https://insipid-sphinx-theme.readthedocs.io/">Insipid Theme</a>.
    </footer>
  </body>
</html>