
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>The pypath book &#8212; pypath 0.14.32 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../_static/styles/bootstrap.css?digest=796348d33e8b1d947c94" rel="stylesheet">
<link href="../_static/styles/pydata-sphinx-theme.css?digest=796348d33e8b1d947c94" rel="stylesheet">

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=796348d33e8b1d947c94" rel="stylesheet">
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2">
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2">

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/pypathdocs.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94">
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/manual';</script>
    <link rel="shortcut icon" href="../_static/omnipath-logo.svg"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="pypath.visual.plot.StackedBarplot" href="../api/pypath.visual.plot.StackedBarplot.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="docsearch:language" content="en">
  </head>
  
  
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="180" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>

  
  <input type="checkbox" class="sidebar-toggle" name="__primary" id="__primary">
  <label class="overlay overlay-primary" for="__primary"></label>

  
  <input type="checkbox" class="sidebar-toggle" name="__secondary" id="__secondary">
  <label class="overlay overlay-secondary" for="__secondary"></label>

  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
      
<form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false">
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
    </div>
  </div>

  
  <nav class="bd-header navbar navbar-expand-lg bd-navbar" id="navbar-main"><div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
      <span class="fa-solid fa-bars"></span>
  </label>
  <div id="navbar-start">
    
    
  


<a class="navbar-brand logo" href="../contents.html">

  
  
  
  
  
  
  

  
    <img src="../_static/img/omnipath-logo.svg" class="logo__image only-light" alt="Pypath">
    <img src="../_static/img/omnipath-logo.svg" class="logo__image only-dark" alt="Pypath">
  
  
    <p class="title logo__title">Pypath</p>
  
</a>
    
  </div>

  
  <div class=" navbar-header-items">
    <div id="navbar-center" class="mr-auto">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        The pypath book
                      </a>
                    </li>
                
    </ul>
</nav>
      </div>
      
    </div>

    <div id="navbar-end">
      
        <div class="navbar-end-item navbar-persistent--container">
          
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
        </div>
      
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
      </ul>
      </div>
      
    </div>
  </div>


  
  
    <div class="navbar-persistent--mobile">
<button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-toggle="tooltip">
  <i class="fa-solid fa-magnifying-glass"></i>
</button>
    </div>
  

  
  <label class="sidebar-toggle secondary-toggle" for="__secondary">
      <span class="fa-solid fa-outdent"></span>
  </label>
  

</div>
  </nav>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        
  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
      
      <div class="navbar-center-item">
        <nav class="navbar-nav">
    <p class="sidebar-header-items__title" role="heading" aria-level="1" aria-label="Site Navigation">
        Site Navigation
    </p>
    <ul id="navbar-main-elements" class="navbar-nav">
        
                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../index.html">
                        Welcome to pypath!
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../webservice.html">
                        Webservice
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../releasehistory.html">
                        Release history
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../reference.html">
                        Reference
                      </a>
                    </li>
                

                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="#">
                        The pypath book
                      </a>
                    </li>
                
    </ul>
</nav>
      </div>
      
      </div>
    

    
    
    <div class="sidebar-header-items__end">
      
      <div class="navbar-end-item">
        <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
</button>
      </div>
      
      <div class="navbar-end-item">
        <ul id="navbar-icon-links" class="navbar-nav" aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://github.com/saezlab/pypath" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-brands fa-square-github"></i></span>
            <label class="sr-only">GitHub</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://twitter.com/omnipathdb" title="Twitter" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-brands fa-square-twitter"></i></span>
            <label class="sr-only">Twitter</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://pypi.org/project/pypath-omnipath" title="PyPI" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-box"></i></span>
            <label class="sr-only">PyPI</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://omnipathdb.org/" title="OmniPath" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-university"></i></span>
            <label class="sr-only">OmniPath</label></a>
        </li>
        <li class="nav-item">
          
          
          
          
          
          
          
          <a href="https://saezlab.org/" title="Saez Lab" class="nav-link" rel="noopener" target="_blank" data-toggle="tooltip"><span><i class="fa-solid fa-users"></i></span>
            <label class="sr-only">Saez Lab</label></a>
        </li>
      </ul>
      </div>
      
    </div>
    
  </div>

  
  <div class="sidebar-start-items sidebar-primary__section">
    <div class="sidebar-start-items__item">
  <div>
    <h3><a href="../contents.html">Table of Contents</a></h3>
    <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Welcome to pypath!</a></li>
<li class="toctree-l1"><a class="reference internal" href="../webservice.html">Webservice</a></li>
<li class="toctree-l1"><a class="reference internal" href="../releasehistory.html">Release history</a></li>
<li class="toctree-l1"><a class="reference internal" href="../reference.html">Reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">The pypath book</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Introduction">Introduction</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Build,-load-and-save-databases">Build, load and save databases</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#The-OmniPath-app">The OmniPath app</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Built-in-database-definitions">Built-in database definitions</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Networks">Networks</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Strictly-literature-curated-network">Strictly literature curated network</a></li>
<li class="toctree-l4"><a class="reference internal" href="#The-OmniPath-network-with-extra-activity-flow,-enzyme-substrate-and-ligand-receptor-interactions">The OmniPath network with extra activity flow, enzyme-substrate and ligand-receptor interactions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Transcriptional-regulation-network-from-DoRothEA-and-other-resources">Transcriptional regulation network from DoRothEA and other resources</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Literature-curated-miRNA-post-transcriptional-regulation-network">Literature curated miRNA post-transcriptional regulation network</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Transcriptional-regulation-of-miRNA">Transcriptional regulation of miRNA</a></li>
<li class="toctree-l4"><a class="reference internal" href="#lncRNA-mRNA-interactions">lncRNA-mRNA interactions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Small-molecule-protein-interactions">Small molecule-protein interactions</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Enzyme-substrate-relationships">Enzyme-substrate relationships</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Protein-complexes">Protein complexes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Annotations">Annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Inter-cellular-communication-roles">Inter-cellular communication roles</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Data-directly-from-the-original-resources">Data directly from the original resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Download-management">Download management</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Cache-management-and-customization">Cache management and customization</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Download-failures">Download failures</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Corrupted-cache-content">Corrupted cache content</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Network-communication-issues:-look-into-the-curl-debug-log">Network communication issues: look into the curl debug log</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Timeouts">Timeouts</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Access-and-inspect-the-Curl-object">Access and inspect the <code class="docutils literal notranslate"><span class="pre">Curl</span></code> object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Is-it-failing-only-for-you?">Is it failing only for you?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Read-the-log">Read the log</a></li>
<li class="toctree-l4"><a class="reference internal" href="#TLS-(SSL,-HTTPS)-errors">TLS (SSL, HTTPS) errors</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Resources">Resources</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Licenses">Licenses</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Resource-information">Resource information</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Resource-definitions-for-a-certain-database-or-dataset">Resource definitions for a certain database or dataset</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Building-networks">Building networks</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Which-network-datasets-are-pre-defined-in-pypath?">Which network datasets are pre-defined in pypath?</a></li>
<li class="toctree-l3"><a class="reference internal" href="#The-Network-object">The <code class="docutils literal notranslate"><span class="pre">Network</span></code> object</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Network-in-pandas.DataFrame">Network in <em>pandas.DataFrame</em></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Translating-identifiers">Translating identifiers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Pre-defined-ID-translation-tables">Pre-defined ID translation tables</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Homology-translation">Homology translation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Homology-translation-tables-as-dictionaries">Homology translation tables as dictionaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Homology-translation-data-frames">Homology translation data frames</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Taxonomy">Taxonomy</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Translating-to-NCBI-Taxonomy,-scientific-names-and-common-names">Translating to NCBI Taxonomy, scientific names and common names</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Organism-from-UniProt-ID">Organism from UniProt ID</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#UniProt">UniProt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#The-UniProt-input-module">The UniProt input module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#All-UniProt-IDs-for-one-organism">All UniProt IDs for one organism</a></li>
<li class="toctree-l4"><a class="reference internal" href="#UniProt-ID-format-validation">UniProt ID format validation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#UniProt-ID-validation">UniProt ID validation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Single-UniProt-protein-datasheet">Single UniProt protein datasheet</a></li>
<li class="toctree-l4"><a class="reference internal" href="#History-of-UniProt-records">History of UniProt records</a></li>
<li class="toctree-l4"><a class="reference internal" href="#UniProt-legacy-API">UniProt legacy API</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Processed-UniProt-annotations">Processed UniProt annotations</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#The-UniProt-utils-module">The UniProt utils module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Datasheets">Datasheets</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Tables">Tables</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Sanitizing-UniProt-IDs">Sanitizing UniProt IDs</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Enzyme-substrate-interactions">Enzyme-substrate interactions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Enzyme-substrate-objects">Enzyme-substrate objects</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Enzyme-substrate-data-frame">Enzyme-substrate data frame</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Protein-sequences">Protein sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id1">Annotations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-a-single-annotation-resource">Load a single annotation resource</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-the-full-annotations-database-by-the-database-manager">Load the full annotations database by the database manager</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-only-selected-annotations">Load only selected annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Data-frames-of-annotations">Data frames of annotations</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Inter-cellular-signaling-roles">Inter-cellular signaling roles</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Build-an-intercellular-communication-network">Build an intercellular communication network</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Quantitative-overview-of-intercell-annotations">Quantitative overview of intercell annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Intercell-database-as-data-frame">Intercell database as data frame</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Browse-intercell-categories">Browse intercell categories</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Gene-Ontology">Gene Ontology</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id2">Protein complexes</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Protein-complex-objects">Protein complex objects</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Protein-complex-data-frame">Protein complex data frame</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Saving-datasets-as-pickles">Saving datasets as pickles</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Log-messages-and-sessions">Log messages and sessions</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Basic-info-about-the-session">Basic info about the session</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Read-the-log-file">Read the log file</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Logging-to-the-console">Logging to the console</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Disable-logging">Disable logging</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Write-to-the-log">Write to the log</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Sending-a-single-message">Sending a single message</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Connect-a-module-or-class-to-the-pypath-logger">Connect a module or class to the pypath logger</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#BEL-export">BEL export</a></li>
<li class="toctree-l2"><a class="reference internal" href="#CellPhoneDB-export">CellPhoneDB export</a></li>
<li class="toctree-l2"><a class="reference internal" href="#The-legacy-igraph-based-network-object">The legacy <em>igraph</em>-based network object</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#I-just-want-a-network-quickly-and-play-around-with-pypath">I just want a network quickly and play around with <em>pypath</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#How-do-I-build-networks-from-any-data-with-pypath?">How do I build networks from any data with <em>pypath</em>?</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Defining-input-formats">Defining input formats</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Creating-PyPath-object-and-loading-the-2-test-files">Creating PyPath object and loading the 2 test files</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Structure-of-the-legacy-network-object">Structure of the legacy network object</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Directions-and-signs">Directions and signs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Accessing-nodes-in-the-network">Accessing nodes in the network</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#Querying-relationships-with-our-without-causality">Querying relationships with our without causality</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Accessing-edges-by-identifiers">Accessing edges by identifiers</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Literature-references">Literature references</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plotting-the-network-with-igraph">Plotting the network with <em>igraph</em></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  </div>
    </div>
  </div>
  

  
  <div class="sidebar-end-items sidebar-primary__section">
    <div class="sidebar-end-items__item">
    </div>
  </div>

  
  <div id="rtd-footer-container"></div>

      </div>
      <main id="main-content" class="bd-main">
        
        
        <div class="bd-content">
          <div class="bd-article-container">
            
            <div class="bd-header-article">
                
            </div>
            
            
            <article class="bd-article" role="main">
              
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars and line breaks on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
    white-space: pre;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<section id="The-pypath-book">
<h1>The pypath book<a class="headerlink" href="#The-pypath-book" title="Permalink to this heading">§</a></h1>
<h1><p>Contents</p>
</h1><div class="toc"><ul class="toc-item"><li><p>1  Introduction</p>
</li><li><p>2  Build, load and save databases</p>
<ul class="toc-item"><li><p>2.1  The OmniPath app</p>
</li><li><p>2.2  Built-in database definitions</p>
</li><li><p>2.3  Networks</p>
<ul class="toc-item"><li><p>2.3.1  Strictly literature curated network</p>
</li><li><p>2.3.2  The OmniPath network with extra activity flow, enzyme-substrate and ligand-receptor interactions</p>
</li><li><p>2.3.3  Transcriptional regulation network from DoRothEA and other resources</p>
</li><li><p>2.3.4  Literature curated miRNA post-transcriptional regulation network</p>
</li><li><p>2.3.5  Transcriptional regulation of miRNA</p>
</li><li><p>2.3.6  lncRNA-mRNA interactions</p>
</li><li><p>2.3.7  Small molecule-protein interactions</p>
</li></ul></li><li><p>2.4  Enzyme-substrate relationships</p>
</li><li><p>2.5  Protein complexes</p>
</li><li><p>2.6  Annotations</p>
</li><li><p>2.7  Inter-cellular communication roles</p>
</li></ul></li><li><p>3  Data directly from the original resources</p>
</li><li><p>4  Download management</p>
<ul class="toc-item"><li><p>4.1  Cache management and customization</p>
</li><li><p>4.2  Download failures</p>
<ul class="toc-item"><li><p>4.2.1  Corrupted cache content</p>
</li><li><p>4.2.2  Network communication issues: look into the curl debug log</p>
</li><li><p>4.2.3  Timeouts</p>
</li><li><p>4.2.4  Access and inspect the Curl object</p>
</li><li><p>4.2.5  Is it failing only for you?</p>
</li><li><p>4.2.6  Read the log</p>
</li><li><p>4.2.7  TLS (SSL, HTTPS) errors</p>
</li></ul></li></ul></li><li><p>5  Resources</p>
<ul class="toc-item"><li><p>5.1  Licenses</p>
</li><li><p>5.2  Resource information</p>
</li><li><p>5.3  Resource definitions for a certain database or dataset</p>
</li></ul></li><li><p>6  Building networks</p>
<ul class="toc-item"><li><p>6.1  Which network datasets are pre-defined in pypath?</p>
</li><li><p>6.2  The Network object</p>
</li><li><p>6.3  Network in pandas.DataFrame</p>
</li></ul></li><li><p>7  Translating identifiers</p>
<ul class="toc-item"><li><p>7.1  Pre-defined ID translation tables</p>
</li></ul></li><li><p>8  Homology translation</p>
<ul class="toc-item"><li><p>8.1  Homology translation tables as dictionaries</p>
</li><li><p>8.2  Homology translation data frames</p>
</li></ul></li><li><p>9  Taxonomy</p>
<ul class="toc-item"><li><p>9.1  Translating to NCBI Taxonomy, scientific names and common names</p>
</li><li><p>9.2  Organism from UniProt ID</p>
</li></ul></li><li><p>10  UniProt</p>
<ul class="toc-item"><li><p>10.1  The UniProt input module</p>
<ul class="toc-item"><li><p>10.1.1  All UniProt IDs for one organism</p>
</li><li><p>10.1.2  UniProt ID format validation</p>
</li><li><p>10.1.3  UniProt ID validation</p>
</li><li><p>10.1.4  Single UniProt protein datasheet</p>
</li><li><p>10.1.5  History of UniProt records</p>
</li><li><p>10.1.6  UniProt legacy API</p>
</li><li><p>10.1.7  Processed UniProt annotations</p>
</li></ul></li><li><p>10.2  The UniProt utils module</p>
<ul class="toc-item"><li><p>10.2.1  Datasheets</p>
</li><li><p>10.2.2  Tables</p>
</li></ul></li><li><p>10.3  Sanitizing UniProt IDs</p>
</li></ul></li><li><p>11  Enzyme-substrate interactions</p>
<ul class="toc-item"><li><p>11.1  Enzyme-substrate objects</p>
</li><li><p>11.2  Enzyme-substrate data frame</p>
</li></ul></li><li><p>12  Protein sequences</p>
</li><li><p>13  Annotations</p>
<ul class="toc-item"><li><p>13.1  Load a single annotation resource</p>
</li><li><p>13.2  Load the full annotations database by the database manager</p>
</li><li><p>13.3  Load only selected annotations</p>
</li><li><p>13.4  Data frames of annotations</p>
</li></ul></li><li><p>14  Inter-cellular signaling roles</p>
<ul class="toc-item"><li><p>14.1  Build an intercellular communication network</p>
</li><li><p>14.2  Quantitative overview of intercell annotations</p>
</li><li><p>14.3  Intercell database as data frame</p>
</li><li><p>14.4  Browse intercell categories</p>
</li></ul></li><li><p>15  Gene Ontology</p>
</li><li><p>16  Protein complexes</p>
<ul class="toc-item"><li><p>16.1  Protein complex objects</p>
</li><li><p>16.2  Protein complex data frame</p>
</li></ul></li><li><p>17  Saving datasets as pickles</p>
</li><li><p>18  Log messages and sessions</p>
<ul class="toc-item"><li><p>18.1  Basic info about the session</p>
</li><li><p>18.2  Read the log file</p>
</li><li><p>18.3  Logging to the console</p>
</li><li><p>18.4  Disable logging</p>
</li><li><p>18.5  Write to the log</p>
<ul class="toc-item"><li><p>18.5.1  Sending a single message</p>
</li><li><p>18.5.2  Connect a module or class to the pypath logger</p>
</li></ul></li></ul></li><li><p>19  BEL export</p>
</li><li><p>20  CellPhoneDB export</p>
</li><li><p>21  The legacy igraph-based network object</p>
<ul class="toc-item"><li><p>21.1  I just want a network quickly and play around with pypath</p>
</li><li><p>21.2  How do I build networks from any data with pypath?</p>
<ul class="toc-item"><li><p>21.2.1  Defining input formats</p>
</li><li><p>21.2.2  Creating PyPath object and loading the 2 test files</p>
</li></ul></li><li><p>21.3  Structure of the legacy network object</p>
<ul class="toc-item"><li><p>21.3.1  Directions and signs</p>
</li><li><p>21.3.2  Accessing nodes in the network</p>
</li></ul></li><li><p>21.4  Querying relationships with our without causality</p>
</li><li><p>21.5  Accessing edges by identifiers</p>
</li><li><p>21.6  Literature references</p>
</li><li><p>21.7  Plotting the network with igraph</p>
</li></ul></li></ul></div><section id="Introduction">
<h2>Introduction<a class="headerlink" href="#Introduction" title="Permalink to this heading">§</a></h2>
<p>OmniPath consists of 5 main database segments: network <em>(interactions)</em>, enzyme-substrate interactions <em>(enz_sub or ptms)</em>, protein complexes <em>(complexes)</em>, molecular entity annotations <em>(annotations)</em> and intercellular communication roles <em>(intercell)</em>. You can access all these by the web service at <a class="reference external" href="https://omnipathdb.org/">https://omnipathdb.org/</a> and the R/Bioconductor package <em>OmnipathR</em>, furthermore the network and some of the annotations by the Cytoscape app. However only <em>pypath</em> is able to build these databases
directly from the original sources with various options for customization and to provide a rich and versatile API for each database enjoying the almost unlimited flexibility of Python. This book attempts to be a guided tour around <em>pypath</em>, however almost all objects, modules, APIs presented here have many more methods, options and features than we have a chance to cover. If you feel like there might be something useful for you, don’t hesitate to ask us by
<a class="reference external" href="https://github.com/saezlab/pypath/issues">github</a>.</p>
<p>This document has been run with the following <em>pypath</em> version:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pypath
pypath.__version__
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;0.14.30&#39;
</pre></div></div>
</div>
</section>
<section id="Build,-load-and-save-databases">
<h2>Build, load and save databases<a class="headerlink" href="#Build,-load-and-save-databases" title="Permalink to this heading">§</a></h2>
<p>We provide a high level interface in the module <em>pypath.omnipath.app</em>. This is the easiest way to build, manage and access the OmniPath databases, hence this is what we present in the <em>Quick start</em> section. In further sections we show the lower level modules more in detail.</p>
<section id="The-OmniPath-app">
<h3>The OmniPath app<a class="headerlink" href="#The-OmniPath-app" title="Permalink to this heading">§</a></h3>
<p><em>pypath.omnipath</em> is an application which contains a database manager at <em>omnipath.db</em>. This manager is empty by default. It builds and loads the databases on demand.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath

omnipath.db
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 14:16:16] [curl] Module `pysftp` not available. Only downloading of a small number of resources relies on this module. Please install by PIP if it is necessary for you.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;pypath.omnipath.app.DatabaseManager at 0x637fc84d83d0&gt;
</pre></div></div>
</div>
</section>
<section id="Built-in-database-definitions">
<h3>Built-in database definitions<a class="headerlink" href="#Built-in-database-definitions" title="Permalink to this heading">§</a></h3>
<p>The databases presented below are <a class="reference external" href="https://github.com/saezlab/pypath/blob/master/pypath/omnipath/databases/builtins.json">pre-defined in pypath</a>. You can also list them by:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
omnipath.db.datasets
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;omnipath&#39;,
 &#39;curated&#39;,
 &#39;complex&#39;,
 &#39;annotations&#39;,
 &#39;intercell&#39;,
 &#39;tf_target&#39;,
 &#39;tf_mirna&#39;,
 &#39;mirna_mrna&#39;,
 &#39;lncrna_mrna&#39;,
 &#39;enz_sub&#39;]
</pre></div></div>
</div>
</section>
<section id="Networks">
<h3>Networks<a class="headerlink" href="#Networks" title="Permalink to this heading">§</a></h3>
<p>OmniPath offers multiple built in network datasets: the OmniPath PPI network the more strict literature curated PPI network, the special ligand-receptor PPI network and various other PPI datasets, the transcriptional regulation network from DoRothEA and other resources, miRNA post-transcriptional regulation network and also transcriptional regulation network for miRNAs.</p>
<section id="Strictly-literature-curated-network">
<h4>Strictly literature curated network<a class="headerlink" href="#Strictly-literature-curated-network" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
cu = omnipath.db.get_db(&#39;curated&#39;)
cu
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 7980 nodes, 35551 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="The-OmniPath-network-with-extra-activity-flow,-enzyme-substrate-and-ligand-receptor-interactions">
<h4>The OmniPath network with extra activity flow, enzyme-substrate and ligand-receptor interactions<a class="headerlink" href="#The-OmniPath-network-with-extra-activity-flow,-enzyme-substrate-and-ligand-receptor-interactions" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
op = omnipath.db.get_db(&#39;omnipath&#39;)
op
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 18558 nodes, 94358 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="Transcriptional-regulation-network-from-DoRothEA-and-other-resources">
<h4>Transcriptional regulation network from DoRothEA and other resources<a class="headerlink" href="#Transcriptional-regulation-network-from-DoRothEA-and-other-resources" title="Permalink to this heading">§</a></h4>
<p>Note: according to the default settings, DoRothEA confidence levels A-D and all original resources will be loaded. To load only DoRothEA, use the key <code class="docutils literal notranslate"><span class="pre">&quot;dorothea&quot;</span></code> instead of <code class="docutils literal notranslate"><span class="pre">&quot;tf_target&quot;</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
tft = omnipath.db.get_db(&#39;tf_target&#39;)
tft
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 18986 nodes, 326708 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="Literature-curated-miRNA-post-transcriptional-regulation-network">
<h4>Literature curated miRNA post-transcriptional regulation network<a class="headerlink" href="#Literature-curated-miRNA-post-transcriptional-regulation-network" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
mi = omnipath.db.get_db(&#39;mirna_mrna&#39;)
mi
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 1264 nodes, 3288 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="Transcriptional-regulation-of-miRNA">
<h4>Transcriptional regulation of miRNA<a class="headerlink" href="#Transcriptional-regulation-of-miRNA" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
tmi = omnipath.db.get_db(&#39;tf_mirna&#39;)
tmi
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 1032 nodes, 4960 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="lncRNA-mRNA-interactions">
<h4>lncRNA-mRNA interactions<a class="headerlink" href="#lncRNA-mRNA-interactions" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
lnc = omnipath.db.get_db(&#39;lncrna_mrna&#39;)
lnc
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 243 nodes, 217 interactions&gt;
</pre></div></div>
</div>
</section>
<section id="Small-molecule-protein-interactions">
<h4>Small molecule-protein interactions<a class="headerlink" href="#Small-molecule-protein-interactions" title="Permalink to this heading">§</a></h4>
<p>These interactions are either ligand-receptor connections, enzyme inhibitions, allosteric regulations or enzyme-metabolite interactions. Currently it is a small, experimental dataset, but will be largely extended in the future.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
smol = omnipath.db.get_db(&#39;small_molecule&#39;)
smol
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 1980 nodes, 3147 interactions&gt;
</pre></div></div>
</div>
</section>
</section>
<section id="Enzyme-substrate-relationships">
<h3>Enzyme-substrate relationships<a class="headerlink" href="#Enzyme-substrate-relationships" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
es = omnipath.db.get_db(&#39;enz_sub&#39;)
es
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Enzyme-substrate database: 41426 relationships&gt;
</pre></div></div>
</div>
</section>
<section id="Protein-complexes">
<h3>Protein complexes<a class="headerlink" href="#Protein-complexes" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
co = omnipath.db.get_db(&#39;complex&#39;)
co
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Complex database: 28173 complexes&gt;
</pre></div></div>
</div>
</section>
<section id="Annotations">
<h3>Annotations<a class="headerlink" href="#Annotations" title="Permalink to this heading">§</a></h3>
<p>The annotations database is huge, building or even loading it takes long time and requires quite some memory.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
an = omnipath.db.get_db(&#39;annotations&#39;)
an
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Annotation database: 5490653 records about 50872 entities from 68 resources&gt;
</pre></div></div>
</div>
</section>
<section id="Inter-cellular-communication-roles">
<h3>Inter-cellular communication roles<a class="headerlink" href="#Inter-cellular-communication-roles" title="Permalink to this heading">§</a></h3>
<p>This database is quick to build, but it requires the <em>annotations</em> database, which is really heavy.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
ic = omnipath.db.get_db(&#39;intercell&#39;)
ic
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Intercell annotations: 301527 records about 48570 entities&gt;
</pre></div></div>
</div>
</section>
</section>
<section id="Data-directly-from-the-original-resources">
<h2>Data directly from the original resources<a class="headerlink" href="#Data-directly-from-the-original-resources" title="Permalink to this heading">§</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.inputs</span></code> module contains clients for more than 150 molecular biology and biomedical resources, and overall almost 500 functions that download data directly from these resources. Maintaining such a large number of clients is troublesome, hence at any time some of them are broken, you can check them in our <a class="reference external" href="https://status.omnipathdb.org/">daily status report</a>. Each submodule of <code class="docutils literal notranslate"><span class="pre">pypath.inputs</span></code> is named after its corresponding resource, all lowercase, e.g. “depod” <em>(DEPOD)</em> or
“cytosig” <em>(CytoSig)</em>. Within these modules each function name starts with the name of the resource, and ends with the kind of data it retrieves. For example, <code class="docutils literal notranslate"><span class="pre">pypath.inputs.signor.signor_interactions</span></code> downloads interactions from <em>SIGNOR.</em> The labels *”_interactions“<em>,</em>”_enz_sub“<em>,</em>”_complexes”* and *”_annotations”* retrieve records intended to these respective databases. However, the records at this stage are not fully processed yet. Some functions have different postfixes,
e.g. <em>“_raw”* means the data is close to the format provided by the resource itself; *”_mapping”* means it is intended for a translation table. The purpose of the input functions is to 1) handle the download; 2) read the raw data; 3) extract the relevant parts; 4) do the specific part of processing, i.e. bring the data to a state when it is suitable for the generic database classes for further processing. The outputs of these functions is not standard in any ways, though you may observ
repeated patterns. The input functions typically return lists or dictionaries. These are arbitrarily designed towards the aims of selecting the relevant fields and give straightforward, accessible Python data structures for processing within or outside of *pypath</em>.</p>
<p>We use SIGNOR as an example because this resource provides data for almost all OmniPath databases. The <code class="docutils literal notranslate"><span class="pre">signor_complexes</span></code> function returns a set of <code class="docutils literal notranslate"><span class="pre">pypath.internals.intera.Complex</span></code> objects, ready to be added to the OmniPath complexes database (built by <code class="docutils literal notranslate"><span class="pre">pypath.core.complex.ComplexAggregator</span></code>).</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import signor
signor.signor_complexes()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;COMPLEX:P23511_P25208_Q13952&#39;: Complex NFY: COMPLEX:P23511_P25208_Q13952,
 &#39;COMPLEX:P68104_P85299_Q6R327_Q8TB45_Q9BVC4&#39;: Complex mTORC2: COMPLEX:P68104_P85299_Q6R327_Q8TB45_Q9BVC4,
 &#39;COMPLEX:P42345_Q8N122_Q8TB45_Q96B36_Q9BVC4&#39;: Complex mTORC1: COMPLEX:P42345_Q8N122_Q8TB45_Q96B36_Q9BVC4,
 &#39;COMPLEX:P63208_Q13616_Q9Y297&#39;: Complex SCF-betaTRCP: COMPLEX:P63208_Q13616_Q9Y297,
 &#39;COMPLEX:Q09472_Q92793&#39;: Complex CBP/p300: COMPLEX:Q09472_Q92793,
 &#39;COMPLEX:Q09472_Q92793_Q92831&#39;: Complex P300/PCAF: COMPLEX:Q09472_Q92793_Q92831,
 &#39;COMPLEX:Q13485_Q15796&#39;: Complex SMAD2/SMAD4: COMPLEX:Q13485_Q15796,
 &#39;COMPLEX:P84022_Q13485&#39;: Complex SMAD3/SMAD4: COMPLEX:P84022_Q13485,
 &#39;COMPLEX:P05412_Q13485&#39;: Complex SMAD4/JUN: COMPLEX:P05412_Q13485,
 &#39;COMPLEX:Q15796_Q9HAU4&#39;: Complex SMAD2/SMURF2: COMPLEX:Q15796_Q9HAU4,
 &#39;COMPLEX:O15105_Q01094_Q13547&#39;: Complex SMAD7/HDAC1/E2F-1: COMPLEX:O15105_Q01094_Q13547,
 &#39;COMPLEX:P19838_Q04206&#39;: Complex NfKb-p65/p50: COMPLEX:P19838_Q04206,
 &#39;COMPLEX:O14920_O15111&#39;: Complex IK
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 17699 characters</em></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">signor_interactions</span></code> function returns a list of arbitrary tuples that represent the most important properties of SIGNOR interaction records in a human readable way, and ready to be processed by the <code class="docutils literal notranslate"><span class="pre">pypath.core.network.Network</span></code> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>signor.signor_interactions()[:10]
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[SignorInteraction(source=&#39;O15530&#39;, target=&#39;O15530&#39;, source_isoform=None, target_isoform=None, source_type=&#39;protein&#39;, target_type=&#39;protein&#39;, effect=&#39;unknown&#39;, mechanism=&#39;phosphorylation&#39;, ncbi_tax_id=&#39;9606&#39;, pubmeds=&#39;10455013&#39;, direct=True, ptm_type=&#39;phosphorylation&#39;, ptm_residue=&#39;Ser396&#39;, ptm_motif=&#39;SSSSSSHsLSASDTG&#39;),
 SignorInteraction(source=&#39;Q9NQ66&#39;, target=&#39;CHEBI:18035&#39;, source_isoform=None, target_isoform=None, source_type=&#39;protein&#39;, target_type=&#39;smallmolecule&#39;, effect=&#39;up-regulates quantity&#39;, mechanism=&#39;&#39;, ncbi_tax_id=&#39;-1&#39;, pubmeds=&#39;23880553&#39;, direct=True, ptm_type=&#39;&#39;, ptm_residue=&#39;Small molecule catalysis&#39;, ptm_motif=&#39;&#39;),
 SignorInteraction(source=&#39;P62136&#39;, target=&#39;O15169&#39;, source_isoform=None, target_isoform=None, source_type=&#39;protein&#39;, target_type=&#39;protein&#39;, effect=&#39;down-regulates activity&#39;, mechanism=&#39;dephosphorylation&#39;, ncbi_tax_id=&#39;9606&#39;, pubmeds=&#39;17318175&#39;, direct=True, ptm_type=&#39;dephosphorylation&#39;, ptm_residue=&#39;Ser77&#39;, ptm_motif=&#39;YEPEGSAsPTPPYLK&#39;),
 SignorInteraction(sou
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 3285 characters</em></div>
</div>
<p>Note, the records above contain also enzyme-PTM data, hence the <code class="docutils literal notranslate"><span class="pre">signor.signor_enzyme_substrate</span></code> function only converts them to an intermediate format to make it easier to process for <code class="docutils literal notranslate"><span class="pre">pypath.core.enz_sub.EnzymeSubstrateAggregator</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>signor.signor_enzyme_substrate()[:2]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[{&#39;typ&#39;: &#39;phosphorylation&#39;,
  &#39;resnum&#39;: 396,
  &#39;instance&#39;: &#39;SSSSSSHSLSASDTG&#39;,
  &#39;substrate&#39;: &#39;O15530&#39;,
  &#39;start&#39;: 389,
  &#39;end&#39;: 403,
  &#39;kinase&#39;: &#39;O15530&#39;,
  &#39;resaa&#39;: &#39;S&#39;,
  &#39;motif&#39;: &#39;SSSSSSHSLSASDTG&#39;,
  &#39;enzyme_isoform&#39;: None,
  &#39;substrate_isoform&#39;: None,
  &#39;references&#39;: {&#39;10455013&#39;}},
 {&#39;typ&#39;: &#39;dephosphorylation&#39;,
  &#39;resnum&#39;: 77,
  &#39;instance&#39;: &#39;YEPEGSASPTPPYLK&#39;,
  &#39;substrate&#39;: &#39;O15169&#39;,
  &#39;start&#39;: 70,
  &#39;end&#39;: 84,
  &#39;kinase&#39;: &#39;P62136&#39;,
  &#39;resaa&#39;: &#39;S&#39;,
  &#39;motif&#39;: &#39;YEPEGSASPTPPYLK&#39;,
  &#39;enzyme_isoform&#39;: None,
  &#39;substrate_isoform&#39;: None,
  &#39;references&#39;: {&#39;17318175&#39;}}]
</pre></div></div>
</div>
<p>Finally, SIGNOR also assigns proteins to pathways. This information is intended for the OmniPath annotations database, and retrieved by the <code class="docutils literal notranslate"><span class="pre">signor.signor_pathway_annotations</span></code> function. This function returns a dict of sets which is typical for *_annotation* functions. This format requires practically no further processing.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>signor.signor_pathway_annotations()[&#39;O14733&#39;]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{SignorPathway(pathway=&#39;TNF alpha&#39;),
 SignorPathway(pathway=&#39;Toll like receptors&#39;)}
</pre></div></div>
</div>
<p>We haven’t mention all functions in the <code class="docutils literal notranslate"><span class="pre">inputs.signor</span></code> module. The rest of the functions retrieve additional information needed by the four functions above, and are of limited direct use for users. For example, <code class="docutils literal notranslate"><span class="pre">signor_protein_families</span></code> returns a dict with the internal ID and members of protein families; this data is used to process the interactions and complexes, but not too interesting on its own.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>signor.signor_protein_families()[&#39;SIGNOR-PF2&#39;]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;Q9HBW0&#39;, &#39;Q92633&#39;, &#39;Q9UBY5&#39;]
</pre></div></div>
</div>
</section>
<section id="Download-management">
<h2>Download management<a class="headerlink" href="#Download-management" title="Permalink to this heading">§</a></h2>
<section id="Cache-management-and-customization">
<h3>Cache management and customization<a class="headerlink" href="#Cache-management-and-customization" title="Permalink to this heading">§</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.omnipath.app</span></code> saves the databases to pickle dumps by default under the <code class="docutils literal notranslate"><span class="pre">~/.pypath/pickles/</span></code> directory and after the first build loads them from there. The very first build of each database might take quite long time (up to &gt;90 min in case of the OmniPath network or annotations) because of the large number of downloads. Subsequent builds will be much faster because <code class="docutils literal notranslate"><span class="pre">pypath</span></code> stores all the downloaded data in a local cache and downloads again only upon request from the user.
Loading the databases from pickle dumps takes only seconds. However if you want to build with different settings you should be aware to set up a different cache file name.</p>
</section>
<section id="Download-failures">
<h3>Download failures<a class="headerlink" href="#Download-failures" title="Permalink to this heading">§</a></h3>
<p>Issuing hundreds of requests to dozens of servers sooner or later comes with failures. These might happen just by accident, especially on slow networks, it is always recommended to try again. The</p>
<section id="Corrupted-cache-content">
<h4>Corrupted cache content<a class="headerlink" href="#Corrupted-cache-content" title="Permalink to this heading">§</a></h4>
<p>Sometimes a truncated or corrupted file remains in the cache, in this case you can use the context managers in <code class="docutils literal notranslate"><span class="pre">pypath.share.curl</span></code> to control the cache. E.g. if the download of the <em>DEPOD</em> database failed and keeps failing due to a corrupted file, use the <code class="docutils literal notranslate"><span class="pre">cache_delete_on</span></code> context:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import curl
from pypath.inputs import depod

with curl.cache_delete_on():
    depod = depod.depod_enzyme_substrate()
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">cache_off</span></code> context forces download even if a cache item is available; the <code class="docutils literal notranslate"><span class="pre">cache_print_on</span></code> context prints paths to the accessed cache files to the terminal, though the paths can always be found in the log; the <code class="docutils literal notranslate"><span class="pre">dry_run_on</span></code> context sets up the <code class="docutils literal notranslate"><span class="pre">pypath.share.curl.Curl</span></code> object and stops just before the actual download.</p>
</section>
<section id="Network-communication-issues:-look-into-the-curl-debug-log">
<h4>Network communication issues: look into the curl debug log<a class="headerlink" href="#Network-communication-issues:-look-into-the-curl-debug-log" title="Permalink to this heading">§</a></h4>
<p>Downloads might fail also due to TLS or HTTP errors, wrong headers or parameters, and many other reasons. In this case a full debug output from <code class="docutils literal notranslate"><span class="pre">curl</span></code> might be very useful. The <code class="docutils literal notranslate"><span class="pre">debug_on</span></code> context writes curl debug into the logfile:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import curl
from pypath.inputs import depod

with curl.debug_on():
    depod = depod.depod_enzyme_substrate()
</pre></div>
</div>
</div>
</section>
<section id="Timeouts">
<h4>Timeouts<a class="headerlink" href="#Timeouts" title="Permalink to this heading">§</a></h4>
<p>From the log we can find out if the download fails due to a timeout. In this case, the timeout parameters can be altered by a settings context. Apart from a timeout for the completion of the download, there is <code class="docutils literal notranslate"><span class="pre">curl_connect_timeout</span></code> (timeout for establishing connection to the server), and <code class="docutils literal notranslate"><span class="pre">curl_extended_timeout</span></code>, that is used for servers that are known to be exceptionally slow. Another parameter, <code class="docutils literal notranslate"><span class="pre">curl_retries</span></code> is the number of attempts before giving up. By default it’s 3, and that should
be more than enough.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import settings
from pypath.inputs import depod

with settings.context(curl_timeout = 360):
    depod = depod.depod_enzyme_substrate()
</pre></div>
</div>
</div>
</section>
<section id="Access-and-inspect-the-Curl-object">
<h4>Access and inspect the <code class="docutils literal notranslate"><span class="pre">Curl</span></code> object<a class="headerlink" href="#Access-and-inspect-the-Curl-object" title="Permalink to this heading">§</a></h4>
<p>Often the <code class="docutils literal notranslate"><span class="pre">Curl</span></code> object is created in a function from the <code class="docutils literal notranslate"><span class="pre">pypath.inputs</span></code> module, deep in a call stack, hence accessing it for investigation is difficult. Using the <code class="docutils literal notranslate"><span class="pre">preserve_on</span></code> context, the last <code class="docutils literal notranslate"><span class="pre">Curl</span></code> instance is kept under the <code class="docutils literal notranslate"><span class="pre">pypath.share.curl.LASTCURL</span></code> attribute:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import curl
from pypath.inputs import depod

with curl.preserve_on():
    depod = depod.depod_enzyme_substrate()

depod_curl = curl.LASTCURL
depod_curl
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;pypath.share.curl.Curl at 0x6947386dc8b0&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>depod_curl.url, depod_curl.req_headers, depod_curl.fileobj, depod_curl.status
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;http://depod.bioss.uni-freiburg.de/download/DEPOD_201405_human_phosphatase-substrate.mitab&#39;,
 [],
 &lt;_io.TextIOWrapper name=&#39;/home/denes/.pypath/cache/6a711369ecf9dcff8c5ed88996685b54-DEPOD_201405_human_phosphatase-substrate.mitab&#39; mode=&#39;r&#39; encoding=&#39;iso-8859-1&#39;&gt;,
 0)
</pre></div></div>
</div>
</section>
<section id="Is-it-failing-only-for-you?">
<h4>Is it failing only for you?<a class="headerlink" href="#Is-it-failing-only-for-you?" title="Permalink to this heading">§</a></h4>
<p>Okay, this is the one you should check first: we run almost all downloads in <code class="docutils literal notranslate"><span class="pre">pypath</span></code> daily, you can always check <a class="reference external" href="https://status.omnipathdb.org/">in the report</a> wether a particular function run successfully last night on our server. If it fails also in our daily build, it still can be a transient error that disappears within a few days, or it can be a permanent error. In the latter case, we first try to fix the issue in pypath (maybe the behaviour or the address of the third party server
has changed). If we have no way to fix it, we start <a class="reference external" href="https://rescued.omnipathdb.org/">hosting the data on our own server</a> and make pypath download it from there.</p>
</section>
<section id="Read-the-log">
<h4>Read the log<a class="headerlink" href="#Read-the-log" title="Permalink to this heading">§</a></h4>
<p>Above we mentioned a lot the pypath log. Here is how to access the log, see more details in the section about logging:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pypath
pypath.log()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 14:57:09] Welcome!
[2022-12-02 14:57:09] Logger started, logging into `/home/denes/pypath/notebooks/pypath_log/pypath-s3e92.log`.
[2022-12-02 14:57:09] Session `s3e92` started.
[2022-12-02 14:57:09] [pypath]
        - session ID: `s3e92`
        - working directory: `/home/denes/pypath/notebooks`
        - logfile: `/home/denes/pypath/notebooks/pypath_log/pypath-s3e92.log`
        - pypath version: 0.14.30
[2022-12-02 14:57:09] [curl] Creating Curl object to retrieve data from `https://www.ensembl.org/info/about/species.html`
[2022-12-02 14:57:09] [curl] Cache file path: `/home/denes/.pypath/cache/535b06d53a59e75bb693369bc5fdc556-species.html`
[2022-12-02 14:57:09] [curl] Cache file found, no need for download.
[2022-12-02 14:57:09] [curl] Opening plain text file `/home/denes/.pypath/cache/535b06d53a59e75bb693369bc5fdc556-species.html`.
[2022-12-02 14:57:09] [curl] Contents of `/home/denes/.pypath/cache/535b06d53a59e75bb693369bc5fdc556-species.html` has been read and the file has been closed.
[2022-1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 112963 characters</em></div>
</div>
</section>
<section id="TLS-(SSL,-HTTPS)-errors">
<h4>TLS (SSL, HTTPS) errors<a class="headerlink" href="#TLS-(SSL,-HTTPS)-errors" title="Permalink to this heading">§</a></h4>
<p>Failed to verify certificate, invalid, expired, self-signed, missing certificates. These might be the most common reasons why people open issues for our software. TLS is a method for encrypted, typically HTTP, communication. The server has a certificate and uses it to sign and encrypt the data before sending it to the client. The client trusts the server certificate because it is signed by another certificate. And that is signed by another one, and so on, until we reach a so called root
certificate that is known and trusted by the client. The number of root certificates used globally is so small that every single computer stores them locally and updates them time to time from trusted sources, such as the provider of the operating system, web browser or programming language. Having up-to-date certificate store and correctly configured TLS clients on your computer is your (or your system admin’s) duty, we can here only give a generic procedure to address these issues. In 97% of
the cases the issue is in your computer, but sometimes the server might be responsible. If you experience a TLS issue:</p>
<ul class="simple">
<li><p>Check the status of the server: initiate a scan at a free TLS checking service, such as <a class="reference external" href="https://www.ssllabs.com/ssltest/">SSL Labs</a>: look for any issue with the certificate chain, such as missing or expired certificates, old or too new ciphers not supported by your client, etc.</p></li>
<li><p>Identify the server that your client failed to establish a TLS connection to (in case of <code class="docutils literal notranslate"><span class="pre">pypath</span></code>, look into the log)</p></li>
<li><p>Identify your software that contains the TLS client: in case of <code class="docutils literal notranslate"><span class="pre">pypath</span></code>, it uses <code class="docutils literal notranslate"><span class="pre">pycurl</span></code>, a Python module built on <code class="docutils literal notranslate"><span class="pre">libcurl</span></code></p></li>
<li><p>Identify the provider of the client software: it can be PyPI, Anaconda, your operating system, etc.</p></li>
<li><p>Find out which certificate store that software uses: most of them uses the store from your operating system, but for example Java or Mozilla Firefox come with their own certificates</p></li>
<li><p>Check if the certificate store is up-to-date, update if necessary</p></li>
<li><p>Alternatively, identify the missing root certificate and add it manually to the store; you can also add a non-root certificate if the server has a serious issue and the chain can not be followed until a valid root certificate</p></li>
</ul>
<p>Please open TLS related issues for our software only if you</p>
<ul class="simple">
<li><p>Experience a server side issue with omnipathdb.org</p></li>
<li><p>You have a strong reason to think the reason is in the code written by us or can be easily fixed within our code</p></li>
</ul>
</section>
</section>
</section>
<section id="Resources">
<h2>Resources<a class="headerlink" href="#Resources" title="Permalink to this heading">§</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[192]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import resources
rc = resources.get_controller()
rc
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[192]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;pypath.resources.controller.ResourceController at 0x689944451d50&gt;
</pre></div></div>
</div>
<section id="Licenses">
<h3>Licenses<a class="headerlink" href="#Licenses" title="Permalink to this heading">§</a></h3>
<p>The license of SIGNOR is CC BY-SA, it allows commercial (for-profit) use:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[195]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rc.license(&#39;SIGNOR&#39;), rc.license(&#39;SIGNOR&#39;).commercial
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[195]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&lt;License CC BY-SA 4.0&gt;, True)
</pre></div></div>
</div>
</section>
<section id="Resource-information">
<h3>Resource information<a class="headerlink" href="#Resource-information" title="Permalink to this heading">§</a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[196]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rc[&#39;MatrixDB&#39;]
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[196]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;yearUsedRelease&#39;: 2015,
 &#39;releases&#39;: [2009, 2011, 2015],
 &#39;urls&#39;: {&#39;articles&#39;: [&#39;http://bioinformatics.oxfordjournals.org/content/25/5/690.long&#39;,
   &#39;http://nar.oxfordjournals.org/content/43/D1/D321.long&#39;,
   &#39;http://nar.oxfordjournals.org/content/39/suppl_1/D235.long&#39;],
  &#39;webpages&#39;: [&#39;http://matrixdb.univ-lyon1.fr/&#39;],
  &#39;omictools&#39;: [&#39;http://omictools.com/matrixdb-tool&#39;]},
 &#39;pubmeds&#39;: [19147664, 20852260, 25378329],
 &#39;taxons&#39;: [&#39;mammalia&#39;],
 &#39;annot&#39;: [&#39;experiment&#39;],
 &#39;recommend&#39;: [&#39;small, literature curated interaction resource; many interactions for&#39;,
  &#39;receptors and extracellular proteins&#39;],
 &#39;descriptions&#39;: [&#39;Protein data were imported from the UniProtKB/Swiss-Prot database (Bairoch et&#39;,
  &#39;al., 2005) and identified by UniProtKB/SwissProt accession numbers. In order to&#39;,
  &#39;list all the partners of a protein, interactions are associated by default to the&#39;,
  &#39;accession number of the human protein. The actual source species used in experiments is&#39;,
  &#39;indicated in the page repor
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 4479 characters</em></div>
</div>
</section>
<section id="Resource-definitions-for-a-certain-database-or-dataset">
<h3>Resource definitions for a certain database or dataset<a class="headerlink" href="#Resource-definitions-for-a-certain-database-or-dataset" title="Permalink to this heading">§</a></h3>
<div class="alert alert-block alert-success"><p>Note: This does not work yet for all databases and datasets, but likely in the near future this will be the preferred method to access resource definitions.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[197]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>rc.collect_enzyme_substrate()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[197]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;EnzymeSubstrateResource: phosphoELM&gt;,
 &lt;EnzymeSubstrateResource: dbPTM&gt;,
 &lt;EnzymeSubstrateResource: SIGNOR&gt;,
 &lt;EnzymeSubstrateResource: HPRD&gt;,
 &lt;EnzymeSubstrateResource: Li2012&gt;,
 &lt;EnzymeSubstrateResource: DEPOD&gt;,
 &lt;EnzymeSubstrateResource: PhosphoSite&gt;,
 &lt;EnzymeSubstrateResource: PhosphoNetworks&gt;,
 &lt;EnzymeSubstrateResource: MIMP&gt;,
 &lt;EnzymeSubstrateResource: ProtMapper&gt;,
 &lt;EnzymeSubstrateResource: KEA&gt;]
</pre></div></div>
</div>
<p>The resource definitions carry all information necessary to load the resource, for example:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[202]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>phosphoelm = rc.collect_enzyme_substrate()[0]
phosphoelm.input_method, phosphoelm.id_type_enzyme
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[202]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;phosphoelm.phosphoelm_enzyme_substrate&#39;, &#39;uniprot&#39;)
</pre></div></div>
</div>
</section>
</section>
<section id="Building-networks">
<h2>Building networks<a class="headerlink" href="#Building-networks" title="Permalink to this heading">§</a></h2>
<p>For this you will need the <code class="docutils literal notranslate"><span class="pre">Network</span></code> class from the <code class="docutils literal notranslate"><span class="pre">pypath.core.network</span></code> module which takes care about building and querying the network. Also you need the <code class="docutils literal notranslate"><span class="pre">pypath.resources.network</span></code> module where you find a number of predefined input settings organized in larger categories (e.g. activity flow, enzyme-substrate, transcriptional regulation, etc). These input settings will tell <code class="docutils literal notranslate"><span class="pre">pypath</span></code> how to download and process the data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import network
from pypath.resources import network as netres
</pre></div>
</div>
</div>
<p>For example the <code class="docutils literal notranslate"><span class="pre">netres.pathway</span></code> is a collection of databases which fit into the activity flow concept, i.e. one protein either stimulates or inhibits the other. It is a dictionary with names as keys and the input settings as values:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>netres.pathway
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;trip&#39;: &lt;NetworkResource: TRIP (post_translational, activity_flow)&gt;,
 &#39;spike&#39;: &lt;NetworkResource: SPIKE (post_translational, activity_flow)&gt;,
 &#39;signalink3&#39;: &lt;NetworkResource: SignaLink3 (post_translational, activity_flow)&gt;,
 &#39;guide2pharma&#39;: &lt;NetworkResource: Guide2Pharma (post_translational, activity_flow)&gt;,
 &#39;ca1&#39;: &lt;NetworkResource: CA1 (post_translational, activity_flow)&gt;,
 &#39;arn&#39;: &lt;NetworkResource: ARN (post_translational, activity_flow)&gt;,
 &#39;nrf2&#39;: &lt;NetworkResource: NRF2ome (post_translational, activity_flow)&gt;,
 &#39;macrophage&#39;: &lt;NetworkResource: Macrophage (post_translational, activity_flow)&gt;,
 &#39;death&#39;: &lt;NetworkResource: DeathDomain (post_translational, activity_flow)&gt;,
 &#39;pdz&#39;: &lt;NetworkResource: PDZBase (post_translational, activity_flow)&gt;,
 &#39;signor&#39;: &lt;NetworkResource: SIGNOR (post_translational, activity_flow)&gt;,
 &#39;adhesome&#39;: &lt;NetworkResource: Adhesome (post_translational, activity_flow)&gt;,
 &#39;icellnet&#39;: &lt;NetworkResource: ICELLNET (post_translational, activity_flow)&gt;,
 &#39;celltalkdb&#39;: &lt;Net
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1864 characters</em></div>
</div>
<p>Such a dictionary you can pass to the <code class="docutils literal notranslate"><span class="pre">load</span></code> method of the <code class="docutils literal notranslate"><span class="pre">network.Network</span></code> object. Then it will download the data from the original sources, translate the identifiers and merge the networks. Pypath stores all downloaded data in a cache, by default <code class="docutils literal notranslate"><span class="pre">~/.pypath/cache</span></code> in your user’s home directory. For this reason when you load a resource for the first time it might take long but next time will be faster as data will be fetched from the cache. First create a <code class="docutils literal notranslate"><span class="pre">pypath.network.Network</span></code>
object, then build the network:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n = network.Network()
n.load(netres.pathway)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 6833 nodes, 25607 interactions&gt;
</pre></div></div>
</div>
<p>You can add more resource sets a similar way:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.load(netres.enzyme_substrate)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 7979 nodes, 35550 interactions&gt;
</pre></div></div>
</div>
<p>To load one single resource simply pass the <code class="docutils literal notranslate"><span class="pre">NetworkResource</span></code> directly:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.load(netres.interaction[&#39;matrixdb&#39;])
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 8002 nodes, 35748 interactions&gt;
</pre></div></div>
</div>
<section id="Which-network-datasets-are-pre-defined-in-pypath?">
<h3>Which network datasets are pre-defined in pypath?<a class="headerlink" href="#Which-network-datasets-are-pre-defined-in-pypath?" title="Permalink to this heading">§</a></h3>
<p>You can find all the pre-defined datasets in the <code class="docutils literal notranslate"><span class="pre">pypath.resources.network</span></code> module. This module currently is a wrapper around an older module, <code class="docutils literal notranslate"><span class="pre">pypath.resources.data_formats</span></code>, the actual definitions are written in this latter. As already we mentined above, the <code class="docutils literal notranslate"><span class="pre">pathway</span></code> dataset contains the literature curated activity flow resources. This was the original focus of pypath and OmniPath, however since then we added a great variety of other kinds of resource definitions. Here we give an
overview of these.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.pathway</span></code>: activity flow networks with literature references</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.activity_flow</span></code>: synonym for <code class="docutils literal notranslate"><span class="pre">pathway</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.pathway_noref</span></code>: activity flow networks without literature references</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.pathway_all</span></code>: all activity flow data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.ptm</span></code>: enzyme-substrate interaction networks with literature references</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.enzyme_substrate</span></code>: synonym for <code class="docutils literal notranslate"><span class="pre">ptm</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.ptm_noref</span></code>: enzyme-substrate networks without literature references</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.ptm_all</span></code>: all enzyme-substrate data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.interaction</span></code>: undirected interactions from both literature curated and high-throughput collections (e.g. IntAct, BioGRID)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.interaction_misc</span></code>: undirected, high-scale interaction networks without the constraint of having any literature reference (e.g. the unbiased human interactome screen from the Vidal lab)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.transcription_onebyone</span></code>: transcriptional regulation databases (TF-target interactions) with all databases downloaded directly and processed by <code class="docutils literal notranslate"><span class="pre">pypath</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.transcription</span></code>: transcriptional regulation only from the DoRothEA data</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.mirna_target</span></code>: miRNA-mRNA interactions from literature curated resources</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.tf_mirna</span></code>: transcriptional regulation of miRNA from literature curated resources</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.lncrna_protein</span></code>: lncRNA-protein interactions from literature curated datasets</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.ligand_receptor</span></code>: ligand-receptor interactions from both literature curated and other kinds of resources</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.pathwaycommons</span></code>: the PathwayCommons database</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.reaction</span></code>: process description databases; not guaranteed to work at this moment</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.reaction_misc</span></code>: alternative definitions to load process description databases; not guaranteed to work at this moment</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pypath.resources.network.small_molecule_protein</span></code>: signaling interactions between small molecules and proteins</p></li>
</ul>
<p>To see the list of the resources in a dataset, you can check the dict keys or the <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute of each element:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>netres.pathway.keys()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dict_keys([&#39;trip&#39;, &#39;spike&#39;, &#39;signalink3&#39;, &#39;guide2pharma&#39;, &#39;ca1&#39;, &#39;arn&#39;, &#39;nrf2&#39;, &#39;macrophage&#39;, &#39;death&#39;, &#39;pdz&#39;, &#39;signor&#39;, &#39;adhesome&#39;, &#39;icellnet&#39;, &#39;celltalkdb&#39;, &#39;cellchatdb&#39;, &#39;connectomedb&#39;, &#39;talklr&#39;, &#39;cellinker&#39;, &#39;scconnect&#39;, &#39;hpmr&#39;, &#39;cellphonedb&#39;, &#39;ramilowski2015&#39;, &#39;lrdb&#39;, &#39;baccin2019&#39;])
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>[resource.name for resource in netres.pathway.values()]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;TRIP&#39;,
 &#39;SPIKE&#39;,
 &#39;SignaLink3&#39;,
 &#39;Guide2Pharma&#39;,
 &#39;CA1&#39;,
 &#39;ARN&#39;,
 &#39;NRF2ome&#39;,
 &#39;Macrophage&#39;,
 &#39;DeathDomain&#39;,
 &#39;PDZBase&#39;,
 &#39;SIGNOR&#39;,
 &#39;Adhesome&#39;,
 &#39;ICELLNET&#39;,
 &#39;CellTalkDB&#39;,
 &#39;CellChatDB&#39;,
 &#39;connectomeDB2020&#39;,
 &#39;talklr&#39;,
 &#39;Cellinker&#39;,
 &#39;scConnect&#39;,
 &#39;HPMR&#39;,
 &#39;CellPhoneDB&#39;,
 &#39;Ramilowski2015&#39;,
 &#39;LRdb&#39;,
 &#39;Baccin2019&#39;]
</pre></div></div>
</div>
<p>The resource definitions above carry all the information about how to load the resource: which function to call, how to process the identifiers, references, directions, and all other attributes from the input. E.g. which column from SPIKE corresponds to the source node? Which identifier type is used? It is the second column, and it has gene symbols in it:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>netres.pathway[&#39;spike&#39;].networkinput.id_col_a, netres.pathway[&#39;spike&#39;].networkinput.id_type_a
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(1, &#39;genesymbol&#39;)
</pre></div></div>
</div>
</section>
<section id="The-Network-object">
<h3>The <code class="docutils literal notranslate"><span class="pre">Network</span></code> object<a class="headerlink" href="#The-Network-object" title="Permalink to this heading">§</a></h3>
<p>Once you built a network you can use it for various purposes and write your own scripts for further processing or analysis. Below we create a <code class="docutils literal notranslate"><span class="pre">Network</span></code> object and populate it with the <code class="docutils literal notranslate"><span class="pre">pathway</span></code> dataset.</p>
<div class="alert alert-block alert-warning"><p>Warning: it is recommended to access databases by the manager. Running the code below takes really long and does not save or reload the database, it builds a fresh copy each time.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import network
from pypath.resources import network as netres

n = network.Network()
n.load(netres.pathway)
n
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Network: 6833 nodes, 25607 interactions&gt;
</pre></div></div>
</div>
<p>Almost all data is stored as a dict node pairs vs. interactions in <code class="docutils literal notranslate"><span class="pre">Network.interactions</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.interactions
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{(&lt;Entity: TRPC1&gt;,
  &lt;Entity: KCNMA1&gt;): &lt;Interaction: TRPC1 ============= KCNMA1 [Evidences: TRIP (2 references)]&gt;,
 (&lt;Entity: TRPC1&gt;,
  &lt;Entity: PPP3CA&gt;): &lt;Interaction: TRPC1 ============= PPP3CA [Evidences: TRIP (1 references)]&gt;,
 (&lt;Entity: CALM2&gt;,
  &lt;Entity: TRPC1&gt;): &lt;Interaction: CALM2 =======(-)==&gt; TRPC1 [Evidences: TRIP (3 references)]&gt;,
 (&lt;Entity: CALM3&gt;,
  &lt;Entity: TRPC1&gt;): &lt;Interaction: CALM3 =======(-)==&gt; TRPC1 [Evidences: TRIP (3 references)]&gt;,
 (&lt;Entity: CALM1&gt;,
  &lt;Entity: TRPC1&gt;): &lt;Interaction: CALM1 =======(-)==&gt; TRPC1 [Evidences: TRIP (3 references)]&gt;,
 (&lt;Entity: CASP1&gt;,
  &lt;Entity: TRPC1&gt;): &lt;Interaction: CASP1 ============= TRPC1 [Evidences: TRIP (1 references)]&gt;,
 (&lt;Entity: TRPC1&gt;,
  &lt;Entity: CASP4&gt;): &lt;Interaction: TRPC1 ============= CASP4 [Evidences: TRIP (1 references)]&gt;,
 (&lt;Entity: TRPC1&gt;,
  &lt;Entity: CACNA1C&gt;): &lt;Interaction: TRPC1 ============= CACNA1C [Evidences: TRIP (1 references)]&gt;,
 (&lt;Entity: TRPC1&gt;,
  &lt;Entity: CAV1&gt;): &lt;Interaction: TRPC1 &lt;=(+)======== CAV1 [Ev
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 118492 characters</em></div>
</div>
<p>The dict under <code class="docutils literal notranslate"><span class="pre">Network.nodes</span></code> is kept in sync with the interactions, and facilitates node access. Keys are primary identifiers (for proteins UniProt IDs by default), values are <code class="docutils literal notranslate"><span class="pre">Entity</span></code> objects:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.nodes
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;P48995&#39;: &lt;Entity: TRPC1&gt;,
 &#39;Q12791&#39;: &lt;Entity: KCNMA1&gt;,
 &#39;Q08209&#39;: &lt;Entity: PPP3CA&gt;,
 &#39;P0DP24&#39;: &lt;Entity: CALM2&gt;,
 &#39;P0DP25&#39;: &lt;Entity: CALM3&gt;,
 &#39;P0DP23&#39;: &lt;Entity: CALM1&gt;,
 &#39;P29466&#39;: &lt;Entity: CASP1&gt;,
 &#39;P49662&#39;: &lt;Entity: CASP4&gt;,
 &#39;Q13936&#39;: &lt;Entity: CACNA1C&gt;,
 &#39;Q03135&#39;: &lt;Entity: CAV1&gt;,
 &#39;P56539&#39;: &lt;Entity: CAV3&gt;,
 &#39;Q14247&#39;: &lt;Entity: CTTN&gt;,
 &#39;P14416&#39;: &lt;Entity: DRD2&gt;,
 &#39;P11532&#39;: &lt;Entity: DMD&gt;,
 &#39;P11362&#39;: &lt;Entity: FGFR1&gt;,
 &#39;Q02790&#39;: &lt;Entity: FKBP4&gt;,
 &#39;Q86YM7&#39;: &lt;Entity: HOMER1&gt;,
 &#39;Q9NSC5&#39;: &lt;Entity: HOMER3&gt;,
 &#39;Q99750&#39;: &lt;Entity: MDFI&gt;,
 &#39;Q14571&#39;: &lt;Entity: ITPR2&gt;,
 &#39;Q14573&#39;: &lt;Entity: ITPR3&gt;,
 &#39;P29966&#39;: &lt;Entity: MARCKS&gt;,
 &#39;Q13255&#39;: &lt;Entity: GRM1&gt;,
 &#39;P20591&#39;: &lt;Entity: MX1&gt;,
 &#39;P62166&#39;: &lt;Entity: NCS1&gt;,
 &#39;Q96D31&#39;: &lt;Entity: ORAI1&gt;,
 &#39;Q96SN7&#39;: &lt;Entity: ORAI2&gt;,
 &#39;Q9BRQ5&#39;: &lt;Entity: ORAI3&gt;,
 &#39;P11171&#39;: &lt;Entity: EPB41&gt;,
 &#39;P61586&#39;: &lt;Entity: RHOA&gt;,
 &#39;Q9Y225&#39;: &lt;Entity: RNF24&gt;,
 &#39;P21817&#39;: &lt;Entity: RYR1&gt;,
 &#39;P16615&#39;: &lt;Entity: ATP2A2&gt;,
 &#39;Q93084&#39;: &lt;Entity: ATP2A3&gt;,
 &#39;P60880&#39;: &lt;Entity: SNAP25&gt;,
 &#39;Q13586&#39;: &lt;Entity: STI
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 30573 characters</em></div>
</div>
<p>An interaction between a pair of entities can be accessed:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.interaction(&#39;EGF&#39;, &#39;EGFR&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Interaction: EGFR &lt;=(+)======== EGF [Evidences: Baccin2019, CellTalkDB, Fantom5, Guide2Pharma, HPMR, HPRD, ICELLNET, LRdb, Ramilowski2015, SIGNOR, SPIKE, SignaLink3, cellsignal.com, connectomeDB2020 (17 references)]&gt;
</pre></div></div>
</div>
<p>Similarly, individual nodes can be looked up:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>n.entity(&#39;EGFR&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Entity: EGFR&gt;
</pre></div></div>
</div>
<p>Labels (gene symbols for proteins by default), identifiers (such as UniProt IDs) and <code class="docutils literal notranslate"><span class="pre">Entity</span></code> objects can be used to refer to nodes. Each node carries some basic information:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>egfr = n.entity(&#39;EGFR&#39;)
egfr.identifier, egfr.label, egfr.entity_type, egfr.id_type, egfr.taxon
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;P00533&#39;, &#39;EGFR&#39;, &#39;protein&#39;, &#39;uniprot&#39;, 9606)
</pre></div></div>
</div>
<p>Interactions feature a number of methods to access various information, such as their types, direction, effect, resources, references, etc. The very same methods are also available for the whole network. Below we only show a few examples of these methods.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia = n.interaction(&#39;EGF&#39;, &#39;EGFR&#39;)
ia
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Interaction: EGFR &lt;=(+)======== EGF [Evidences: Baccin2019, CellTalkDB, Fantom5, Guide2Pharma, HPMR, HPRD, ICELLNET, LRdb, Ramilowski2015, SIGNOR, SPIKE, SignaLink3, cellsignal.com, connectomeDB2020 (17 references)]&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_resource_names()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Baccin2019&#39;,
 &#39;CellTalkDB&#39;,
 &#39;HPMR&#39;,
 &#39;ICELLNET&#39;,
 &#39;LRdb&#39;,
 &#39;SIGNOR&#39;,
 &#39;SPIKE&#39;,
 &#39;SignaLink3&#39;,
 &#39;connectomeDB2020&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_references()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&lt;Reference: 10085134&gt;,
 &lt;Reference: 10209155&gt;,
 &lt;Reference: 10788520&gt;,
 &lt;Reference: 12093292&gt;,
 &lt;Reference: 12297050&gt;,
 &lt;Reference: 12620237&gt;,
 &lt;Reference: 12648462&gt;,
 &lt;Reference: 15620700&gt;,
 &lt;Reference: 16274239&gt;,
 &lt;Reference: 17145710&gt;,
 &lt;Reference: 19531499&gt;,
 &lt;Reference: 20458382&gt;,
 &lt;Reference: 21071413&gt;,
 &lt;Reference: 23331499&gt;,
 &lt;Reference: 3494473&gt;,
 &lt;Reference: 6289330&gt;,
 &lt;Reference: 8639530&gt;}
</pre></div></div>
</div>
<p>This is a valid direction for this interaction:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_direction((&#39;EGF&#39;, &#39;EGFR&#39;))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<p>The opposite direction is not supported by any of the resources:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_direction((&#39;EGFR&#39;, &#39;EGF&#39;))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False
</pre></div></div>
</div>
<p>However, some resources provide no direction information, these are classified as <em>“undirected”</em>:</p>
<p>ia.get_direction(‘undirected’)</p>
<p>We can check which resources are those exactly:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_direction(&#39;undirected&#39;, sources = True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;HPMR&#39;, &#39;SPIKE&#39;}
</pre></div></div>
</div>
<p>Effect signs (stimulation, inhibition) are available in a similar way. The first one of the Boolean values mean stimulation (activation), the second one inhibition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_sign((&#39;EGF&#39;, &#39;EGFR&#39;))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[True, False]
</pre></div></div>
</div>
<p>Which resources support the effect signs:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_sign((&#39;EGF&#39;, &#39;EGFR&#39;), sources = True)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[{&#39;SIGNOR&#39;, &#39;SPIKE&#39;, &#39;SignaLink3&#39;}, set()]
</pre></div></div>
</div>
<p>Many methods start by <code class="docutils literal notranslate"><span class="pre">get_...</span></code>, such as:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_interaction_types()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;post_translational&#39;}
</pre></div></div>
</div>
<p>Others are called <code class="docutils literal notranslate"><span class="pre">..._by_...</span></code>, these combine two <code class="docutils literal notranslate"><span class="pre">get_...</span></code> methods:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.references_by_resource()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;ICELLNET&#39;: {&lt;Reference: 8639530&gt;},
 &#39;SIGNOR&#39;: {&lt;Reference: 12297050&gt;, &lt;Reference: 12648462&gt;},
 &#39;SignaLink3&#39;: {&lt;Reference: 10085134&gt;,
  &lt;Reference: 10209155&gt;,
  &lt;Reference: 19531499&gt;,
  &lt;Reference: 21071413&gt;,
  &lt;Reference: 23331499&gt;},
 &#39;Baccin2019&#39;: {&lt;Reference: 10788520&gt;,
  &lt;Reference: 12093292&gt;,
  &lt;Reference: 12297050&gt;,
  &lt;Reference: 12620237&gt;,
  &lt;Reference: 15620700&gt;,
  &lt;Reference: 16274239&gt;,
  &lt;Reference: 6289330&gt;},
 &#39;LRdb&#39;: {&lt;Reference: 10788520&gt;,
  &lt;Reference: 12093292&gt;,
  &lt;Reference: 12297050&gt;,
  &lt;Reference: 12620237&gt;,
  &lt;Reference: 15620700&gt;,
  &lt;Reference: 16274239&gt;,
  &lt;Reference: 6289330&gt;},
 &#39;SPIKE&#39;: {&lt;Reference: 12297050&gt;,
  &lt;Reference: 17145710&gt;,
  &lt;Reference: 20458382&gt;,
  &lt;Reference: 3494473&gt;},
 &#39;CellTalkDB&#39;: {&lt;Reference: 12093292&gt;},
 &#39;connectomeDB2020&#39;: {&lt;Reference: 10788520&gt;,
  &lt;Reference: 12093292&gt;,
  &lt;Reference: 12297050&gt;,
  &lt;Reference: 12620237&gt;,
  &lt;Reference: 15620700&gt;,
  &lt;Reference: 16274239&gt;,
  &lt;Reference: 6289330&gt;},
 &#39;HPMR&#39;: {&lt;Reference: 6289330&gt;}}
</pre></div></div>
</div>
<p>And all these methods accept the same filtering parameters. E.g. if you are interested only in certain resources, it’s possible to restrict the query to those. For example, the two resources below provide no positive sign interaction:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_interactions_positive(resources = {&#39;ICELLNET&#39;, &#39;HPMR&#39;})
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
()
</pre></div></div>
</div>
<p>While some other resources do:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_interactions_positive(resources = {&#39;SignaLink3&#39;})
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
((&lt;Entity: EGF&gt;, &lt;Entity: EGFR&gt;),)
</pre></div></div>
</div>
<p>Or see the references that do or do not provide effect sign:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ia.get_references(effect = True), ia.get_references(effect = False)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
({&lt;Reference: 10085134&gt;,
  &lt;Reference: 10209155&gt;,
  &lt;Reference: 12297050&gt;,
  &lt;Reference: 12648462&gt;,
  &lt;Reference: 19531499&gt;,
  &lt;Reference: 20458382&gt;,
  &lt;Reference: 21071413&gt;,
  &lt;Reference: 23331499&gt;},
 {&lt;Reference: 10085134&gt;,
  &lt;Reference: 10209155&gt;,
  &lt;Reference: 10788520&gt;,
  &lt;Reference: 12093292&gt;,
  &lt;Reference: 12297050&gt;,
  &lt;Reference: 12620237&gt;,
  &lt;Reference: 12648462&gt;,
  &lt;Reference: 15620700&gt;,
  &lt;Reference: 16274239&gt;,
  &lt;Reference: 17145710&gt;,
  &lt;Reference: 19531499&gt;,
  &lt;Reference: 20458382&gt;,
  &lt;Reference: 21071413&gt;,
  &lt;Reference: 23331499&gt;,
  &lt;Reference: 3494473&gt;,
  &lt;Reference: 6289330&gt;,
  &lt;Reference: 8639530&gt;})
</pre></div></div>
</div>
</section>
<section id="Network-in-pandas.DataFrame">
<h3>Network in <em>pandas.DataFrame</em><a class="headerlink" href="#Network-in-pandas.DataFrame" title="Permalink to this heading">§</a></h3>
<p>Contents of a <code class="docutils literal notranslate"><span class="pre">pypath.core.network.Network</span></code> object can be exported to a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
cu = omnipath.db.get_db(&#39;curated&#39;)
cu.make_df()
cu.df
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>id_a</th>
      <th>id_b</th>
      <th>type_a</th>
      <th>type_b</th>
      <th>directed</th>
      <th>effect</th>
      <th>type</th>
      <th>dmodel</th>
      <th>sources</th>
      <th>references</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>P48995</td>
      <td>Q12791</td>
      <td>protein</td>
      <td>protein</td>
      <td>False</td>
      <td>0</td>
      <td>post_translational</td>
      <td>{activity_flow}</td>
      <td>{TRIP}</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>P48995</td>
      <td>Q08209</td>
      <td>protein</td>
      <td>protein</td>
      <td>False</td>
      <</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 3616 characters</em></div>
</div>
</section>
</section>
<section id="Translating-identifiers">
<h2>Translating identifiers<a class="headerlink" href="#Translating-identifiers" title="Permalink to this heading">§</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping</span></code> module is for ID translation, most of the time you can simply call the <code class="docutils literal notranslate"><span class="pre">map_name</span></code> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import mapping
mapping.map_name(&#39;P00533&#39;, &#39;uniprot&#39;, &#39;genesymbol&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;EGFR&#39;}
</pre></div></div>
</div>
<p>By default the <code class="docutils literal notranslate"><span class="pre">map_name</span></code> function returns a <code class="docutils literal notranslate"><span class="pre">set</span></code> because it accounts for ambiguous mapping. However most often the ID translation is unambiguous, and you want to retrieve only one ID. The <code class="docutils literal notranslate"><span class="pre">map_name0</span></code> returns a string, even in case of ambiguity, it returns a random element from the resulted set:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.map_name0(&#39;GABARAPL3&#39;, &#39;genesymbol&#39;, &#39;uniprot&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;Q9BY60&#39;
</pre></div></div>
</div>
<p>Molecules have large variety of identifiers, but in pypath two identifier types are special:</p>
<ul class="simple">
<li><p>The <em>primary identifier</em> defines the molecule category, e.g. if UniProt is the primary identifier for proteins, then a protein is anything that has a UniProt ID</p></li>
<li><p>The <em>label</em> is a human readable identifier, for proteins it’s gene symbol</p></li>
</ul>
<p>The primary ID and label types are configured for each molecule type (protein, miRNA, drug, etc) in the module settings. The <code class="docutils literal notranslate"><span class="pre">mapping</span></code> module provides shortcuts to translate between these identifiers: <code class="docutils literal notranslate"><span class="pre">label</span></code> and <code class="docutils literal notranslate"><span class="pre">id_from_label</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.label(&#39;O75385&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;ULK1&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.id_from_label(&#39;ULK1&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;O75385&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.id_from_label0(&#39;ULK1&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;O75385&#39;
</pre></div></div>
</div>
<p>Multiple IDs can be translated in one call, however, it’s not possible to know certainly which output corresponds to which input.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.map_names([&#39;ULK1&#39;, &#39;EGFR&#39;, &#39;SMAD2&#39;], &#39;genesymbol&#39;, &#39;uniprot&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;O75385&#39;, &#39;P00533&#39;, &#39;Q15796&#39;}
</pre></div></div>
</div>
<p>The default organism is defined in the module settings, it is human by default. Translating for other organisms requires the <code class="docutils literal notranslate"><span class="pre">ncbi_tax_id</span></code> argument. Most of the functions in <code class="docutils literal notranslate"><span class="pre">pypath</span></code> accepts also common or latin names, but <code class="docutils literal notranslate"><span class="pre">map_name</span></code> accepts only numeric taxon IDs for efficiency. Let’s translate a mouse identifier:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.map_name(&#39;Smad2&#39;, &#39;genesymbol&#39;, &#39;uniprot&#39;, ncbi_tax_id = 10090)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q62432&#39;}
</pre></div></div>
</div>
<p>If no direct translation table is available between two ID types, <code class="docutils literal notranslate"><span class="pre">pypath</span></code> will try to translate by an intermediate ID type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.map_name(&#39;8408&#39;, &#39;entrez&#39;, &#39;genesymbol&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;ULK1&#39;}
</pre></div></div>
</div>
<p>Behind the scenes the <code class="docutils literal notranslate"><span class="pre">chain_map</span></code> function is called:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m = mapping.get_mapper()
m.chain_map(&#39;8408&#39;, id_type = &#39;entrez&#39;, target_id_type = &#39;genesymbol&#39;, by_id_type = &#39;uniprot&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;ULK1&#39;}
</pre></div></div>
</div>
<p>And the procedure corresponds to the following:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>mapping.map_names(
    mapping.map_name(&#39;8408&#39;, &#39;entrez&#39;, &#39;uniprot&#39;),
    &#39;uniprot&#39;,
    &#39;genesymbol&#39;,
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;ULK1&#39;}
</pre></div></div>
</div>
<section id="Pre-defined-ID-translation-tables">
<h3>Pre-defined ID translation tables<a class="headerlink" href="#Pre-defined-ID-translation-tables" title="Permalink to this heading">§</a></h3>
<p>A number of mapping tables are pre-defined, these load automatically on demand, and are removed from the memory if not used for some time (5 minutes by default). New mapping tables are saved directly into pickle files in the cache for a quick reload. Tables are either organism specific (hence loaded for each organism one-by-one), or non-organism specific, such as drug IDs (<code class="docutils literal notranslate"><span class="pre">pypath</span></code> uses integer <code class="docutils literal notranslate"><span class="pre">0</span></code> in this case in place of the numeric NCBI Taxonomy ID). The identifier translation data is
retrieved from the following sources:</p>
<ul class="simple">
<li><p>UniProt legacy API (main UniProt API until autumn 2022): <code class="docutils literal notranslate"><span class="pre">internals.input_formats.UniprotMapping</span></code></p></li>
<li><p>UniProt uploadlists API (also outdated, replaced by the new UniProt API): <code class="docutils literal notranslate"><span class="pre">internals.inputs_formats.UniprotListMapping</span></code></p></li>
<li><p>Ensembl Biomart: <code class="docutils literal notranslate"><span class="pre">internals.input_formats.BiomartMapping</span></code> and <code class="docutils literal notranslate"><span class="pre">internals.input_formats.ArrayMapping</span></code> (for microarray probes)</p></li>
<li><p>Protein Ontology Consortium: <code class="docutils literal notranslate"><span class="pre">internals.input_formats.ProMapping</span></code></p></li>
<li><p>UniChem: <code class="docutils literal notranslate"><span class="pre">internals.input_formats.UnichemMapping</span></code></p></li>
<li><p>Arbitrary files: <code class="docutils literal notranslate"><span class="pre">internals.input_formats.FileMapping</span></code> (this class is used to process data from miRBase, some files from the UniProt FTP site, and also user defined, custom cases)</p></li>
</ul>
<p>Some of the classes above are instantiated in <code class="docutils literal notranslate"><span class="pre">internals.maps</span></code>, but most of the instances are created on the fly when loading a mapping table in <code class="docutils literal notranslate"><span class="pre">utils.mapping.MapReader</span></code>. This latter class is responsible to take a table definition and load a <code class="docutils literal notranslate"><span class="pre">utils.mapping.MappingTable</span></code> instance. The whole process is managed by <code class="docutils literal notranslate"><span class="pre">utils.mapping.Mapper</span></code>, this is the object all the ID translation queries are dispatched to. It has a method to list the defined ID translation tables:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m = mapping.get_mapper()
m.mapping_tables()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[MappingTableDefinition(id_type_a=&#39;embl&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_class=&#39;UniprotMapping&#39;, resource_id_type_a=&#39;database(embl)&#39;, resource_id_type_b=&#39;id&#39;),
 MappingTableDefinition(id_type_a=&#39;genesymbol&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_class=&#39;UniprotMapping&#39;, resource_id_type_a=&#39;genes(PREFERRED)&#39;, resource_id_type_b=&#39;id&#39;),
 MappingTableDefinition(id_type_a=&#39;genesymbol-syn&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_class=&#39;UniprotMapping&#39;, resource_id_type_a=&#39;genes(ALTERNATIVE)&#39;, resource_id_type_b=&#39;id&#39;),
 MappingTableDefinition(id_type_a=&#39;entrez&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_class=&#39;UniprotMapping&#39;, resource_id_type_a=&#39;database(geneid)&#39;, resource_id_type_b=&#39;id&#39;),
 MappingTableDefinition(id_type_a=&#39;hgnc&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_class=&#39;UniprotMapping&#39;, resource_id_type_a=&#39;database(HGNC)&#39;, resource_id_type_b=&#39;id&#39;),
 MappingTableDefinition(id_type_a=&#39;refseqp&#39;, id_type_b=&#39;uniprot&#39;, resource=&#39;uniprot&#39;, input_cl
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 22169 characters</em></div>
</div>
<p>Pypath uses synonyms to refer to ID types: these are intended to be short, clear and lowercase for ease of use. Most of the synonyms are defined in <code class="docutils literal notranslate"><span class="pre">internals.input_formats</span></code>, in the <code class="docutils literal notranslate"><span class="pre">AC_QUERY</span></code>, <code class="docutils literal notranslate"><span class="pre">AC_MAPPING</span></code>, <code class="docutils literal notranslate"><span class="pre">BIOMART_MAPPING</span></code>, <code class="docutils literal notranslate"><span class="pre">PRO_MAPPING</span></code> and <code class="docutils literal notranslate"><span class="pre">ARRAY_MAPPING</span></code> dictionaries. UniChem ID types are used exactly as provided by UniChem. To list all available ID types (below <em>pypath</em> is the synonym used here, <em>original</em> is the name in the original resource):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m = mapping.get_mapper()
m.id_types()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{IdType(pypath=&#39;MedChemExpress&#39;, original=&#39;MedChemExpress&#39;),
 IdType(pypath=&#39;actor&#39;, original=&#39;actor&#39;),
 IdType(pypath=&#39;affy&#39;, original=&#39;affy&#39;),
 IdType(pypath=&#39;affymetrix&#39;, original=&#39;affymetrix&#39;),
 IdType(pypath=&#39;agilent&#39;, original=&#39;agilent&#39;),
 IdType(pypath=&#39;alzforum&#39;, original=&#39;Alzforum_mut&#39;),
 IdType(pypath=&#39;araport&#39;, original=&#39;Araport&#39;),
 IdType(pypath=&#39;atlas&#39;, original=&#39;atlas&#39;),
 IdType(pypath=&#39;bindingdb&#39;, original=&#39;bindingdb&#39;),
 IdType(pypath=&#39;brenda&#39;, original=&#39;brenda&#39;),
 IdType(pypath=&#39;carotenoiddb&#39;, original=&#39;carotenoiddb&#39;),
 IdType(pypath=&#39;cgnc&#39;, original=&#39;CGNC&#39;),
 IdType(pypath=&#39;chebi&#39;, original=&#39;chebi&#39;),
 IdType(pypath=&#39;chembl&#39;, original=&#39;chembl&#39;),
 IdType(pypath=&#39;chemicalbook&#39;, original=&#39;chemicalbook&#39;),
 IdType(pypath=&#39;clinicaltrials&#39;, original=&#39;clinicaltrials&#39;),
 IdType(pypath=&#39;codelink&#39;, original=&#39;codelink&#39;),
 IdType(pypath=&#39;comptox&#39;, original=&#39;comptox&#39;),
 IdType(pypath=&#39;dailymed&#39;, original=&#39;dailymed&#39;),
 IdType(pypath=&#39;dailymed_old&#39;, original=&#39;dailymed_old&#39;),
 IdType(py
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 6649 characters</em></div>
</div>
</section>
</section>
<section id="Homology-translation">
<h2>Homology translation<a class="headerlink" href="#Homology-translation" title="Permalink to this heading">§</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">utils.homology</span></code> module handles translation of data between organism by orthologous gene pairs. Its most important function is <code class="docutils literal notranslate"><span class="pre">translate</span></code>. The source organism is human by default, the target must be provided, below we use mouse (NCBI Taxonomy 10090):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import homology
homology.translate(&#39;P00533&#39;, target = 10090)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q01279&#39;}
</pre></div></div>
</div>
<p>ID translation and homology translation are integrated, hence not only UniProt IDs can be translated:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>homology.translate(&#39;EGFR&#39;, target = 10090, id_type = &#39;genesymbol&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Egfr&#39;}
</pre></div></div>
</div>
<p>This module uses data from NCBI HomoloGene and Ensembl. The latter covers more organisms, and accepts some parameters (high confidence, one-to-one vs. one-to-many mapping). These parameters can be controlled by the settings module, or passed to the functions above and below. For some reason the settings below fail to find any ortholog of our example protein:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>homology.translate(&#39;P00533&#39;, target = 10090, homologene = False, ensembl = False, ensembl_hc = False, ensembl_types = &#39;one2one&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
set()
</pre></div></div>
</div>
<section id="Homology-translation-tables-as-dictionaries">
<h3>Homology translation tables as dictionaries<a class="headerlink" href="#Homology-translation-tables-as-dictionaries" title="Permalink to this heading">§</a></h3>
<p>The translation tables are available as dicts of sets, these are convenient for use outside of pypath:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>human_mouse_genesymbols = homology.get_dict(target = &#39;mouse&#39;, id_type = &#39;genesymbol&#39;)
human_mouse_genesymbols[&#39;EGFR&#39;]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Egfr&#39;}
</pre></div></div>
</div>
</section>
<section id="Homology-translation-data-frames">
<h3>Homology translation data frames<a class="headerlink" href="#Homology-translation-data-frames" title="Permalink to this heading">§</a></h3>
<p>Similarly, <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>s are available:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>human_mouse_genesymbols = homology.get_df(target = &#39;mouse&#39;, id_type = &#39;genesymbol&#39;)
human_mouse_genesymbols
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>target</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MICOS13</td>
      <td>Micos13</td>
    </tr>
    <tr>
      <th>1</th>
      <td>FAT4</td>
      <td>Fat4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>RARS2</td>
      <td>Rars2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ZFP36L2</td>
      <td>Zfp36l2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>LAMC1</td>
      <td>Lamc1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>24843</th>
      <td>KRTAP4-16</td>
      <td>Krtap5-2</td>
    </tr>
    <tr>
      <th>24844</th></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1386 characters</em></div>
</div>
</section>
</section>
<section id="Taxonomy">
<h2>Taxonomy<a class="headerlink" href="#Taxonomy" title="Permalink to this heading">§</a></h2>
<p>Organisms matter everywhere, both in the input, output and processing parts of pypath. For this reason we created a utility module to deal with translation of organism identifiers. We prefer NCBI Taxonomy IDs as the primary organism identifier. These are simple numbers, 9606 is human, 10090 is mouse, etc. Many databases use common English names or latin (scientific) names. Then some databases use custom codes, such as <em>hsapiens</em> in Ensmebl (first letter of genus name + species name, without
space, all lowercase); <em>hsa</em> in miRBase and KEGG (first letter of genus name, first two letters of species name). The <code class="docutils literal notranslate"><span class="pre">pypath.utils.taxonomy</span></code> module features some convenient functions for handling all these names.</p>
<section id="Translating-to-NCBI-Taxonomy,-scientific-names-and-common-names">
<h3>Translating to NCBI Taxonomy, scientific names and common names<a class="headerlink" href="#Translating-to-NCBI-Taxonomy,-scientific-names-and-common-names" title="Permalink to this heading">§</a></h3>
<p>The most often used is <code class="docutils literal notranslate"><span class="pre">ensure_ncbi_tax_id</span></code>, which returns the NCBI Taxonomy ID for any comprehensible input:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import taxonomy
taxonomy.ensure_ncbi_tax_id(&#39;human&#39;), taxonomy.ensure_ncbi_tax_id(&#39;H sapiens&#39;), taxonomy.ensure_ncbi_tax_id(&#39;hsapiens&#39;), taxonomy.ensure_ncbi_tax_id(9606), taxonomy.ensure_ncbi_tax_id(&#39;Homo sapiens&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(9606, 9606, 9606, 9606, 9606)
</pre></div></div>
</div>
<p>To access scientific names or common names:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>taxonomy.ensure_latin_name(&#39;cow&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;Bos taurus&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>taxonomy.ensure_common_name(&#39;Erithacus rubecula&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;European robin&#39;
</pre></div></div>
</div>
</section>
<section id="Organism-from-UniProt-ID">
<h3>Organism from UniProt ID<a class="headerlink" href="#Organism-from-UniProt-ID" title="Permalink to this heading">§</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">uniprot_taxid</span></code> function returns the taxonomy ID for a SwissProt ID. Unfortunately it does not work for TrEMBL IDs, that would require to keep too much data in memory.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>taxonomy.ensure_latin_name(taxonomy.uniprot_taxid(&#39;P53104&#39;))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;Saccharomyces cerevisiae&#39;
</pre></div></div>
</div>
</section>
</section>
<section id="UniProt">
<h2>UniProt<a class="headerlink" href="#UniProt" title="Permalink to this heading">§</a></h2>
<p>UniProt is a huge, diverse resource that is essential for <em>pypath</em> as we use it as a reference set for proteomes and it provides ID translation data. Its input module <code class="docutils literal notranslate"><span class="pre">pypath.inputs.uniprot</span></code> is already more complex than an average input module. It harbors a little database manager that loads and unloads tables on demand, ensuring fast and convenient operation. Further services are available in the <code class="docutils literal notranslate"><span class="pre">pypath.utils.uniprot</span></code> module.</p>
<section id="The-UniProt-input-module">
<h3>The UniProt input module<a class="headerlink" href="#The-UniProt-input-module" title="Permalink to this heading">§</a></h3>
<section id="All-UniProt-IDs-for-one-organism">
<h4>All UniProt IDs for one organism<a class="headerlink" href="#All-UniProt-IDs-for-one-organism" title="Permalink to this heading">§</a></h4>
<p>The complete set of UniProt IDs for an organism is considered to be the proteome of the organism, and it is used in many procedures across <em>pypath</em>. All SwissProt IDs, all TrEMBL IDs or both together can be retrieved:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[119]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
(
    len(iuniprot.all_uniprots(organism = 10090)),
    len(iuniprot.all_swissprots(organism = 10090)),
    len(iuniprot.all_trembls(organism = 10090)),
)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[119]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(86440, 17131, 69300)
</pre></div></div>
</div>
</section>
<section id="UniProt-ID-format-validation">
<h4>UniProt ID format validation<a class="headerlink" href="#UniProt-ID-format-validation" title="Permalink to this heading">§</a></h4>
<p>UniProt defines a format for its accessions, any string can be checked against this template to tell if it’s possibly a valid ID:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[124]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
iuniprot.valid_uniprot(&#39;A0A8D0H0C2&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[124]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
</section>
<section id="UniProt-ID-validation">
<h4>UniProt ID validation<a class="headerlink" href="#UniProt-ID-validation" title="Permalink to this heading">§</a></h4>
<p>Another functions check if an ID indeed exists in UniProt. These functions require loading the list of all UniProt IDs for the organism, hence calling them the first time might take even a few minutes (in case new download is necessary). Subsequent calls will be much faster.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[125]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
iuniprot.is_uniprot(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[125]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[122]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.is_swissprot(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[122]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
True
</pre></div></div>
</div>
<p>If the organism doesn’t match:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[123]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.is_uniprot(&#39;P00533&#39;, organism = 10090)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[123]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
False
</pre></div></div>
</div>
</section>
<section id="Single-UniProt-protein-datasheet">
<h4>Single UniProt protein datasheet<a class="headerlink" href="#Single-UniProt-protein-datasheet" title="Permalink to this heading">§</a></h4>
<p>Raw contents of protein datasheets can be retrieved. The structure is a Python list with tuples of two elements, the first is the tag of the line, the second is the line content.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[126]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
iuniprot.protein_datasheet(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[126]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[(&#39;ID&#39;, &#39;EGFR_HUMAN              Reviewed;        1210 AA.&#39;),
 (&#39;AC&#39;,
  &#39;P00533; O00688; O00732; P06268; Q14225; Q68GS5; Q92795; Q9BZS2; Q9GZX1;&#39;),
 (&#39;AC&#39;, &#39;Q9H2C9; Q9H3C9; Q9UMD7; Q9UMD8; Q9UMG5;&#39;),
 (&#39;DT&#39;, &#39;21-JUL-1986, integrated into UniProtKB/Swiss-Prot.&#39;),
 (&#39;DT&#39;, &#39;01-NOV-1997, sequence version 2.&#39;),
 (&#39;DT&#39;, &#39;12-OCT-2022, entry version 283.&#39;),
 (&#39;DE&#39;, &#39;RecName: Full=Epidermal growth factor receptor {ECO:0000305};&#39;),
 (&#39;DE&#39;, &#39;EC=2.7.10.1;&#39;),
 (&#39;DE&#39;, &#39;AltName: Full=Proto-oncogene c-ErbB-1;&#39;),
 (&#39;DE&#39;, &#39;AltName: Full=Receptor tyrosine-protein kinase erbB-1;&#39;),
 (&#39;DE&#39;, &#39;Flags: Precursor;&#39;),
 (&#39;GN&#39;, &#39;Name=EGFR {ECO:0000312|HGNC:HGNC:3236}; Synonyms=ERBB, ERBB1, HER1;&#39;),
 (&#39;OS&#39;, &#39;Homo sapiens (Human).&#39;),
 (&#39;OC&#39;,
  &#39;Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;&#39;),
 (&#39;OC&#39;,
  &#39;Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;&#39;),
 (&#39;OC&#39;, &#39;Homo.&#39;),
 (&#39;OX&#39;, &#39;NCBI_TaxID=9606;&#39;),
 (&#39;RN&#39;, &#39;[1]&#39;),
 (&#39;RP&#39;,
  &#39;NUCLEOTIDE SEQUENCE [MRNA] (ISOFORM
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 58080 characters</em></div>
</div>
</section>
<section id="History-of-UniProt-records">
<h4>History of UniProt records<a class="headerlink" href="#History-of-UniProt-records" title="Permalink to this heading">§</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[131]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
egfr_history = list(iuniprot.uniprot_history(&#39;P00533&#39;))
egfr_history
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[131]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[UniprotRecordHistory(entry_version=&#39;283&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2022_04&#39;, date=&#39;2022-10-12&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;),
 UniprotRecordHistory(entry_version=&#39;282&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2022_03&#39;, date=&#39;2022-08-03&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;),
 UniprotRecordHistory(entry_version=&#39;281&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2022_02&#39;, date=&#39;2022-05-25&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;),
 UniprotRecordHistory(entry_version=&#39;280&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2022_01&#39;, date=&#39;2022-02-23&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;),
 UniprotRecordHistory(entry_version=&#39;279&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2021_04&#39;, date=&#39;2021-09-29&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;),
 UniprotRecordHistory(entry_version=&#39;278&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 50933 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[132]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_recent_version(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[132]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
UniprotRecordHistory(entry_version=&#39;283&#39;, sequence_version=&#39;2&#39;, entry_name=&#39;EGFR_HUMAN&#39;, database=&#39;Swiss-Prot&#39;, number=&#39;2022_04&#39;, date=&#39;2022-10-12&#39;, replaces=&#39;&#39;, replaced_by=&#39;&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[133]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_history_recent_datasheet(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[133]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[(&#39;ID&#39;, &#39;EGFR_HUMAN              Reviewed;        1210 AA.&#39;),
 (&#39;AC&#39;,
  &#39;P00533; O00688; O00732; P06268; Q14225; Q68GS5; Q92795; Q9BZS2; Q9GZX1;&#39;),
 (&#39;AC&#39;, &#39;Q9H2C9; Q9H3C9; Q9UMD7; Q9UMD8; Q9UMG5;&#39;),
 (&#39;DT&#39;, &#39;21-JUL-1986, integrated into UniProtKB/Swiss-Prot.&#39;),
 (&#39;DT&#39;, &#39;01-NOV-1997, sequence version 2.&#39;),
 (&#39;DT&#39;, &#39;12-OCT-2022, entry version 283.&#39;),
 (&#39;DE&#39;, &#39;RecName: Full=Epidermal growth factor receptor {ECO:0000305};&#39;),
 (&#39;DE&#39;, &#39;EC=2.7.10.1;&#39;),
 (&#39;DE&#39;, &#39;AltName: Full=Proto-oncogene c-ErbB-1;&#39;),
 (&#39;DE&#39;, &#39;AltName: Full=Receptor tyrosine-protein kinase erbB-1;&#39;),
 (&#39;DE&#39;, &#39;Flags: Precursor;&#39;),
 (&#39;GN&#39;, &#39;Name=EGFR {ECO:0000312|HGNC:HGNC:3236}; Synonyms=ERBB, ERBB1, HER1;&#39;),
 (&#39;OS&#39;, &#39;Homo sapiens (Human).&#39;),
 (&#39;OC&#39;,
  &#39;Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi; Mammalia;&#39;),
 (&#39;OC&#39;,
  &#39;Eutheria; Euarchontoglires; Primates; Haplorrhini; Catarrhini; Hominidae;&#39;),
 (&#39;OC&#39;, &#39;Homo.&#39;),
 (&#39;OX&#39;, &#39;NCBI_TaxID=9606;&#39;),
 (&#39;RN&#39;, &#39;[1]&#39;),
 (&#39;RP&#39;,
  &#39;NUCLEOTIDE SEQUENCE [MRNA] (ISOFORM
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 58080 characters</em></div>
</div>
<p>The functions above are able to retrieve the latest datasheet of deleted UniProt records. However, they are slow as several queries are performed to process a single protein.</p>
</section>
<section id="UniProt-legacy-API">
<h4>UniProt legacy API<a class="headerlink" href="#UniProt-legacy-API" title="Permalink to this heading">§</a></h4>
<p>UniProt deployed its new API in the autumn of 2022, since then the old API is available as a legacy option. In <em>pypath</em> this API is well supported. It is accessed by the <code class="docutils literal notranslate"><span class="pre">inputs.uniprot.uniprot_data</span></code> function, though higher level functions are more convenient for the users. For the function above, a list of fields can be passed. By default it uses only SwissProt. The output is a dict of dicts with fields as top level keys and UniProt IDs as second level keys. The results often contain notes,
additional info in parentheses, prefixes and postfixes for identifiers, that are not needed in every situation. Using <code class="docutils literal notranslate"><span class="pre">uniprot_preprocess</span></code> instead of <code class="docutils literal notranslate"><span class="pre">uniprot_data</span></code> cleans up some of this clutter.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[135]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
iuniprot.uniprot_data(field = (&#39;family&#39;, &#39;keywords&#39;, &#39;transmembrane&#39;))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[135]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;family&#39;: {&#39;P63120&#39;: &#39;Peptidase A2 family, HERV class-II K(HML-2) subfamily&#39;,
  &#39;Q96EC8&#39;: &#39;YIP1 family&#39;,
  &#39;Q6ZMS4&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q8N8L2&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q3MIS6&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q86UK7&#39;: &#39;ZNF598 family&#39;,
  &#39;Q6P280&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q969W1&#39;: &#39;DHHC palmitoyltransferase family&#39;,
  &#39;O14978&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q15937&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q9P2J8&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q8IUH4&#39;: &#39;DHHC palmitoyltransferase family, AKR/ZDHHC17 subfamily&#39;,
  &#39;Q9Y2D9&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q14588&#39;: &#39;Krueppel C2H2-type zinc-finger protein family&#39;,
  &#39;Q6XR72&#39;: &#39;Cation diffusion facilitator (CDF) transporter (TC 2.A.4) family, SLC30A subfamily&#39;,
  &#39;P58557&#39;: &#39;Endoribonuclease YbeY family&#39;,
  &#39;Q9Y5A9&#39;: &#39;YTHDF family, YTHDF2 subfamily&#39;,
  &#39;Q8N9L1
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 501768 characters</em></div>
</div>
</section>
<section id="Processed-UniProt-annotations">
<h4>Processed UniProt annotations<a class="headerlink" href="#Processed-UniProt-annotations" title="Permalink to this heading">§</a></h4>
<p>For a few important fields we have dedicated processing functions with the aim of making their format cleaner and better usable. Sometimes even these do an imperfect job, and certain fields are badly truncated or contain residual fragments of the stripped labels.</p>
<div class="alert alert-block alert-success"><p>Note: All the data presented below is part of the OmniPath annotations database, the recommended way to access it is by the database manager.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[136]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.inputs import uniprot as iuniprot
iuniprot.uniprot_taxonomy()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[136]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;P00521&#39;: {&#39;Abelson murine leukemia virus&#39;},
 &#39;P03333&#39;: {&#39;Abelson murine leukemia virus&#39;},
 &#39;H8ZM73&#39;: {&#39;Abies balsamea&#39;, &#39;Balsam fir&#39;, &#39;Pinus balsamea&#39;},
 &#39;H8ZM71&#39;: {&#39;Abies balsamea&#39;, &#39;Balsam fir&#39;, &#39;Pinus balsamea&#39;},
 &#39;Q9MV51&#39;: {&#39;Abies firma&#39;, &#39;Momi fir&#39;},
 &#39;O81086&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;O24474&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;O24475&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;O64404&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;O64405&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;Q948Z0&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;Q9M7D1&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;Q9M7D0&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;O22340&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;Q9M7C9&#39;: {&#39;Abies grandis&#39;, &#39;Grand fir&#39;, &#39;Pinus grandis&#39;},
 &#39;Q5K3V1&#39;: {&#39;Abies homolepis&#39;, &#39;Nikko fir&#39;},
 &#39;P21715&#39;: {&#39;Abrothrix jelskii&#39;, &#39;Akodon jelskii&#39;, &#34;Jelski&#39;s altiplano mouse&#34;},
 &#39;P11140&#39;: {&#39;Abru
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 56985 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[139]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_ncbi_taxids_2()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[139]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{648330: Taxon(ncbi_id=648330, latin=&#39;Aedes albopictus densovirus (isolate Boublik/1994)&#39;, english=&#39;AalDNV&#39;, latin_synonym=None),
 10804: Taxon(ncbi_id=10804, latin=&#39;Adeno-associated virus 2&#39;, english=&#39;AAV-2&#39;, latin_synonym=None),
 648242: Taxon(ncbi_id=648242, latin=&#39;Adeno-associated virus 2 (isolate Srivastava/1982)&#39;, english=&#39;AAV-2&#39;, latin_synonym=None),
 118452: Taxon(ncbi_id=118452, latin=&#39;Abacion magnum&#39;, english=&#39;Millipede&#39;, latin_synonym=None),
 72259: Taxon(ncbi_id=72259, latin=&#39;Abaeis nicippe&#39;, english=&#39;Sleepy orange butterfly&#39;, latin_synonym=&#39;Eurema nicippe&#39;),
 102642: Taxon(ncbi_id=102642, latin=&#39;Abax parallelepipedus&#39;, english=&#39;Ground beetle&#39;, latin_synonym=None),
 392897: Taxon(ncbi_id=392897, latin=&#39;Abalistes stellaris&#39;, english=&#39;Starry triggerfish&#39;, latin_synonym=&#39;Balistes stellaris&#39;),
 75332: Taxon(ncbi_id=75332, latin=&#39;Abbottina rivularis&#39;, english=&#39;Chinese false gudgeon&#39;, latin_synonym=&#39;Gobio rivularis&#39;),
 515833: Taxon(ncbi_id=515833, latin=&#39;Abdopus aculeatus&#39;, engl
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 118050 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[140]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_locations()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[140]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q96EC8&#39;: {UniprotLocation(location=&#39;Golgi apparatus membrane&#39;, features=(&#39;Multi-pass membrane protein&#39;,))},
 &#39;Q6ZMS4&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q8N8L2&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q15916&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q3MIS6&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q6P280&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q969W1&#39;: {UniprotLocation(location=&#39;Endoplasmic reticulum membrane&#39;, features=(&#39;Multi-pass membrane protein&#39;,))},
 &#39;O14978&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q66K41&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q15937&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q9P2J8&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q8ND82&#39;: {UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;Q9NP64&#39;: {UniprotLocation(location=&#39;Nucleolus&#39;, features=None),
  UniprotLocation(location=&#39;Nucleus&#39;, features=None)},
 &#39;P
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 143466 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[141]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_keywords()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[141]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;P63120&#39;: {UniprotKeyword(keyword=&#39;Aspartyl protease&#39;),
  UniprotKeyword(keyword=&#39;Autocatalytic cleavage&#39;),
  UniprotKeyword(keyword=&#39;ERV&#39;),
  UniprotKeyword(keyword=&#39;Hydrolase&#39;),
  UniprotKeyword(keyword=&#39;Protease&#39;),
  UniprotKeyword(keyword=&#39;Reference proteome&#39;),
  UniprotKeyword(keyword=&#39;Ribosomal frameshifting&#39;),
  UniprotKeyword(keyword=&#39;Transposable element&#39;)},
 &#39;Q96EC8&#39;: {UniprotKeyword(keyword=&#39;Acetylation&#39;),
  UniprotKeyword(keyword=&#39;Alternative splicing&#39;),
  UniprotKeyword(keyword=&#39;Golgi apparatus&#39;),
  UniprotKeyword(keyword=&#39;Membrane&#39;),
  UniprotKeyword(keyword=&#39;Phosphoprotein&#39;),
  UniprotKeyword(keyword=&#39;Reference proteome&#39;),
  UniprotKeyword(keyword=&#39;Transmembrane&#39;),
  UniprotKeyword(keyword=&#39;Transmembrane helix&#39;)},
 &#39;Q6ZMS4&#39;: {UniprotKeyword(keyword=&#39;Metal-binding&#39;),
  UniprotKeyword(keyword=&#39;Nucleus&#39;),
  UniprotKeyword(keyword=&#39;Phosphoprotein&#39;),
  UniprotKeyword(keyword=&#39;Reference proteome&#39;),
  UniprotKeyword(keyword=&#39;Repeat&#39;),
  UniprotKeyword(keyword=&#39;Zinc&#39;),
  Unipro
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 445111 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[142]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_families()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[142]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;P63120&#39;: {UniprotFamily(family=&#39;Peptidase A2&#39;, subfamily=&#39;HERV class-II K(HML-2)&#39;)},
 &#39;Q96EC8&#39;: {UniprotFamily(family=&#39;YIP1&#39;, subfamily=None)},
 &#39;Q6ZMS4&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q8N8L2&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q3MIS6&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q86UK7&#39;: {UniprotFamily(family=&#39;ZNF598&#39;, subfamily=None)},
 &#39;Q6P280&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q969W1&#39;: {UniprotFamily(family=&#39;DHHC palmitoyltransferase&#39;, subfamily=None)},
 &#39;O14978&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q15937&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q9P2J8&#39;: {UniprotFamily(family=&#39;Krueppel C2H2-type zinc-finger protein&#39;, subfamily=None)},
 &#39;Q8IUH4&#39;: {UniprotFamily(family=&#39;DHHC palmitoyltransferase&#39;,
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 77892 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[143]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_tissues()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[143]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q15916&#39;: {UniprotTissue(tissue=&#39;Brain&#39;, level=&#39;high&#39;),
  UniprotTissue(tissue=&#39;Wide&#39;, level=&#39;high&#39;)},
 &#39;Q969W1&#39;: {UniprotTissue(tissue=&#39;Wide&#39;, level=&#39;undefined&#39;)},
 &#39;O14978&#39;: {UniprotTissue(tissue=&#39;Brain&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Colon&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Heart&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Kidney&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Leukocyte&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Liver&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Lung&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Ovary&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Pancreas&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Placenta&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Prostate&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Skeletal muscle&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Small intestine&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Spleen&#39;, level=&#39;undefined&#39;),
  UniprotTissue(tissue=&#39;Testis&#39;, level=&#39;undefined&#39;),
  Uniprot
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 318790 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[144]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>iuniprot.uniprot_topology()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[144]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q96EC8&#39;: {UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=2, end=84),
  UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=137, end=146),
  UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=206, end=212),
  UniprotTopology(topology=&#39;Lumenal&#39;, start=106, end=115),
  UniprotTopology(topology=&#39;Lumenal&#39;, start=168, end=184),
  UniprotTopology(topology=&#39;Lumenal&#39;, start=234, end=236),
  UniprotTopology(topology=&#39;Transmembrane&#39;, start=85, end=105),
  UniprotTopology(topology=&#39;Transmembrane&#39;, start=116, end=136),
  UniprotTopology(topology=&#39;Transmembrane&#39;, start=147, end=167),
  UniprotTopology(topology=&#39;Transmembrane&#39;, start=185, end=205),
  UniprotTopology(topology=&#39;Transmembrane&#39;, start=213, end=233)},
 &#39;Q969W1&#39;: {UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=1, end=77),
  UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=138, end=198),
  UniprotTopology(topology=&#39;Cytoplasmic&#39;, start=288, end=377),
  UniprotTopology(topology=&#39;Lumenal&#39;, start=99, end=116),
  UniprotTopology(topology=&#39;Lumenal&#39;, start=220,
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 544230 characters</em></div>
</div>
</section>
</section>
<section id="The-UniProt-utils-module">
<h3>The UniProt utils module<a class="headerlink" href="#The-UniProt-utils-module" title="Permalink to this heading">§</a></h3>
<section id="Datasheets">
<h4>Datasheets<a class="headerlink" href="#Datasheets" title="Permalink to this heading">§</a></h4>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.utils.uniprot</span></code> module is an API around UniProt protein datasheets. It is not suitable for bulk retrieval: that would work but take really long time. Calling its bulk methods with more than a few dozens or hundreds of proteins might take minutes, as it downloads protein datasheets one-by-one. To retrieve the full datasheets of one or more proteins use <code class="docutils literal notranslate"><span class="pre">query</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[153]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import uniprot
uniprot.query(&#39;P00533&#39;, &#39;O75385&#39;, &#39;Q14457&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[153]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;UniProt datasheet P00533 (EGFR)&gt;,
 &lt;UniProt datasheet O75385 (ULK1)&gt;,
 &lt;UniProt datasheet Q14457 (BECN1)&gt;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[154]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ulk1 = uniprot.query(&#39;O75385&#39;)
ulk1
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[154]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;UniProt datasheet O75385 (ULK1)&gt;
</pre></div></div>
</div>
<p>Many attributes are available from the datasheet objects, just a few examples:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[156]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ulk1.weight, ulk1.length, ulk1.subcellular_location, ulk1.sequence
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[156]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(112631,
 1050,
 &#39;Cytoplasm, cytosol. Preautophagosomal structure. Note=Under starvation conditions, is localized to puncate structures primarily representing the isolation membrane that sequesters a portion of the cytoplasm resulting in the formation of an autophagosome.&#39;,
 &#39;MEPGRGGTETVGKFEFSRKDLIGHGAFAVVFKGRHREKHDLEVAVKCINKKNLAKSQTLLGKEIKILKELKHENIVALYDFQEMANSVYLVMEYCNGGDLADYLHAMRTLSEDTIRLFLQQIAGAMRLLHSKGIIHRDLKPQNILLSNPAGRRANPNSIRVKIADFGFARYLQSNMMAATLCGSPMYMAPEVIMSQHYDGKADLWSIGTIVYQCLTGKAPFQASSPQDLRLFYEKNKTLVPTIPRETSAPLRQLLLALLQRNHKDRMDFDEFFHHPFLDASPSVRKSPPVPVPSYPSSGSGSSSSSSSTSHLASPPSLGEMQQLQKTLASPADTAGFLHSSRDSGGSKDSSCDTDDFVMVPAQFPGDLVAEAPSAKPPPDSLMCSGSSLVASAGLESHGRTPSPSPPCSSSPSPSGRAGPFSSSRCGASVPIPVPTQVQNYQRIERNLQSPTQFQTPRSSAIRRSGSTSPLGFARASPSPPAHAEHGGVLARKMSLGGGRPYTPSPQVGTIPERPGWSGTPSPQGAEMRGGRSPRPGSSAPEHSPRTSGLGCRLHSAPNLSDLHVVRPKLPKPPTDPLGAVFSPPQASPPQPSHGLQSCRNLRGSPKLPDFLQRNPLPPILGSPTKAVPSFDFPKTPSSQNLLALLARQGVVMTPPRNRTLPDLSEVGPFHGQPLGPGLRPGEDPKGPFGRSFSTSRLTDLLLKAAFGTQAPDPGSTESLQEK
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1329 characters</em></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">collect</span></code> function collects certain features for a set of proteins.</p>
<div class="alert alert-block alert-warning"><p>Warning: This is a really inefficient way of retrieving data from UniProt. If you work with more than a handful of proteins, go for pypath.inputs.uniprot_data instead.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[158]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>uniprot.collect([&#39;P00533&#39;, &#39;O75385&#39;, &#39;Q14457&#39;], &#39;weight&#39;, &#39;length&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[158]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
OrderedDict([(&#39;ac&#39;, [&#39;P00533&#39;, &#39;O75385&#39;, &#39;Q14457&#39;]),
             (&#39;weight&#39;, [134277, 112631, 51896]),
             (&#39;length&#39;, [1210, 1050, 450])])
</pre></div></div>
</div>
</section>
<section id="Tables">
<h4>Tables<a class="headerlink" href="#Tables" title="Permalink to this heading">§</a></h4>
<p>UniProt data can be printed to the console in a tabular format:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[159]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>uniprot.print_features([&#39;P00533&#39;, &#39;O75385&#39;, &#39;Q14457&#39;], &#39;weight&#39;, &#39;length&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
╒═══════╤════════╤══════════╤══════════╕
│   No. │ ac     │   weight │   length │
╞═══════╪════════╪══════════╪══════════╡
│     1 │ P00533 │   134277 │     1210 │
├───────┼────────┼──────────┼──────────┤
│     2 │ O75385 │   112631 │     1050 │
├───────┼────────┼──────────┼──────────┤
│     3 │ Q14457 │    51896 │      450 │
╘═══════╧════════╧══════════╧══════════╛
</pre></div></div>
</div>
<p>There is a shortcut to print essential characterization of proteins as such a table. The <code class="docutils literal notranslate"><span class="pre">info</span></code> function is really useful if you get to a set of proteins at some point of your analysis and you want to quickly check what kind they are. To iterate through multiple groups of proteins, use <code class="docutils literal notranslate"><span class="pre">utils.uniprot.browse</span></code>. The columns and format of these tables can be customized by <code class="docutils literal notranslate"><span class="pre">kwargs</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[160]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>uniprot.info([&#39;P00533&#39;, &#39;O75385&#39;, &#39;Q14457&#39;])
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
=====&gt; [3 proteins] &lt;=====
╒═══════╤════════╤══════════════╤══════════╤══════════╤═════════════╤══════════════╤════════════╤══════════════╕
│   No. │ ac     │ genesymbol   │   length │   weight │ full_name   │ function_o   │ keywords   │ subcellula   │
│       │        │              │          │          │             │ r_genecard   │            │ r_location   │
│       │        │              │          │          │             │ s            │            │              │
╞═══════╪════════╪══════════════╪══════════╪══════════╪═════════════╪══════════════╪════════════╪══════════════╡
│     1 │ P00533 │ EGFR         │     1210 │   134277 │ Epidermal   │ Receptor     │ 3D-        │ Cell         │
│       │        │              │          │          │ growth      │ tyrosine     │ structure, │ membrane;    │
│       │        │              │          │          │ factor      │ kinase       │ Alternativ │ Single-      │
│       │        │              │          │          │ receptor    │
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 20254 characters</em></div>
</div>
</section>
</section>
<section id="Sanitizing-UniProt-IDs">
<h3>Sanitizing UniProt IDs<a class="headerlink" href="#Sanitizing-UniProt-IDs" title="Permalink to this heading">§</a></h3>
<p>It is important to know that the ID translation module always do a number of checks when translating to UniProt IDs. Unless the <code class="docutils literal notranslate"><span class="pre">uniprot_cleanup</span></code> parameter is disabled. It translates secondary IDs to primary, attempts to map TrEMBL IDs to SwissProts by gene symbols, removes IDs of other organisms or invalid format. To exploit this behaviour it’s enough to map from UniProt to UniProt:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[162]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import mapping
mapping.map_name(&#39;Q9UQ28&#39;, &#39;uniprot&#39;, &#39;uniprot&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[162]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;O75385&#39;}
</pre></div></div>
</div>
</section>
</section>
<section id="Enzyme-substrate-interactions">
<h2>Enzyme-substrate interactions<a class="headerlink" href="#Enzyme-substrate-interactions" title="Permalink to this heading">§</a></h2>
<p>The database is an instance of <code class="docutils literal notranslate"><span class="pre">pypath.core.enz_sub.EnzymeSubstrateAggregator</span></code> class. The database is built with the default or current configuration by the <code class="docutils literal notranslate"><span class="pre">core.enz_sub.get_db</span></code> method.</p>
<div class="alert alert-block alert-warning"><p>Warning: it is recommended to access databases by the manager. Running the code below takes really long and does not save or reload the database, it builds a fresh copy each time.</p>
</div><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import enz_sub
es = enz_sub.get_db()
</pre></div>
</div>
</div>
<p>Instead, let’s acquire the database from the manager:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
es = omnipath.db.get_db(&#39;enz_sub&#39;)
</pre></div>
</div>
</div>
<p>The database itself is stored as a dictionary (<code class="docutils literal notranslate"><span class="pre">EnzymeSubstrateAggregator.enz_sub</span></code>) with pairs of proteins as keys and a list of special objects representing enzyme-substrate interactions as values. These can be accessed by pairs of labels, identifiers or <code class="docutils literal notranslate"><span class="pre">Entity</span></code> objects, e.g. mTOR phosphorylates AKT1:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>es[(&#39;MTOR&#39;, &#39;AKT1&#39;)]
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;MTOR =&gt; Residue AKT1-1:S473:phosphorylation [Evidences: HPRD, KEA, MIMP, PhosphoSite, ProtMapper, REACH, SIGNOR, Sparser, dbPTM, phosphoELM (15 references)]&gt;,
 &lt;MTOR =&gt; Residue AKT1-1:T450:phosphorylation [Evidences: HPRD, MIMP, PhosphoSite, ProtMapper, phosphoELM (0 references)]&gt;,
 &lt;MTOR =&gt; Residue AKT1-1:T308:phosphorylation [Evidences: ProtMapper, Sparser (1 references)]&gt;]
</pre></div></div>
</div>
<section id="Enzyme-substrate-objects">
<h3>Enzyme-substrate objects<a class="headerlink" href="#Enzyme-substrate-objects" title="Permalink to this heading">§</a></h3>
<p>Let’s take a closer look at one of the enzyme-PTM relationships, represented by <code class="docutils literal notranslate"><span class="pre">pypath.internals.intera.DomainMotif</span></code> objects. Below some of the attributes are shown:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>e_ptm = es[(&#39;MTOR&#39;, &#39;AKT1&#39;)][0]
e_ptm.ptm.protein, e_ptm.ptm.protein.identifier, e_ptm.ptm.isoform, e_ptm.ptm.residue, e_ptm.ptm.residue.name, e_ptm.ptm.residue.number, e_ptm.ptm.typ, e_ptm.domain.protein
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&lt;Entity: AKT1&gt;,
 &#39;P31749&#39;,
 1,
 &lt;Residue AKT1-1:S473&gt;,
 &#39;S&#39;,
 473,
 &#39;phosphorylation&#39;,
 &lt;Entity: MTOR&gt;)
</pre></div></div>
</div>
<p>The resources and references are available in <code class="docutils literal notranslate"><span class="pre">Evidences</span></code> objects:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>e_ptm.evidences
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Evidences: HPRD, KEA, MIMP, PhosphoSite, ProtMapper, REACH, SIGNOR, Sparser, dbPTM, phosphoELM (15 references)&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>e_ptm.evidences.get_resource_names()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;KEA&#39;, &#39;MIMP&#39;, &#39;PhosphoSite&#39;, &#39;ProtMapper&#39;, &#39;SIGNOR&#39;, &#39;dbPTM&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>e_ptm.evidences.get_references()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&lt;Reference: 14761976&gt;,
 &lt;Reference: 15047712&gt;,
 &lt;Reference: 15364915&gt;,
 &lt;Reference: 15718470&gt;,
 &lt;Reference: 15899889&gt;,
 &lt;Reference: 16221682&gt;,
 &lt;Reference: 17013611&gt;,
 &lt;Reference: 19844585&gt;,
 &lt;Reference: 20333297&gt;,
 &lt;Reference: 20489726&gt;,
 &lt;Reference: 21157483&gt;,
 &lt;Reference: 21592956&gt;,
 &lt;Reference: 23006971&gt;,
 &lt;Reference: 8978681&gt;,
 &lt;Reference: 9736715&gt;}
</pre></div></div>
</div>
</section>
<section id="Enzyme-substrate-data-frame">
<h3>Enzyme-substrate data frame<a class="headerlink" href="#Enzyme-substrate-data-frame" title="Permalink to this heading">§</a></h3>
<p>The dabase object is able to export its contents into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[116]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>es.make_df()
es.df
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[116]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enzyme</th>
      <th>enzyme_genesymbol</th>
      <th>substrate</th>
      <th>substrate_genesymbol</th>
      <th>isoforms</th>
      <th>residue_type</th>
      <th>residue_offset</th>
      <th>modification</th>
      <th>sources</th>
      <th>references</th>
      <th>curation_effort</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>P31749</td>
      <td>AKT1</td>
      <td>P63104</td>
      <td>YWHAZ</td>
      <td>1</td>
      <td>S</td>
      <td>58</td>
      <td>phosphorylation</td>
      <td>HPRD;HPRD_MIMP;KEA;MIMP;PhosphoSite;PhosphoSit...</td>
      <td>HPRD:11956222;KEA:11956222;KEA:12861023;KEA:16...</td></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 4163 characters</em></div>
</div>
</section>
</section>
<section id="Protein-sequences">
<h2>Protein sequences<a class="headerlink" href="#Protein-sequences" title="Permalink to this heading">§</a></h2>
<p>The APIs for sequences are very basic, because we’ve never really needed them; but the fundamentals are probably there to make a nice, powerful API. Still, I don’t believe <em>pypath</em> will ever be strong in sequences, it’s just not our main topic.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[186]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import homology
seqc = homology.SequenceContainer(preload_seq = [9606])
akt1 = seqc.get_seq(&#39;P31749&#39;)
akt1.get_region(start = 10, end = 19, isoform = 2)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[186]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(10, 19, &#39;TFIIRCLQWT&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[187]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import seq
human_proteome = seq.swissprot_seq()
human_proteome
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[187]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;P63120&#39;: &lt;pypath.utils.seq.Seq at 0x689900d45cc0&gt;,
 &#39;Q96EC8&#39;: &lt;pypath.utils.seq.Seq at 0x689908ea8f70&gt;,
 &#39;Q6ZMS4&#39;: &lt;pypath.utils.seq.Seq at 0x689908eaa4a0&gt;,
 &#39;Q8N8L2&#39;: &lt;pypath.utils.seq.Seq at 0x6899223538b0&gt;,
 &#39;Q15916&#39;: &lt;pypath.utils.seq.Seq at 0x689922353c70&gt;,
 &#39;O60384&#39;: &lt;pypath.utils.seq.Seq at 0x689922350730&gt;,
 &#39;Q3MIS6&#39;: &lt;pypath.utils.seq.Seq at 0x689922353310&gt;,
 &#39;Q86UK7&#39;: &lt;pypath.utils.seq.Seq at 0x689922353760&gt;,
 &#39;Q6P280&#39;: &lt;pypath.utils.seq.Seq at 0x689922353190&gt;,
 &#39;Q969W1&#39;: &lt;pypath.utils.seq.Seq at 0x689922350d90&gt;,
 &#39;O14978&#39;: &lt;pypath.utils.seq.Seq at 0x689922353220&gt;,
 &#39;P61129&#39;: &lt;pypath.utils.seq.Seq at 0x689922353370&gt;,
 &#39;Q66K41&#39;: &lt;pypath.utils.seq.Seq at 0x6899223534f0&gt;,
 &#39;Q15937&#39;: &lt;pypath.utils.seq.Seq at 0x689922350c70&gt;,
 &#39;Q9P2J8&#39;: &lt;pypath.utils.seq.Seq at 0x689922351450&gt;,
 &#39;Q8ND82&#39;: &lt;pypath.utils.seq.Seq at 0x689922353910&gt;,
 &#39;Q9NP64&#39;: &lt;pypath.utils.seq.Seq at 0x6899223502b0&gt;,
 &#39;P98182&#39;: &lt;pypath.utils.seq.Seq at 0x689922350280&gt;,
 &#39;Q8IUH4&#39;: &lt;pypath.utils.seq.Seq at 0x68992235
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 53045 characters</em></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[191]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>list(human_proteome[&#39;P00533&#39;].findall(&#39;YGCT&#39;))
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[191]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[SeqLookup(isoform=1, offset=625)]
</pre></div></div>
</div>
</section>
<section id="id1">
<h2>Annotations<a class="headerlink" href="#id1" title="Permalink to this heading">§</a></h2>
<p>This database provides various annotations about the function, structure, localization and many other properties of the proteins and genes. The database is an instance of <code class="docutils literal notranslate"><span class="pre">pypath.core.annot.AnnotationTable</span></code> class. The database is built with the default or current configuration by the <code class="docutils literal notranslate"><span class="pre">core.annot.get_db</span></code> method.</p>
<div class="alert alert-block alert-warning"><p>Warning: it is recommended to access databases by the manager. Running the code below takes really long and does not save or reload the database, it builds a fresh copy each time.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import annot
an = annot.get_db()
an
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Annotation database: 3788067 records about 51636 entities from 78 resources&gt;
</pre></div></div>
</div>
<section id="Load-a-single-annotation-resource">
<h3>Load a single annotation resource<a class="headerlink" href="#Load-a-single-annotation-resource" title="Permalink to this heading">§</a></h3>
<p>The annotations database is huge, on disk it takes up 1-2 GB of space, it consists of 60-70 resources. But all these resources are not integrated with each other, each can be loaded individually, by their dedicated classes in the <code class="docutils literal notranslate"><span class="pre">core.annot</span></code> module. This practice can be recommended and will be supported better in the future. Let’s load one resource:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import annot
cpad = annot.Cpad()
cpad
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;CPAD annotations: 2309 records about 1357 entities&gt;
</pre></div></div>
</div>
<p>The resulted object is derived from the <code class="docutils literal notranslate"><span class="pre">AnnotationBase</span></code> class, its data is stored under the <code class="docutils literal notranslate"><span class="pre">annot</span></code> attribute, in a dict where identifiers are keys and sets of annotation records are the values. The keys of the records are shown by the <code class="docutils literal notranslate"><span class="pre">get_names</span></code> method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cpad.get_names()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;regulator_type&#39;,
 &#39;effect_on_pathway&#39;,
 &#39;pathway&#39;,
 &#39;effect_on_cancer&#39;,
 &#39;effect_on_cancer_outcome&#39;,
 &#39;cancer&#39;,
 &#39;pathway_category&#39;)
</pre></div></div>
</div>
<p>For each name we can list the possible values:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cpad.get_values(&#39;cancer&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Acute lymphoblastic leukemia (ALL) (precursor T lymphoblastic leukemia)&#39;,
 &#39;Acute myeloid leukemia (AML)&#39;,
 &#39;Basal cell carcinoma&#39;,
 &#39;Bladder cancer&#39;,
 &#39;Breast cancer&#39;,
 &#39;Cervical cancer&#39;,
 &#39;Cholangiocarcinoma&#39;,
 &#39;Choriocarcinoma&#39;,
 &#39;Chronic lymphocytic leukemia (CLL)&#39;,
 &#39;Chronic myeloid leukemia (CML)&#39;,
 &#39;Colorectal cancer&#39;,
 &#39;Endometrial cancer&#39;,
 &#39;Esophageal cancer&#39;,
 &#34;Ewing&#39;s sarcoma&#34;,
 &#39;Gallbladder cancer&#39;,
 &#39;Gastric cancer&#39;,
 &#39;Glioma&#39;,
 &#39;Hepatocellular carcinoma&#39;,
 &#39;Hodgkin lymphoma&#39;,
 &#39;Infantile hemangioma&#39;,
 &#39;Laryngeal cancer&#39;,
 &#39;Malignant melanoma&#39;,
 &#39;Malignant pleural mesothelioma&#39;,
 &#39;Mantle cell lymphoma&#39;,
 &#39;Multiple myeloma&#39;,
 &#39;Nasopharyngeal cancer&#39;,
 &#39;Neuroblastoma&#39;,
 &#39;Non-small cell lung cancer&#39;,
 &#39;Oral cancer&#39;,
 &#39;Osteosarcoma&#39;,
 &#39;Ovarian cancer&#39;,
 &#39;Pancreatic cancer&#39;,
 &#39;Pituitary adenomas&#39;,
 &#39;Prostate cancer&#39;,
 &#39;Renal cell carcinoma&#39;,
 &#39;Small cell lung cancer&#39;,
 &#39;Squamous cell carcinoma&#39;,
 &#39;Synovial sarcoma&#39;,
 &#39;Testicular cancer&#39;,
 &#39;Thyroid cancer&#39;}
</pre></div></div>
</div>
<p>Based on their annotations the <code class="docutils literal notranslate"><span class="pre">select</span></code> method filters the annotated molecules. For example, 78 complexes, miRNAs and proteins are annotated as inhibiting colorectal cancer:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cpad.select(cancer = &#39;Colorectal cancer&#39;, effect_on_cancer = &#39;Inhibiting&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;A6NDV4&#39;,
 Complex: COMPLEX:O14745,
 Complex: COMPLEX:O14862,
 Complex: COMPLEX:O15169_P25054,
 Complex: COMPLEX:O94813,
 Complex: COMPLEX:O94953,
 Complex: COMPLEX:P00533,
 Complex: COMPLEX:P06733,
 Complex Glucose transporter complex 1: COMPLEX:P11166,
 Complex: COMPLEX:P25054,
 Complex: COMPLEX:P40261,
 Complex: COMPLEX:P49327,
 Complex: COMPLEX:P54687,
 Complex PTEN phosphatase complex: COMPLEX:P60484,
 Complex: COMPLEX:Q01973,
 Complex: COMPLEX:Q12888,
 Complex: COMPLEX:Q13620,
 Complex: COMPLEX:Q96CX2,
 Complex: COMPLEX:Q99558,
 &#39;MIMAT0000069&#39;,
 &#39;MIMAT0000089&#39;,
 &#39;MIMAT0000093&#39;,
 &#39;MIMAT0000262&#39;,
 &#39;MIMAT0000274&#39;,
 &#39;MIMAT0000422&#39;,
 &#39;MIMAT0000427&#39;,
 &#39;MIMAT0000437&#39;,
 &#39;MIMAT0000449&#39;,
 &#39;MIMAT0000455&#39;,
 &#39;MIMAT0000460&#39;,
 &#39;MIMAT0000461&#39;,
 &#39;MIMAT0000617&#39;,
 &#39;MIMAT0003266&#39;,
 &#39;MIMAT0003320&#39;,
 &#39;O14745&#39;,
 &#39;O14862&#39;,
 &#39;O15169&#39;,
 &#39;O75473&#39;,
 &#39;O75888&#39;,
 &#39;O76041&#39;,
 &#39;O94813&#39;,
 &#39;O94953&#39;,
 &#39;P00533&#39;,
 &#39;P06733&#39;,
 &#39;P06756&#39;,
 &#39;P11166&#39;,
 &#39;P13631&#39;,
 &#39;P22676&#39;,
 &#39;P25054&#39;,
 &#39;P25791&#39;,
 &#39;P40261&#39;,
 &#39;P49327&#39;,
 &#39;P546
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1279 characters</em></div>
</div>
</section>
<section id="Load-the-full-annotations-database-by-the-database-manager">
<h3>Load the full annotations database by the database manager<a class="headerlink" href="#Load-the-full-annotations-database-by-the-database-manager" title="Permalink to this heading">§</a></h3>
<p>Alternatively, the full annotations database can be accessed in the usual way:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[215]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
an = omnipath.db.get_db(&#39;annotations&#39;)
an
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[215]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Annotation database: 5490653 records about 50872 entities from 68 resources&gt;
</pre></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">AnnotationTable</span></code> object contains the resource specific annotation objects under the <code class="docutils literal notranslate"><span class="pre">annots</span></code> attribute:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>an.annots
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[40]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;CellTypist&#39;: &lt;CellTypist annotations: 927 records about 473 entities&gt;,
 &#39;Integrins&#39;: &lt;Integrins annotations: 62 records about 62 entities&gt;,
 &#39;CellCellInteractions&#39;: &lt;CellCellInteractions annotations: 5544 records about 4960 entities&gt;,
 &#39;PanglaoDB&#39;: &lt;PanglaoDB annotations: 8479 records about 4813 entities&gt;,
 &#39;Lambert2018&#39;: &lt;Lambert2018 annotations: 3281 records about 3277 entities&gt;,
 &#39;CancerSEA&#39;: &lt;CancerSEA annotations: 2515 records about 1992 entities&gt;,
 &#39;Phobius&#39;: &lt;Phobius annotations: 35382 records about 35382 entities&gt;,
 &#39;GO_Intercell&#39;: &lt;GO_Intercell annotations: 48799 records about 18377 entities&gt;,
 &#39;MatrixDB&#39;: &lt;MatrixDB annotations: 18127 records about 15903 entities&gt;,
 &#39;Surfaceome&#39;: &lt;Surfaceome annotations: 3558 records about 3558 entities&gt;,
 &#39;Matrisome&#39;: &lt;Matrisome annotations: 1514 records about 1514 entities&gt;,
 &#39;HPA_secretome&#39;: &lt;HPA_secretome annotations: 3568 records about 3568 entities&gt;,
 &#39;HPMR&#39;: &lt;HPMR annotations: 1748 records about 1695 entities&gt;,
 &#39;CPAD&#39;: &lt;CPAD annotati
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 5842 characters</em></div>
</div>
<p>For each of these you can query the names of the fields, their possible values and the set of proteins annotated with any combination of the values, just like for CPAD above. As another exemple, let’s take a look into the Matrisome database:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>matrisome = an.annots[&#39;Matrisome&#39;]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>matrisome.get_names()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(&#39;mainclass&#39;, &#39;subclass&#39;, &#39;subsubclass&#39;)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>matrisome.get_values(&#39;subclass&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Collagens&#39;,
 &#39;ECM Glycoproteins&#39;,
 &#39;ECM Regulators&#39;,
 &#39;ECM-affiliated Proteins&#39;,
 &#39;Proteoglycans&#39;,
 &#39;Secreted Factors&#39;,
 &#39;n/a&#39;}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>matrisome.get_subset(subclass = &#39;Collagens&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;A6NMZ7&#39;,
 &#39;A8TX70&#39;,
 &#39;B4DZ39&#39;,
 Complex Collagen type I homotrimer: COMPLEX:P02452,
 Complex HT_DM_Cluster278: COMPLEX:P02452_P02462_P08572_P29400_P53420_Q01955_Q02388_Q14031_Q17RW2_Q8NFW1,
 Complex Collagen type I trimer: COMPLEX:P02452_P08123,
 Complex Collagen type II trimer: COMPLEX:P02458,
 Complex Collagen type XI trimer variant 1: COMPLEX:P02458_P12107_P13942,
 Complex: COMPLEX:P02458_P20908_P25067,
 Complex: COMPLEX:P02458_P20908_P25067_P29400,
 Complex: COMPLEX:P02458_P25067_P29400,
 Complex Collagen type III trimer: COMPLEX:P02461,
 Complex: COMPLEX:P02462,
 Complex Collagen type IV trimer variant 1: COMPLEX:P02462_P08572,
 Complex Collagen type XI trimer variant 2: COMPLEX:P05997_P12107,
 Complex Collagen type XI trimer variant 3: COMPLEX:P05997_P12107_P20908,
 Complex Collagen type V trimer variant 1: COMPLEX:P05997_P20908,
 Complex Collagen type V trimer variant 2: COMPLEX:P05997_P20908_P25940,
 Complex: COMPLEX:P08572,
 Complex: COMPLEX:P12109_P12110,
 Complex Collagen
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 3072 characters</em></div>
</div>
</section>
<section id="Load-only-selected-annotations">
<h3>Load only selected annotations<a class="headerlink" href="#Load-only-selected-annotations" title="Permalink to this heading">§</a></h3>
<p>Another option is to load only certain annotation resources into an <code class="docutils literal notranslate"><span class="pre">AnnotationTable</span></code> object. We refer to the resources by class names. For example, if you only want to load the pathway membership annotations from SIGNOR, SignaLink, NetPath and KEGG, you can provide the names of the appropriate classes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pathways = annot.AnnotationTable(
    protein_sources = (
        &#39;SignalinkPathways&#39;,
        &#39;KeggPathways&#39;,
        &#39;NetpathPathways&#39;,
        &#39;SignorPathways&#39;,
    ),
    complex_sources = (),
)
pathways
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[47]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Annotation database: 28745 records about 6762 entities from 4 resources&gt;
</pre></div></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">AnnotationTable</span></code> object provides methods to query all resources together, or build a boolean array out of them. To see all annotations of one protein:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pathways.all_annotations(&#39;P00533&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[48]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[SignalinkPathway(pathway=&#39;Receptor tyrosine kinase&#39;),
 SignalinkPathway(pathway=&#39;JAK/STAT&#39;),
 KeggPathway(pathway=&#39;Proteoglycans in cancer&#39;),
 KeggPathway(pathway=&#39;Regulation of actin cytoskeleton&#39;),
 KeggPathway(pathway=&#39;Oxytocin signaling pathway&#39;),
 KeggPathway(pathway=&#39;Phospholipase D signaling pathway&#39;),
 KeggPathway(pathway=&#39;Pathways in cancer&#39;),
 KeggPathway(pathway=&#39;Hepatocellular carcinoma&#39;),
 KeggPathway(pathway=&#39;Colorectal cancer&#39;),
 KeggPathway(pathway=&#39;Melanoma&#39;),
 KeggPathway(pathway=&#39;EGFR tyrosine kinase inhibitor resistance&#39;),
 KeggPathway(pathway=&#39;Human papillomavirus infection&#39;),
 KeggPathway(pathway=&#39;Pancreatic cancer&#39;),
 KeggPathway(pathway=&#39;Non-small cell lung cancer&#39;),
 KeggPathway(pathway=&#39;Central carbon metabolism in cancer&#39;),
 KeggPathway(pathway=&#39;Endocytosis&#39;),
 KeggPathway(pathway=&#39;Endometrial cancer&#39;),
 KeggPathway(pathway=&#39;Choline metabolism in cancer&#39;),
 KeggPathway(pathway=&#39;Bladder cancer&#39;),
 KeggPathway(pathway=&#39;Parathyroid hormone synthesis, secretion
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 2540 characters</em></div>
</div>
</section>
<section id="Data-frames-of-annotations">
<h3>Data frames of annotations<a class="headerlink" href="#Data-frames-of-annotations" title="Permalink to this heading">§</a></h3>
<p>Data from annotation objects can be exported to a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cpad.make_df()
cpad.df
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[49]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>uniprot</th>
      <th>genesymbol</th>
      <th>entity_type</th>
      <th>source</th>
      <th>label</th>
      <th>value</th>
      <th>record_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Q16181</td>
      <td>SEPT7</td>
      <td>protein</td>
      <td>CPAD</td>
      <td>regulator_type</td>
      <td>protein</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Q16181</td>
      <td>SEPT7</td>
      <td>protein</td>
      <td>CPAD</td>
      <td>effect_on_pathway</td>
      <td>Upregulation</td>
      <td>0</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Q16181</td>
      <td>SEPT7</td>
      <td</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 2867 characters</em></div>
</div>
<p>The data frame has a long format. It can be converted to the more conventional wide format using standard <code class="docutils literal notranslate"><span class="pre">pandas</span></code> procedures (well, in tidyverse you would simply call <code class="docutils literal notranslate"><span class="pre">tidyr::pivot_wider</span></code>, in <code class="docutils literal notranslate"><span class="pre">pandas</span></code> you have to do an unintuitive sequence of 6 calls):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>index_cols = [&#39;record_id&#39;, &#39;uniprot&#39;, &#39;genesymbol&#39;, &#39;label&#39;, &#39;entity_type&#39;]

(
    cpad.df.drop(&#39;source&#39;, axis=1).
    set_index(index_cols).
    unstack(&#39;label&#39;).
    droplevel(axis=1, level=0).
    reset_index().
    drop(&#39;record_id&#39;, axis=1)
)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th>label</th>
      <th>uniprot</th>
      <th>genesymbol</th>
      <th>entity_type</th>
      <th>cancer</th>
      <th>effect_on_cancer</th>
      <th>effect_on_cancer_outcome</th>
      <th>effect_on_pathway</th>
      <th>pathway</th>
      <th>pathway_category</th>
      <th>regulator_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Q16181</td>
      <td>SEPT7</td>
      <td>protein</td>
      <td>Glioma</td>
      <td>Inhibiting</td>
      <td>inhibit glioma cell migration</td>
      <td>Upregulation</td>
      <td>Actin cytoskeleton pathway</td>
      <td>Regulation of actin cytoskeleton</td>
      <td>protein</td>
    </tr>
    <t</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 4209 characters</em></div>
</div>
</section>
</section>
<section id="Inter-cellular-signaling-roles">
<h2>Inter-cellular signaling roles<a class="headerlink" href="#Inter-cellular-signaling-roles" title="Permalink to this heading">§</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pypath</span></code> does not combine the annotations in the <code class="docutils literal notranslate"><span class="pre">annot</span></code> module, exactly what goes in goes out. For example, WNT pathway from Signor and SignaLink won’t be merged automatically. However with the <code class="docutils literal notranslate"><span class="pre">pypath.core.annot.CustomAnnotation</span></code> class anyone can do it. For inter-cellular communication categories the <code class="docutils literal notranslate"><span class="pre">pypath.core.intercell</span></code> module combines the data from all the relevant resources and creates categories based on a combination of evidences. The database is an instance of the
<code class="docutils literal notranslate"><span class="pre">IntercellAnnotation</span></code> object, and the build is executed by the <code class="docutils literal notranslate"><span class="pre">pypath.core.intercell.get_db</span></code> function.</p>
<div class="alert alert-block alert-warning"><p>Warning: it is recommended to access databases by the manager. Running the code below takes really long and does not save or reload the database, it builds a fresh copy each time.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import intercell
ic = intercell.get_db() # this takes quite some time
                       # unless you load annotations from a pickle cache
ic
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[53]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Intercell annotations: 310033 records about 43617 entities&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath import omnipath
ic = omnipath.db.get_db(&#39;intercell&#39;)
ic
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Intercell annotations: 301527 records about 48570 entities&gt;
</pre></div></div>
</div>
<p>This object stores its data under the <code class="docutils literal notranslate"><span class="pre">classes</span></code> attribute. Classes are defined in <code class="docutils literal notranslate"><span class="pre">pypath.core.intercell_annot.annot_combined_classes</span></code>. In addition, we manually revised and excluded some proteins from the more generic classes, these are listed in <code class="docutils literal notranslate"><span class="pre">pypath.core.intercell_annot.excludes</span></code>. Each class has the following properties:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">name</span></code>: all lowercase, human understandable name, without repeating the parent class (e.g. <em>WNT receptor</em>s will be simply <em>wnt</em>, and the parent class will be <em>receptor</em>)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">parent</span></code>: for a <em>specific</em> class the parent is the <em>generic</em> category it belongs to; for generic classes the <code class="docutils literal notranslate"><span class="pre">name</span></code> and <code class="docutils literal notranslate"><span class="pre">parent</span></code> are the same</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">resource</span></code>: the resource the data comes from, or <em>OmniPath</em> for composite classes (combined from multiple resources)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">scope</span></code>: <em>specific</em> or <em>generic</em>; e.g. <em>TGF ligand</em> is specific, <em>ligand</em> is generic</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">aspect</span></code>: <em>locational</em> (e.g. <em>plasma membrane</em>) or <em>functional</em> (e.g. <em>transporter</em>)</p></li>
</ul>
<p>Read more about the design of the intercell database <a class="reference external" href="https://www.embopress.org/doi/full/10.15252/msb.20209923">in our paper</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ic.classes
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[55]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{AnnotDefKey(name=&#39;transmembrane&#39;, parent=&#39;transmembrane&#39;, resource=&#39;UniProt_location&#39;): &lt;AnnotationGroup `transmembrane` from UniProt_location, 5150 elements&gt;,
 AnnotDefKey(name=&#39;transmembrane&#39;, parent=&#39;transmembrane&#39;, resource=&#39;UniProt_topology&#39;): &lt;AnnotationGroup `transmembrane` from UniProt_topology, 5760 elements&gt;,
 AnnotDefKey(name=&#39;transmembrane&#39;, parent=&#39;transmembrane&#39;, resource=&#39;UniProt_keyword&#39;): &lt;AnnotationGroup `transmembrane` from UniProt_keyword, 7041 elements&gt;,
 AnnotDefKey(name=&#39;transmembrane&#39;, parent=&#39;transmembrane_predicted&#39;, resource=&#39;Phobius&#39;): &lt;AnnotationGroup `transmembrane` from Phobius, 6444 elements&gt;,
 AnnotDefKey(name=&#39;transmembrane_phobius&#39;, parent=&#39;transmembrane_predicted&#39;, resource=&#39;Almen2009&#39;): &lt;AnnotationGroup `transmembrane_phobius` from Almen2009, 2072 elements&gt;,
 AnnotDefKey(name=&#39;transmembrane_sosui&#39;, parent=&#39;transmembrane_predicted&#39;, resource=&#39;Almen2009&#39;): &lt;AnnotationGroup `transmembrane_sosui` from Almen2009, 1663 elements&gt;,
 AnnotDefKey(name=&#39;trans
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 143945 characters</em></div>
</div>
<p>An easy way to access the classes is the <code class="docutils literal notranslate"><span class="pre">select</span></code> method. The <code class="docutils literal notranslate"><span class="pre">AnnotationGroup</span></code> objects behave as plain Python <code class="docutils literal notranslate"><span class="pre">set</span></code>s, and besides that, they feature many further attributes and methods.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gaba_receptors = ic.select(&#39;gaba&#39;, parent = &#39;receptor&#39;)
gaba_receptors
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[56]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AnnotationGroup `gaba` from HGNC, 40 elements&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>gaba_receptors.members
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;A8MPY1&#39;,
 Complex GABA-A receptor (GABRA1, GABRB2, GABRD): COMPLEX:O14764_P14867_P47870,
 Complex GABA-A receptor, alpha-4/beta-3/delta: COMPLEX:O14764_P28472_P48169,
 Complex GABA-A receptor, alpha-6/beta-3/delta: COMPLEX:O14764_P28472_Q16445,
 Complex GABA-A receptor, alpha-4/beta-2/delta: COMPLEX:O14764_P47870_P48169,
 Complex GABA-A receptor, alpha-6/beta-2/delta: COMPLEX:O14764_P47870_Q16445,
 Complex GABBR1-GABBR2 complex: COMPLEX:O75899_Q9UBS5,
 Complex: COMPLEX:P14867,
 Complex GABA-A receptor, alpha-1/beta-3/gamma-2: COMPLEX:P14867_P18507_P28472,
 Complex GABA-A receptor (GABRA1, GABRB2, GABRG2): COMPLEX:P14867_P18507_P47870,
 Complex GABA-A receptor, alpha-5/beta-3/gamma-2: COMPLEX:P18507_P28472_P31644,
 Complex GABA-A receptor, alpha-3/beta-3/gamma-2: COMPLEX:P18507_P28472_P34903,
 Complex GABA-A receptor, alpha-2/beta-3/gamma-2: COMPLEX:P18507_P28472_P47869,
 Complex GABA-A receptor, alpha-6/beta-3/gamma-2: COMPLEX:P18507_P28472_Q16445,
 Complex: COMPLEX:P18507_Q8N1C3,
 C
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1368 characters</em></div>
</div>
<section id="Build-an-intercellular-communication-network">
<h3>Build an intercellular communication network<a class="headerlink" href="#Build-an-intercellular-communication-network" title="Permalink to this heading">§</a></h3>
<p>The intercell database can be connected to a <code class="docutils literal notranslate"><span class="pre">Network</span></code> object to create an intercellular communication network:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cu = omnipath.db.get_db(&#39;curated&#39;)
ic.register_network(cu)
</pre></div>
</div>
</div>
</section>
<section id="Quantitative-overview-of-intercell-annotations">
<h3>Quantitative overview of intercell annotations<a class="headerlink" href="#Quantitative-overview-of-intercell-annotations" title="Permalink to this heading">§</a></h3>
<p>A data frame with basic statistics is available:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>i.counts_df()
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[59]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>category</th>
      <th>parent</th>
      <th>database</th>
      <th>scope</th>
      <th>aspect</th>
      <th>source</th>
      <th>consensus_score</th>
      <th>transmitter</th>
      <th>receiver</th>
      <th>secreted</th>
      <th>plasma_membrane_transmembrane</th>
      <th>plasma_membrane_peripheral</th>
      <th>n_uniprot</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>transmembrane</td>
      <td>transmembrane</td>
      <td>UniProt_location</td>
      <td>generic</td>
      <td>locational</td>
      <td>resource_specific</td>
      <td>4</td>
      <td>False</td>
      <td>False</td>
      <td>False</td></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 4706 characters</em></div>
</div>
</section>
<section id="Intercell-database-as-data-frame">
<h3>Intercell database as data frame<a class="headerlink" href="#Intercell-database-as-data-frame" title="Permalink to this heading">§</a></h3>
<p>Just like the other databases, the object can be exported into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>i.make_df()
i.df[:10]
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[60]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>category</th>
      <th>parent</th>
      <th>database</th>
      <th>scope</th>
      <th>aspect</th>
      <th>source</th>
      <th>uniprot</th>
      <th>genesymbol</th>
      <th>entity_type</th>
      <th>consensus_score</th>
      <th>transmitter</th>
      <th>receiver</th>
      <th>secreted</th>
      <th>plasma_membrane_transmembrane</th>
      <th>plasma_membrane_peripheral</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>transmembrane</td>
      <td>transmembrane</td>
      <td>UniProt_location</td>
      <td>generic</td>
      <td>locational</td>
      <td>resource_specific</td>
      <td>A0A1B0GWG4</td>
      <</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 4762 characters</em></div>
</div>
</section>
<section id="Browse-intercell-categories">
<h3>Browse intercell categories<a class="headerlink" href="#Browse-intercell-categories" title="Permalink to this heading">§</a></h3>
<p>Use the <code class="docutils literal notranslate"><span class="pre">select</span></code> method to access intercell classes:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ic.select(definition = &#39;neurotensin&#39;, parent = &#39;receptor&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[72]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;AnnotationGroup `neurotensin` from HGNC, 2 elements&gt;
</pre></div></div>
</div>
<p>Proteins in each category can be listed with their descriptions from UniProt. Loading the UniProt datasheets for each protein is a slow process, we don’t recomment calling this method on more than a few dozens of proteins.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[79]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ic.show(&#39;neurotensin&#39;, parent = &#39;receptor&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
=====&gt; [2 proteins] &lt;=====
╒═══════╤════════╤══════════════╤══════════╤══════════╤═════════════╤══════════════╤════════════╤══════════════╕
│   No. │ ac     │ genesymbol   │   length │   weight │ full_name   │ function_o   │ keywords   │ subcellula   │
│       │        │              │          │          │             │ r_genecard   │            │ r_location   │
│       │        │              │          │          │             │ s            │            │              │
╞═══════╪════════╪══════════════╪══════════╪══════════╪═════════════╪══════════════╪════════════╪══════════════╡
│     1 │ O95665 │ NTSR2        │      410 │    45385 │ Neurotensi  │ Receptor     │ Cell       │ Cell         │
│       │        │              │          │          │ n receptor  │ for the tr   │ membrane,  │ membrane;    │
│       │        │              │          │          │ type 2      │ idecapepti   │ Disulfide  │ Multi-pass   │
│       │        │              │          │          │             │
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 7598 characters</em></div>
</div>
</section>
</section>
<section id="Gene-Ontology">
<h2>Gene Ontology<a class="headerlink" href="#Gene-Ontology" title="Permalink to this heading">§</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">pypath.utils.go</span></code> is an almost standalone module for management of the Gene Ontology tree and annotations. The main objects here are <code class="docutils literal notranslate"><span class="pre">GeneOntology</span></code> and <code class="docutils literal notranslate"><span class="pre">GOAnnotation</span></code>. The former represents the ontology tree, i.e. terms and their relationships, the latter their assignment to gene products. Both provides many versatile methods for querying.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[80]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.utils import go
goa = go.GOAnnotation()
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[81]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>goa.ontology # the GeneOntology object
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[81]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;pypath.utils.go.GeneOntology at 0x689946b55570&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[82]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>goa # the GOAnnotation object
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[82]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;pypath.utils.go.GOAnnotation at 0x68991cdc9b40&gt;
</pre></div></div>
</div>
<p>Among many others, the most versatile method is <code class="docutils literal notranslate"><span class="pre">select</span></code> which is able to select the annotated gene products by various expressions built from GO terms or IDs. It understands <code class="docutils literal notranslate"><span class="pre">AND</span></code>, <code class="docutils literal notranslate"><span class="pre">OR</span></code>, <code class="docutils literal notranslate"><span class="pre">NOT</span></code> and parentheses.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[83]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>query = &quot;&quot;&quot;(cell surface OR
        external side of plasma membrane OR
        extracellular region) AND
        (regulation of transmembrane transporter activity OR
        channel regulator activity)&quot;&quot;&quot;
result = goa.select(query)
print(list(result)[:7])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;P21333&#39;, &#39;P80108&#39;, &#39;P62258&#39;, &#39;Q9NRX4&#39;, &#39;P54710&#39;, &#39;Q8NER1&#39;, &#39;P01303&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[84]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>goa.ontology.get_all_descendants(&#39;GO:0005576&#39;)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[84]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;GO:0001507&#39;,
 &#39;GO:0001527&#39;,
 &#39;GO:0003351&#39;,
 &#39;GO:0003355&#39;,
 &#39;GO:0005201&#39;,
 &#39;GO:0005576&#39;,
 &#39;GO:0005577&#39;,
 &#39;GO:0005582&#39;,
 &#39;GO:0005583&#39;,
 &#39;GO:0005584&#39;,
 &#39;GO:0005585&#39;,
 &#39;GO:0005586&#39;,
 &#39;GO:0005587&#39;,
 &#39;GO:0005588&#39;,
 &#39;GO:0005590&#39;,
 &#39;GO:0005591&#39;,
 &#39;GO:0005592&#39;,
 &#39;GO:0005595&#39;,
 &#39;GO:0005596&#39;,
 &#39;GO:0005599&#39;,
 &#39;GO:0005601&#39;,
 &#39;GO:0005602&#39;,
 &#39;GO:0005604&#39;,
 &#39;GO:0005606&#39;,
 &#39;GO:0005607&#39;,
 &#39;GO:0005608&#39;,
 &#39;GO:0005609&#39;,
 &#39;GO:0005610&#39;,
 &#39;GO:0005611&#39;,
 &#39;GO:0005612&#39;,
 &#39;GO:0005614&#39;,
 &#39;GO:0005615&#39;,
 &#39;GO:0005616&#39;,
 &#39;GO:0006858&#39;,
 &#39;GO:0006859&#39;,
 &#39;GO:0006860&#39;,
 &#39;GO:0009519&#39;,
 &#39;GO:0010367&#39;,
 &#39;GO:0016914&#39;,
 &#39;GO:0016942&#39;,
 &#39;GO:0020003&#39;,
 &#39;GO:0020004&#39;,
 &#39;GO:0020005&#39;,
 &#39;GO:0020006&#39;,
 &#39;GO:0030020&#39;,
 &#39;GO:0030021&#39;,
 &#39;GO:0030023&#39;,
 &#39;GO:0030197&#39;,
 &#39;GO:0030345&#39;,
 &#39;GO:0030934&#39;,
 &#39;GO:0030935&#39;,
 &#39;GO:0030938&#39;,
 &#39;GO:0031012&#39;,
 &#39;GO:0031395&#39;,
 &#39;GO:0032311&#39;,
 &#39;GO:0032579&#39;,
 &#39;GO:0033165&#39;,
 &#39;GO:0033166&#39;,
 &#39;GO:0034358&#39;,
 &#39;GO:0034359&#39;,
 &#39;GO:0034360&#39;,
 &#39;GO:0034361&#39;,
 &#39;GO:0034362&#39;,
 &#39;GO:0034363&#39;,
 &#39;GO:0034364&#39;,
 &#39;GO:0034365&#39;,
 &#39;GO:00343
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 3104 characters</em></div>
</div>
</section>
<section id="id2">
<h2>Protein complexes<a class="headerlink" href="#id2" title="Permalink to this heading">§</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">pypath.complex</span></code> module builds a non-redundant list of complexes from about 12 original resources. Complexes are unique considering their set of components, and optionally carry stoichiometry information. Homomultimers are also included, hence some complexes consist only of a single kind of protein. The database is an instance of <code class="docutils literal notranslate"><span class="pre">pypath.core.complex.ComplexAggregator</span></code> object and the built by the <code class="docutils literal notranslate"><span class="pre">pypath.core.complex.get_db</span></code> function.</p>
<div class="alert alert-block alert-warning"><p>Warning: it is recommended to access databases by the manager. Running the code below takes really long and does not save or reload the database, it builds a fresh copy each time.</p>
</div><div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.core import complex
co = complex.get_db()
co.update_index()
co
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[90]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Complex database: 28173 complexes&gt;
</pre></div></div>
</div>
<p>To retrieve all complexes containing a specific protein, here MTOR:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[91]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>co.proteins[&#39;P42345&#39;]
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[91]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{Complex: COMPLEX:O00141_O15530_O75879_P23443_P34931_P42345_Q6R327_Q8N122_Q9BPZ7_Q9BVC4_Q9H672,
 Complex: COMPLEX:O00141_O15530_P07900_P23443_P31749_P31751_P42345_P78527_Q05513_Q05655_Q6R327_Q8N122_Q9BPZ7_Q9BVC4,
 Complex: COMPLEX:O00141_O15530_P0CG47_P0CG48_P23443_P42345_Q15118_Q6R327_Q8N122_Q96BR1_Q9BPZ7_Q9BVC4,
 Complex: COMPLEX:O00141_O15530_P23443_P42345_Q15118_Q6R327_Q8N122_Q96BR1_Q96J02_Q9BPZ7_Q9BVC4,
 Complex: COMPLEX:O00141_O75879_P0CG48_P23443_P34931_P42345_P62753_Q6R327_Q8N122_Q9BPZ7_Q9BVC4_Q9NY26,
 Complex: COMPLEX:O00141_P0CG48_P23443_P36894_P42345_P62942_P68106_Q15427_Q6R327_Q8N122_Q9BPZ7_Q9BVC4,
 Complex: COMPLEX:O00141_P0CG48_P23443_P42345_P46781_P62753_Q6R327_Q8N122_Q96KQ7_Q9BPZ7_Q9BVC4_Q9NY26,
 Complex: COMPLEX:O00141_P0CG48_P23443_P42345_P62753_P62942_Q6R327_Q8N122_Q9BPZ7_Q9BVC4_Q9NY26,
 Complex: COMPLEX:O00141_P0CG48_P23443_P42345_P62753_Q15172_Q6R327_Q8IW41_Q9BPZ7_Q9BVC4_Q9H672,
 Complex: COMPLEX:O00141_P0CG48_P23443_P42345_P62753_Q6R327_Q70Z35_Q8N122_Q8TCU6_Q9BPZ7
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 5348 characters</em></div>
</div>
<p>Note some of the complexes have human readable names, these are preferred at printing if available from any of the databases. Otherwise the complexes are labelled by <code class="docutils literal notranslate"><span class="pre">COMPLEX:list-of-components</span></code>.</p>
<section id="Protein-complex-objects">
<h3>Protein complex objects<a class="headerlink" href="#Protein-complex-objects" title="Permalink to this heading">§</a></h3>
<p>Take a closer look on one complex object. The hash of the is equivalent with the string representation below, where the UniProt IDs are unique and alphabetically sorted. Hence you can look up complexes using strings as keys despite the dict keys are in fact <code class="docutils literal notranslate"><span class="pre">pypath.intera.Complex</span></code> objects:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[97]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cplex = co.complexes[&#39;COMPLEX:Q09472_Q92793&#39;]
cplex
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[97]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Complex CBP/p300: COMPLEX:Q09472_Q92793
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[98]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cplex.components # stoichiometry
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[98]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Q92793&#39;: 1, &#39;Q09472&#39;: 1}
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[99]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>cplex.sources # resources
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[99]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
{&#39;Signor&#39;}
</pre></div></div>
</div>
</section>
<section id="Protein-complex-data-frame">
<h3>Protein complex data frame<a class="headerlink" href="#Protein-complex-data-frame" title="Permalink to this heading">§</a></h3>
<p>The database can be exported into a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[117]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>co.make_df()
co.df
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[117]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>name</th>
      <th>components</th>
      <th>components_genesymbols</th>
      <th>stoichiometry</th>
      <th>sources</th>
      <th>references</th>
      <th>identifiers</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>NFY</td>
      <td>P23511_P25208_Q13952</td>
      <td>NFYA_NFYB_NFYC</td>
      <td>1:1:1</td>
      <td>PDB;ComplexPortal;hu.MAP;CORUM;hu.MAP2;Signor;...</td>
      <td>9372932;14755292;15243141</td>
      <td>Signor:SIGNOR-C1;CORUM:4478;Compleat:HC1449;in...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>mTORC2</td>
      <td>P68104_P85299_Q6R327_Q8TB45_Q9BVC4</td>
      <td>DEPTOR_EEF1A1_MLST8_PRR5_</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 3293 characters</em></div>
</div>
</section>
</section>
<section id="Saving-datasets-as-pickles">
<h2>Saving datasets as pickles<a class="headerlink" href="#Saving-datasets-as-pickles" title="Permalink to this heading">§</a></h2>
<p>The large datasets above are compiled from many resources. Even if these are already available in the cache, the data processing often takes longer than convenient, e.g. from a few minutes up to half an hour. Most of the data integration objects in <code class="docutils literal notranslate"><span class="pre">pypath</span></code> provide methods to save and load their contents as pickle dumps. In fact, the database manager does this all the time, in a coordinated way – for this reason, the methods below should be used only with good reason, and relying on the
database manager is preferred.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># for `pypath.annot.AnnotationTable` objects:
a.save_to_pickle(&#39;myannots.pickle&#39;)
a = annot.AnnotationTable(pickle_file = &#39;myannots.pickle&#39;)
# for `pypath.complex.ComplexAggregator` objects:
complexdb.save_to_pickle(&#39;mycomplexes.pickle&#39;)
complexdb = complex.ComplexAggregator(pickle_file = &#39;mycomplexes.pickle&#39;)
</pre></div>
</div>
</div>
</section>
<section id="Log-messages-and-sessions">
<h2>Log messages and sessions<a class="headerlink" href="#Log-messages-and-sessions" title="Permalink to this heading">§</a></h2>
<p>In <code class="docutils literal notranslate"><span class="pre">pypath</span></code> all modules sends messages to a log file named by default by the session ID (a 5 char random string). The default path to the log file is <code class="docutils literal notranslate"><span class="pre">./pypath_log/pypath-xxxxx.log</span></code> where <code class="docutils literal notranslate"><span class="pre">xxxxx</span></code> is the session ID.</p>
<div class="alert alert-block alert-warning"><p>Warning: The logger of pypath is really verbose, the log files can grow huge: several tens of thousands of lines, few MBs. It is recommended to empty the pypath_log directories time to time.</p>
</div><section id="Basic-info-about-the-session">
<h3>Basic info about the session<a class="headerlink" href="#Basic-info-about-the-session" title="Permalink to this heading">§</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">info</span></code> function prints the most important information about the current session:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[100]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pypath
pypath.info()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 16:41:55] [pypath]
        - session ID: `l0n17`
        - working directory: `/home/denes/pypath/notebooks`
        - logfile: `/home/denes/pypath/notebooks/pypath_log/pypath-l0n17.log`
        - pypath version: 0.14.31
</pre></div></div>
</div>
<p>Another function prints a disclaimer about licenses. Until recently this message was printed every time upon import, it is still important, but we removed it as in certain situations it can be annoying.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[101]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.disclaimer()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>

        === d i s c l a i m e r ===

        All data accessed through this module,
        either as redistributed copy or downloaded using the
        programmatic interfaces included in the present module,
        are free to use at least for academic research or
        education purposes.
        Please be aware of the licenses of all the datasets
        you use in your analysis, and please give appropriate
        credits for the original sources when you publish your
        results. To find out more about data sources please
        look at `pypath/resources/data/resources.json` or
        https://omnipathdb.org/info and
        `pypath.resources.urls.urls`.

</pre></div></div>
</div>
</section>
<section id="Read-the-log-file">
<h3>Read the log file<a class="headerlink" href="#Read-the-log-file" title="Permalink to this heading">§</a></h3>
<p>Calling <code class="docutils literal notranslate"><span class="pre">pypath.log</span></code> opens the logfile by the default console application for paginating text files (in GNU systems typically <code class="docutils literal notranslate"><span class="pre">less</span></code>):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.log()
</pre></div>
</div>
</div>
<p>The logger and the log file are bound to the session (the 5 random characters is the session ID):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[104]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[104]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Session l0n17&gt;
</pre></div></div>
</div>
<p>The logger:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[105]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session.log
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[105]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Logger [/home/denes/pypath/notebooks/pypath_log/pypath-l0n17.log]
</pre></div></div>
</div>
<p>The path to the log file:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[106]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session.log.fname
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[106]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;/home/denes/pypath/notebooks/pypath_log/pypath-l0n17.log&#39;
</pre></div></div>
</div>
</section>
<section id="Logging-to-the-console">
<h3>Logging to the console<a class="headerlink" href="#Logging-to-the-console" title="Permalink to this heading">§</a></h3>
<p>Each log message has a numeric priority level, and messages with lower level than a threshold are printed to the console. By default only important warnings are dispatched to the console. To log everything to the console, set the threshold to a large number:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[107]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session.log.console_level = 10

from pypath.inputs import signor

si = signor.signor_interactions()
pypath.session.log.console_level = -1
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 16:42:55] [curl] Creating Curl object to retrieve data from `https://signor.uniroma2.it/download_complexes.php`
[2022-12-02 16:42:55] [curl] Cache file path: `/home/denes/.pypath/cache/d7b8673e83e43a01c533f9de5a2b04b9-download_complexes.php`
[2022-12-02 16:42:55] [curl] Cache file found, no need for download.
[2022-12-02 16:42:55] [curl] Opening plain text file `/home/denes/.pypath/cache/d7b8673e83e43a01c533f9de5a2b04b9-download_complexes.php`.
[2022-12-02 16:42:55] [curl] Creating Curl object to retrieve data from `https://signor.uniroma2.it/download_complexes.php`
[2022-12-02 16:42:55] [curl] Cache file path: `/home/denes/.pypath/cache/d7b8673e83e43a01c533f9de5a2b04b9-download_complexes.php`
[2022-12-02 16:42:55] [curl] Cache file found, no need for download.
[2022-12-02 16:42:55] [curl] Opening plain text file `/home/denes/.pypath/cache/d7b8673e83e43a01c533f9de5a2b04b9-download_complexes.php`.
[2022-12-02 16:42:55] [curl] Creating Curl object to retrieve data from `https
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<em>Output truncated: showing 1000 of 1046 characters</em></div>
</div>
</section>
<section id="Disable-logging">
<h3>Disable logging<a class="headerlink" href="#Disable-logging" title="Permalink to this heading">§</a></h3>
<p>To avoid creation of a log file (and the directory <code class="docutils literal notranslate"><span class="pre">pypath_log</span></code>) set the environment variable <code class="docutils literal notranslate"><span class="pre">PYPATH_LOG</span></code> or the <code class="docutils literal notranslate"><span class="pre">builtins.PYPATH_LOG</span></code> attribute:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># shell:
export PYPATH_LOG=&quot;/dev/null&quot;
# then, start Python and use pypath
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[108]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import os
import builtins
builtins.PYPATH_LOG=os.devnull
import pypath
</pre></div>
</div>
</div>
</section>
<section id="Write-to-the-log">
<h3>Write to the log<a class="headerlink" href="#Write-to-the-log" title="Permalink to this heading">§</a></h3>
<section id="Sending-a-single-message">
<h4>Sending a single message<a class="headerlink" href="#Sending-a-single-message" title="Permalink to this heading">§</a></h4>
<p>First we change the console level so we can see the log messages. The label is optional. The priority of the message is given by the <code class="docutils literal notranslate"><span class="pre">level</span></code>, notice that the second message won’t be printed to the console as its level is higher than 10:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[109]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session.log.console_level = 10
pypath.session.log.msg(&#39;Greetings from the pypath tutorial notebook! :)&#39;, label = &#39;book&#39;)
pypath.session.log.msg(&#39;Not important, not shown on console but printed to the logfile.&#39;, level = 11)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 16:43:13] [book] Greetings from the pypath tutorial notebook! :)
</pre></div></div>
</div>
</section>
<section id="Connect-a-module-or-class-to-the-pypath-logger">
<h4>Connect a module or class to the pypath logger<a class="headerlink" href="#Connect-a-module-or-class-to-the-pypath-logger" title="Permalink to this heading">§</a></h4>
<p>The preferred way of connecting to the logger is to make a class inherit from the <code class="docutils literal notranslate"><span class="pre">Logger</span></code> class. Here the <code class="docutils literal notranslate"><span class="pre">name</span></code> will be the default label for all messages coming from the instances of this class:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[110]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import session

class ChildOfLogger(session.Logger):

    def __init__(self):

        session.Logger.__init__(self, name = &#39;child&#39;)

    def say_something(self):

        self._log(&#39;Have a nice day! :D&#39;)


col = ChildOfLogger()
col.say_something()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 16:43:17] [child] Have a nice day! :D
</pre></div></div>
</div>
<p>Alternatively, a logger can be created anywhere and used from any module or function:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[111]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.share import session

_logger = session.Logger(name = &#39;mylogger&#39;)
_log = _logger._log

_log(&#39;Message from a stray logger&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[2022-12-02 16:43:20] [mylogger] Message from a stray logger
</pre></div></div>
</div>
<p>Finally we just set the console level to a lower value, to avoid flooding the rest of this book with log messages:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[112]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pypath.session.log.console = -1
</pre></div>
</div>
</div>
</section>
</section>
</section>
<section id="BEL-export">
<h2>BEL export<a class="headerlink" href="#BEL-export" title="Permalink to this heading">§</a></h2>
<div class="alert alert-block alert-danger"><p>Warning: This section hasn’t been thoroughly revised for long time, some parts might be outdated or broken.</p>
</div><p>Biological Expression Language (BEL, <a class="reference external" href="https://bel-commons.scai.fraunhofer.de/">https://bel-commons.scai.fraunhofer.de/</a>) is a versatile description language to capture relationships between various biological entities spanning wide range of the levels of biological organization. <code class="docutils literal notranslate"><span class="pre">pypath</span></code> has a dedicated module to convert the network and the enzyme-substrate interactions to BEL format:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.legacy import main
from pypath.resources import data_formats
from pypath.omnipath import bel
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa = main.PyPath()
pa.init_network(data_formats.pathway)
</pre></div>
</div>
</div>
<p>You can provide one or more resources to the <code class="docutils literal notranslate"><span class="pre">Bel</span></code> class. Supported resources currently are <code class="docutils literal notranslate"><span class="pre">pypath.main.PyPath</span></code> and <code class="docutils literal notranslate"><span class="pre">pypath.ptm.PtmAggregator</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>b = bel.Bel(resource = pa)
</pre></div>
</div>
</div>
<p>From the resources we compile a <code class="docutils literal notranslate"><span class="pre">BELGraph</span></code> object which provides a Python interface for various operations and you can also export the data in BEL format:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>b.main()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>b.bel_graph
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>b.bel_graph.summarize()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>b.export_relationships(&#39;omnipath_pathways.bel&#39;)
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>with open(&#39;omnipath_pathways.bel&#39;, &#39;r&#39;) as fp:
    bel_str = fp.read()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(bel_str[:333])
</pre></div>
</div>
</div>
</section>
<section id="CellPhoneDB-export">
<h2>CellPhoneDB export<a class="headerlink" href="#CellPhoneDB-export" title="Permalink to this heading">§</a></h2>
<div class="alert alert-block alert-danger"><p>Warning: This section hasn’t been thoroughly revised for long time, some parts might be outdated or broken.</p>
</div><p>CellPhoneDB is a statistical method and a database for inferring inter-cellular communication pathways between specific cell types from single-cell data. OmniPath/pypath uses CellPhoneDB as a resource for interaction, protein complex and annotation data. Apart from this, pypath is able to export its data in the appropriate format to provide input for the CellPhoneDB Python module. For this you can use the <code class="docutils literal notranslate"><span class="pre">pypath.cellphonedb</span></code> module:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.omnipath import cellphonedb
from pypath.share import settings

settings.setup(network_expand_complexes = False)
</pre></div>
</div>
</div>
<p>Here you can provide parameters for the network or provide an already built network. Also you can provide the datasets as pickles to make them load really fast. Otherwise this step will take quite long.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>c = cellphonedb.CellPhoneDB()
</pre></div>
</div>
</div>
<p>You can access each of the CellPhoneDB input files as a <code class="docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code> and also they’ve been exported to csv files. For example the <code class="docutils literal notranslate"><span class="pre">interaction_input.csv</span></code> contains interactions from all the resources used for building the network (here Signor, SingnaLink, etc.):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>c.interaction_dataframe[:10]
</pre></div>
</div>
</div>
<p>The proteins and complexes are annotated (transmembrane, peripheral, secreted, etc.) using data from the <code class="docutils literal notranslate"><span class="pre">pypath.intercell</span></code> module (identical to the <a class="reference external" href="http://omnipathdb.org/intercell">http://omnipathdb.org/intercell</a> query of the web service):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>c.protein_dataframe[:10]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
<section id="The-legacy-igraph-based-network-object">
<h2>The legacy <em>igraph</em>-based network object<a class="headerlink" href="#The-legacy-igraph-based-network-object" title="Permalink to this heading">§</a></h2>
<div class="alert alert-block alert-danger"><p>Warning: This section hasn’t been thoroughly revised for long time, some parts might be outdated or broken.</p>
</div><p>Until about 2019 (before pypath version 0.9) <code class="docutils literal notranslate"><span class="pre">pypath</span></code> used an <code class="docutils literal notranslate"><span class="pre">igraph.Graph</span></code> object (<a class="reference external" href="http://igraph.org/">igraph.org</a>) to organize all data structures around. This legacy API still present in <code class="docutils literal notranslate"><span class="pre">pypath.legacy.main</span></code>, however it is not maintained. This section of the book is still here, but will be removed soon, along with the <code class="docutils literal notranslate"><span class="pre">legacy</span></code> module.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[43]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.legacy import main
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
No module `cairo` available.
Some plotting functionalities won&#39;t be accessible.
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa = main.PyPath()
#pa.load_omnipath() # This is commented out because it takes &gt; 1h
                    # to run it for the first time due to the vast
                    # amount of data download.
                    # Once you populated the cache it still takes
                    # approx. 30 min to build the entire OmniPath
                    # as the process consists of quite some data
                    # processing. If you dump it in a pickle, you
                    # can load the network in &lt; 1 min
</pre></div>
</div>
</div>
<section id="I-just-want-a-network-quickly-and-play-around-with-pypath">
<h3>I just want a network quickly and play around with <em>pypath</em><a class="headerlink" href="#I-just-want-a-network-quickly-and-play-around-with-pypath" title="Permalink to this heading">§</a></h3>
<p>You can find the predefined formats in the <code class="docutils literal notranslate"><span class="pre">pypath.resources.network</span></code> module. For example, to load one resource from there, let’s say SIGNOR:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.legacy import main
from pypath.resources import network as netres
pa = main.PyPath()
pa.load_resources({&#39;signor&#39;: netres.pathway[&#39;signor&#39;]})
</pre></div>
</div>
</div>
<p>Or to load all <em>activity flow</em> resources with <em>literature references:</em></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.legacy import main
from pypath.resources import network as netres
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa = main.PyPath()
pa.init_network(netres.pathway)
</pre></div>
</div>
</div>
<p>Or to load all <em>activity flow</em> resources, including the ones without literature references:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa = main.PyPath()
pa.init_network(data_formats.pathway_all)
</pre></div>
</div>
</div>
</section>
<section id="How-do-I-build-networks-from-any-data-with-pypath?">
<h3>How do I build networks from any data with <em>pypath</em>?<a class="headerlink" href="#How-do-I-build-networks-from-any-data-with-pypath?" title="Permalink to this heading">§</a></h3>
<p>Here we show how to build a network from your own files. The advantage of building network with pypath is that you don’t need to worry about merging redundant elements, neither about different formats and identifiers. Let’s say you have two files with network data:</p>
<p><strong>network1.csv</strong></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>entrezA,entrezB,effect
1950,1956,inhibition
5290,207,stimulation
207,2932,inhibition
1956,5290,stimulation
</pre></div>
</div>
<p><strong>network2.sif</strong></p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>EGF + EGFR
EGFR + PIK3CA
EGFR + SOS1
PIK3CA + RAC1
RAC1 + MAP3K1
SOS1 + HRAS
HRAS + MAP3K1
PIK3CA + AKT1
AKT1 - GSK3B
</pre></div>
</div>
<p><em>Note: you need to create these files in order to load them.</em></p>
<section id="Defining-input-formats">
<h4>Defining input formats<a class="headerlink" href="#Defining-input-formats" title="Permalink to this heading">§</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import pypath
import pypath.iinput_formats as input_formats

input1 = input_formats.ReadSettings(
    name = &#39;egf1&#39;,
    input = &#39;network1.csv&#39;,
    header = True,
    separator = &#39;,&#39;,
    id_col_a = 0,
    id_col_b = 1,
    id_type_a = &#39;entrez&#39;,
    id_type_b = &#39;entrez&#39;,
    sign = (2, &#39;stimulation&#39;, &#39;inhibition&#39;),
    ncbi_tax_id = 9606,
)

input2 = input_formats.ReadSettings(
    name = &#39;egf2&#39;,
    input = &#39;network2.sif&#39;,
    separator = &#39; &#39;,
    id_col_a = 0,
    id_col_b = 2,
    id_type_a = &#39;genesymbol&#39;,
    id_type_b = &#39;genesymbol&#39;,
    sign = (1, &#39;+&#39;, &#39;-&#39;),
    ncbi_tax_id = 9606,
)
</pre></div>
</div>
</div>
</section>
<section id="Creating-PyPath-object-and-loading-the-2-test-files">
<h4>Creating PyPath object and loading the 2 test files<a class="headerlink" href="#Creating-PyPath-object-and-loading-the-2-test-files" title="Permalink to this heading">§</a></h4>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>inputs = {
    &#39;egf1&#39;: input1,
    &#39;egf2&#39;: input2
}

pa = main.PyPath()
pa.reload()
pa.init_network(lst = inputs)
</pre></div>
</div>
</div>
</section>
</section>
<section id="Structure-of-the-legacy-network-object">
<h3>Structure of the legacy network object<a class="headerlink" href="#Structure-of-the-legacy-network-object" title="Permalink to this heading">§</a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from pypath.legacy import main as legacy
pa = legacy.PyPath()
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.graph
</pre></div>
</div>
</div>
<p>Number of edges and nodes:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.ecount, pa.vcount
</pre></div>
</div>
</div>
<p>The edge and vertex sequences you can access in the <code class="docutils literal notranslate"><span class="pre">es</span></code> and <code class="docutils literal notranslate"><span class="pre">vs</span></code> attributes, you can iterate these or index by integers. The edge and vertex attributes you can access by string keys. E.g. get the sources of edge 0:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.graph.es[81][&#39;sources&#39;]
</pre></div>
</div>
</div>
<section id="Directions-and-signs">
<h4>Directions and signs<a class="headerlink" href="#Directions-and-signs" title="Permalink to this heading">§</a></h4>
<p>By default the <code class="docutils literal notranslate"><span class="pre">igraph</span></code> object is undirected but it carries all direction information in Python objects assigned to each edge. Pypath can convert it to a directed <code class="docutils literal notranslate"><span class="pre">igraph</span></code> object, but you still need the <code class="docutils literal notranslate"><span class="pre">Direction</span></code> objects to have the signs, as <code class="docutils literal notranslate"><span class="pre">igraph</span></code> has no signed network representation. Certain methods need the directed <code class="docutils literal notranslate"><span class="pre">igraph</span></code> object and they will automatically create it, but you can create it manually:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.get_directed()
</pre></div>
</div>
</div>
<p>You find the directed network in the <code class="docutils literal notranslate"><span class="pre">pa.dgraph</span></code> attribute:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.dgraph
</pre></div>
</div>
</div>
<p>Now let’s take a look on the <code class="docutils literal notranslate"><span class="pre">pypath.main.Direction</span></code> objects which contain details about directions and signs. First as an example, select a random edge:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>edge = pa.graph.es[3241]
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">Direction</span></code> object is in the <code class="docutils literal notranslate"><span class="pre">dirs</span></code> edge attribute:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d = edge[&#39;dirs&#39;]
</pre></div>
</div>
</div>
<p>It has a method to print its content a human readable way:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(pa.graph.es[3241][&#39;dirs&#39;])
</pre></div>
</div>
</div>
<p>From this we see the databases phosphoELM and Signor agree that protein <code class="docutils literal notranslate"><span class="pre">P17252</span></code> has an effect on <code class="docutils literal notranslate"><span class="pre">Q15139</span></code> and Signor in addition tells us this effect is stimulatory. However in your scripts you can query the <code class="docutils literal notranslate"><span class="pre">Direction</span></code> objects a number of ways. Each <code class="docutils literal notranslate"><span class="pre">Direction</span></code> object calls the two possible directions either straight or reverse:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.straight
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.reverse
</pre></div>
</div>
</div>
<p>It can tell you if one of these directions is supported by any of the network resources:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.get_dir(d.straight)
</pre></div>
</div>
</div>
<p>Or it can return those resources:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.get_dir(d.straight, sources = True)
</pre></div>
</div>
</div>
<p>The opposite direction is not supported by any resource:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.get_dir(d.reverse, sources = True)
</pre></div>
</div>
</div>
<p>Similar way the signs can be queried. The returned pair of boolean values mean if the interaction in this direction is stimulatory or inhibitory, respectively.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.get_sign(d.straight)
</pre></div>
</div>
</div>
<p>Or you can ask whether it is inhibition:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.is_inhibition(d.straight)
</pre></div>
</div>
</div>
<p>Or if the interaction is directed at all:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.is_directed()
</pre></div>
</div>
</div>
<p>Sometimes resources don’t agree, for example one tells an interaction is inhibition while according to others it is stimulation; or one tells A effects B and another resource the other way around. Here we preserve all these potentially contradicting information in the <code class="docutils literal notranslate"><span class="pre">Direction</span></code> object and at the end you decide what to do with it depending on your purpose. If you want to get rid of ambiguity there is a method to get a consensus direction and sign which returns the attributes the most
resources agree on:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>d.consensus_edges()
</pre></div>
</div>
</div>
</section>
<section id="Accessing-nodes-in-the-network">
<h4>Accessing nodes in the network<a class="headerlink" href="#Accessing-nodes-in-the-network" title="Permalink to this heading">§</a></h4>
<p>In <code class="docutils literal notranslate"><span class="pre">igraph</span></code> the vertices are numbered but this numbering can change at certain operations. Instead the we can use the vertex attributes. In <code class="docutils literal notranslate"><span class="pre">PyPath</span></code> for proteins the <code class="docutils literal notranslate"><span class="pre">name</span></code> attribute is UniProt ID by default and the <code class="docutils literal notranslate"><span class="pre">label</span></code> is Gene Symbol.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.graph.vs[&#39;name&#39;][:5]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.graph.vs[&#39;label&#39;][:5]
</pre></div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">PyPath</span></code> object offers a number of helper methods to access the nodes by their names. For example, <code class="docutils literal notranslate"><span class="pre">uniprot</span></code> or <code class="docutils literal notranslate"><span class="pre">up</span></code> returns the <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> for a UniProt ID:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>type(pa.up(&#39;P00533&#39;))
</pre></div>
</div>
</div>
<p>Similarly <code class="docutils literal notranslate"><span class="pre">genesymbol</span></code> or <code class="docutils literal notranslate"><span class="pre">gs</span></code> for Gene Symbols:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>type(pa.gs(&#39;ESR1&#39;))
</pre></div>
</div>
</div>
<p>Each of these has a “plural” version:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>len(list(pa.gss([&#39;MTOR&#39;, &#39;ATG16L2&#39;, &#39;ULK1&#39;])))
</pre></div>
</div>
</div>
<p>And a generic method where you can mix UniProts and Gene Symbols:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>len(list(pa.proteins([&#39;MTOR&#39;, &#39;P00533&#39;])))
</pre></div>
</div>
</div>
</section>
</section>
<section id="Querying-relationships-with-our-without-causality">
<h3>Querying relationships with our without causality<a class="headerlink" href="#Querying-relationships-with-our-without-causality" title="Permalink to this heading">§</a></h3>
<p>Above you could see how to query the directions and names of individual edges and nodes. Building on top of these, other methods give a way to query causality, i.e. which proteins are affected by an other one, and which others are its regulators. The example below returns the nodes PIK3CA is stimulated by, the <code class="docutils literal notranslate"><span class="pre">gs</span></code> prefix tells we query by the Gene Symbol:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>pa.gs_stimulated_by(&#39;PIK3CA&#39;)
</pre></div>
</div>
</div>
<p>It returns a so called <code class="docutils literal notranslate"><span class="pre">_NamedVertexSeq</span></code> object, which you can get a series of <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> objects or Gene Symbols or UniProt IDs from:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>list(pa.gs_stimulated_by(&#39;PIK3CA&#39;).gs())[:5]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>list(pa.gs_stimulated_by(&#39;PIK3CA&#39;).up())[:5]
</pre></div>
</div>
</div>
<p>Note, the names of these methods are a bit contraintuitive, the for example the <code class="docutils literal notranslate"><span class="pre">gs_stimulates</span></code> returns the genes stimulated by PIK3CA:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>list(pa.gs_stimulates(&#39;PIK3CA&#39;).gs())[:5]
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>&#39;PIK3CA&#39; in set(pa.affected_by(&#39;AKT1&#39;).gs())
</pre></div>
</div>
</div>
<p>There are many similary methods, <code class="docutils literal notranslate"><span class="pre">inhibited_by</span></code> returns negative regulators, <code class="docutils literal notranslate"><span class="pre">affected_by</span></code> does not consider +/- signs, without <code class="docutils literal notranslate"><span class="pre">gs_</span></code> and <code class="docutils literal notranslate"><span class="pre">up_</span></code> prefixes you can provide either of these identifiers, <code class="docutils literal notranslate"><span class="pre">neighbors</span></code> does not consider the direction. At the end <code class="docutils literal notranslate"><span class="pre">.gs()</span></code> converts the result for a list of Gene Symbols, <code class="docutils literal notranslate"><span class="pre">up()</span></code> to UniProts, <code class="docutils literal notranslate"><span class="pre">.ids()</span></code> to vertex IDs and by default it yields <code class="docutils literal notranslate"><span class="pre">igraph.Vertex</span></code> objects:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>list(pa.neighbors(&#39;AKT1&#39;).ids())[:5]
</pre></div>
</div>
</div>
<p>Finally, with <code class="docutils literal notranslate"><span class="pre">neighborhood</span></code> methods return the indirect neighborhood in custom number of steps (however size of the neighborhood increases rapidly with number of steps):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(list(pa.neighborhood(&#39;ATG3&#39;, 1).gs()))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(list(pa.neighborhood(&#39;ATG3&#39;, 2).gs()))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>len(list(pa.neighborhood(&#39;ATG3&#39;, 3).gs()))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>len(list(pa.neighborhood(&#39;ATG3&#39;, 4).gs()))
</pre></div>
</div>
</div>
</section>
<section id="Accessing-edges-by-identifiers">
<h3>Accessing edges by identifiers<a class="headerlink" href="#Accessing-edges-by-identifiers" title="Permalink to this heading">§</a></h3>
<p>Just like nodes also edges can be accessed by identifiers like Gene Symbols. <code class="docutils literal notranslate"><span class="pre">get_edge</span></code> returns an <code class="docutils literal notranslate"><span class="pre">igraph.Edge</span></code> if the edge exists otherwise <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>type(pa.get_edge(&#39;EGF&#39;, &#39;EGFR&#39;))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>type(pa.get_edge(&#39;EGF&#39;, &#39;P00533&#39;))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>type(pa.get_edge(&#39;EGF&#39;, &#39;AKT1&#39;))
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>print(pa.get_edge(&#39;EGF&#39;, &#39;EGFR&#39;)[&#39;dirs&#39;])
</pre></div>
</div>
</div>
</section>
<section id="Literature-references">
<h3>Literature references<a class="headerlink" href="#Literature-references" title="Permalink to this heading">§</a></h3>
<p>Select a random edge and in the <code class="docutils literal notranslate"><span class="pre">references</span></code> attribute you find a list of references:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>edge = pa.get_edge( &#39;MAP1LC3B&#39;, &#39;SQSTM1&#39;)
edge[&#39;references&#39;]
</pre></div>
</div>
</div>
<p>Each reference has a PubMed ID:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>edge[&#39;references&#39;][0].pmid
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>edge[&#39;references&#39;][0].open()
</pre></div>
</div>
</div>
<p>These 3 references come from 3 different databases, but there must be 2 overlaps between them:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>edge[&#39;refs_by_source&#39;]
</pre></div>
</div>
</div>
</section>
<section id="Plotting-the-network-with-igraph">
<h3>Plotting the network with <em>igraph</em><a class="headerlink" href="#Plotting-the-network-with-igraph" title="Permalink to this heading">§</a></h3>
<p>Here we use the network created above (because it is reasonable size, not like the networks we could get from most of the network databases). Igraph has excellent plotting abilities built on top of the <em>cairo</em> library.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import igraph
plot = igraph.plot(pa.graph, target = &#39;egf_network.png&#39;,
            edge_width = 0.3, edge_color = &#39;#777777&#39;,
            vertex_color = &#39;#97BE73&#39;, vertex_frame_width = 0,
            vertex_size = 70.0, vertex_label_size = 15,
            vertex_label_color = &#39;#FFFFFF&#39;,
            # due to a bug in either igraph or IPython,
            # vertex labels are not visible on inline plots:
            inline = False, margin = 120)
from IPython.display import Image
Image(filename=&#39;egf_network.png&#39;)
</pre></div>
</div>
</div>
</section>
</section>
</section>


            </article>
            
            
            
            <footer class="bd-footer-article">
                <!-- Previous / next buttons -->
<div class='prev-next-area'>
  <a class='left-prev' id="prev-link" href="../api/pypath.visual.plot.StackedBarplot.html" title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
          <p class="prev-next-subtitle">previous</p>
          <p class="prev-next-title">pypath.visual.plot.StackedBarplot</p>
      </div>
  </a>
</div>
            </footer>
            
          </div>
          
          
          
            <div class="bd-sidebar-secondary bd-toc">
              
<div class="toc-item">
  
<div class="tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
</div>
<nav id="bd-toc-nav" class="page-toc">
    <ul class="current visible nav section-nav flex-column">
 <li class="toctree-l1 toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="../index.html">
   Welcome to pypath!
  </a>
 </li>
 <li class="toctree-l1 toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="../webservice.html">
   Webservice
  </a>
 </li>
 <li class="toctree-l1 toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="../releasehistory.html">
   Release history
  </a>
 </li>
 <li class="toctree-l1 toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="../reference.html">
   Reference
  </a>
 </li>
 <li class="toctree-l1 current toc-h1 nav-item toc-entry">
  <a class="current reference internal nav-link" href="#">
   The pypath book
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Introduction">
     Introduction
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Build,-load-and-save-databases">
     Build, load and save databases
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#The-OmniPath-app">
       The OmniPath app
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Built-in-database-definitions">
       Built-in database definitions
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Networks">
       Networks
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Strictly-literature-curated-network">
         Strictly literature curated network
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#The-OmniPath-network-with-extra-activity-flow,-enzyme-substrate-and-ligand-receptor-interactions">
         The OmniPath network with extra activity flow, enzyme-substrate and ligand-receptor interactions
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Transcriptional-regulation-network-from-DoRothEA-and-other-resources">
         Transcriptional regulation network from DoRothEA and other resources
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Literature-curated-miRNA-post-transcriptional-regulation-network">
         Literature curated miRNA post-transcriptional regulation network
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Transcriptional-regulation-of-miRNA">
         Transcriptional regulation of miRNA
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#lncRNA-mRNA-interactions">
         lncRNA-mRNA interactions
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Small-molecule-protein-interactions">
         Small molecule-protein interactions
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Enzyme-substrate-relationships">
       Enzyme-substrate relationships
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Protein-complexes">
       Protein complexes
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Annotations">
       Annotations
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Inter-cellular-communication-roles">
       Inter-cellular communication roles
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Data-directly-from-the-original-resources">
     Data directly from the original resources
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Download-management">
     Download management
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Cache-management-and-customization">
       Cache management and customization
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Download-failures">
       Download failures
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Corrupted-cache-content">
         Corrupted cache content
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Network-communication-issues:-look-into-the-curl-debug-log">
         Network communication issues: look into the curl debug log
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Timeouts">
         Timeouts
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Access-and-inspect-the-Curl-object">
         Access and inspect the
         <code class="docutils literal notranslate">
          <span class="pre">
           Curl
          </span>
         </code>
         object
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Is-it-failing-only-for-you?">
         Is it failing only for you?
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Read-the-log">
         Read the log
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#TLS-(SSL,-HTTPS)-errors">
         TLS (SSL, HTTPS) errors
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Resources">
     Resources
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Licenses">
       Licenses
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Resource-information">
       Resource information
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Resource-definitions-for-a-certain-database-or-dataset">
       Resource definitions for a certain database or dataset
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Building-networks">
     Building networks
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Which-network-datasets-are-pre-defined-in-pypath?">
       Which network datasets are pre-defined in pypath?
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#The-Network-object">
       The
       <code class="docutils literal notranslate">
        <span class="pre">
         Network
        </span>
       </code>
       object
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Network-in-pandas.DataFrame">
       Network in
       <em>
        pandas.DataFrame
       </em>
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Translating-identifiers">
     Translating identifiers
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Pre-defined-ID-translation-tables">
       Pre-defined ID translation tables
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Homology-translation">
     Homology translation
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Homology-translation-tables-as-dictionaries">
       Homology translation tables as dictionaries
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Homology-translation-data-frames">
       Homology translation data frames
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Taxonomy">
     Taxonomy
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Translating-to-NCBI-Taxonomy,-scientific-names-and-common-names">
       Translating to NCBI Taxonomy, scientific names and common names
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Organism-from-UniProt-ID">
       Organism from UniProt ID
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#UniProt">
     UniProt
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#The-UniProt-input-module">
       The UniProt input module
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#All-UniProt-IDs-for-one-organism">
         All UniProt IDs for one organism
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#UniProt-ID-format-validation">
         UniProt ID format validation
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#UniProt-ID-validation">
         UniProt ID validation
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Single-UniProt-protein-datasheet">
         Single UniProt protein datasheet
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#History-of-UniProt-records">
         History of UniProt records
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#UniProt-legacy-API">
         UniProt legacy API
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Processed-UniProt-annotations">
         Processed UniProt annotations
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#The-UniProt-utils-module">
       The UniProt utils module
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Datasheets">
         Datasheets
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Tables">
         Tables
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Sanitizing-UniProt-IDs">
       Sanitizing UniProt IDs
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Enzyme-substrate-interactions">
     Enzyme-substrate interactions
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Enzyme-substrate-objects">
       Enzyme-substrate objects
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Enzyme-substrate-data-frame">
       Enzyme-substrate data frame
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Protein-sequences">
     Protein sequences
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id1">
     Annotations
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Load-a-single-annotation-resource">
       Load a single annotation resource
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Load-the-full-annotations-database-by-the-database-manager">
       Load the full annotations database by the database manager
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Load-only-selected-annotations">
       Load only selected annotations
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Data-frames-of-annotations">
       Data frames of annotations
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Inter-cellular-signaling-roles">
     Inter-cellular signaling roles
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Build-an-intercellular-communication-network">
       Build an intercellular communication network
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Quantitative-overview-of-intercell-annotations">
       Quantitative overview of intercell annotations
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Intercell-database-as-data-frame">
       Intercell database as data frame
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Browse-intercell-categories">
       Browse intercell categories
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Gene-Ontology">
     Gene Ontology
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#id2">
     Protein complexes
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Protein-complex-objects">
       Protein complex objects
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Protein-complex-data-frame">
       Protein complex data frame
      </a>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Saving-datasets-as-pickles">
     Saving datasets as pickles
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Log-messages-and-sessions">
     Log messages and sessions
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Basic-info-about-the-session">
       Basic info about the session
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Read-the-log-file">
       Read the log file
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Logging-to-the-console">
       Logging to the console
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Disable-logging">
       Disable logging
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Write-to-the-log">
       Write to the log
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Sending-a-single-message">
         Sending a single message
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Connect-a-module-or-class-to-the-pypath-logger">
         Connect a module or class to the pypath logger
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#BEL-export">
     BEL export
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#CellPhoneDB-export">
     CellPhoneDB export
    </a>
   </li>
   <li class="toctree-l2 toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#The-legacy-igraph-based-network-object">
     The legacy
     <em>
      igraph
     </em>
     -based network object
    </a>
    <ul class="visible nav section-nav flex-column">
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#I-just-want-a-network-quickly-and-play-around-with-pypath">
       I just want a network quickly and play around with
       <em>
        pypath
       </em>
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#How-do-I-build-networks-from-any-data-with-pypath?">
       How do I build networks from any data with
       <em>
        pypath
       </em>
       ?
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Defining-input-formats">
         Defining input formats
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Creating-PyPath-object-and-loading-the-2-test-files">
         Creating PyPath object and loading the 2 test files
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Structure-of-the-legacy-network-object">
       Structure of the legacy network object
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Directions-and-signs">
         Directions and signs
        </a>
       </li>
       <li class="toctree-l4 toc-h4 nav-item toc-entry">
        <a class="reference internal nav-link" href="#Accessing-nodes-in-the-network">
         Accessing nodes in the network
        </a>
       </li>
      </ul>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Querying-relationships-with-our-without-causality">
       Querying relationships with our without causality
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Accessing-edges-by-identifiers">
       Accessing edges by identifiers
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Literature-references">
       Literature references
      </a>
     </li>
     <li class="toctree-l3 toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#Plotting-the-network-with-igraph">
       Plotting the network with
       <em>
        igraph
       </em>
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
</ul>

</nav>
</div>

<div class="toc-item">
  
<div id="searchbox"></div>
</div>

<div class="toc-item">
  
</div>

<div class="toc-item">
  
<div class="tocsection sourcelink">
    <a href="../_sources/notebooks/manual.ipynb.txt">
        <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
</div>

</div>

            </div>
          
          
        </div>
        <footer class="bd-footer-content">
          <div class="bd-footer-content__inner">
            
          </div>
        </footer>
        
      </main>
    </div>
  </div>

  
    
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=796348d33e8b1d947c94"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=796348d33e8b1d947c94"></script>

  <footer class="bd-footer"><div class="bd-footer__inner container">
  
  <div class="footer-item">
    
<p class="copyright">

    &copy; Copyright 2022, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Erva Ulusoy, Ahmet Rifaioglu, Charles Tapley Hoyt.<br>

</p>

  </div>
  
  <div class="footer-item">
    <p class="theme-version">
    Built with the
    <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">
        PyData Sphinx Theme
    </a>
    0.12.0.
</p>
  </div>
  
  <div class="footer-item">
    
<p class="sphinx-version">
Created using <a href="http://sphinx-doc.org/">Sphinx</a> 5.3.0.<br>
</p>

  </div>
  
</div>
  </footer>
  </body>
</html>