
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />

    <title>pypath.utils.homology.ProteinHomology &#8212; pypath 0.14.29 documentation</title>
    <link rel="stylesheet" href="../_static/basic.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/insipid.css" />
    <link rel="stylesheet" type="text/css" href="../_static/css/pypathdocs.css" />

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script defer src="../_static/insipid.js"></script>
    <script defer src="../_static/insipid-sidebar.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pypath.utils.homology.Proteomes" href="pypath.utils.homology.Proteomes.html" />
    <link rel="prev" title="pypath.utils.homology.Ortholog" href="pypath.utils.homology.Ortholog.html" />
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-capable" content="yes">
  </head><body>
    <script type="text/javascript">
        document.body.classList.add('js');
    </script>
    <input type="checkbox" id="sidebar-checkbox" style="display: none;">
    <label id="overlay" for="sidebar-checkbox"></label>
    <div class="sidebar-resize-handle"></div>
    <script type="text/javascript">
        try {
            let sidebar = localStorage.getItem('sphinx-sidebar');
            const sidebar_width = localStorage.getItem('sphinx-sidebar-width');
            if (sidebar_width) {
                document.documentElement.style.setProperty('--sidebar-width', sidebar_width);
            }
            // show sidebar on wide screen
            if (!sidebar && window.matchMedia('(min-width: 100rem)').matches) {
                sidebar = 'visible';
                // NB: We don't store the value in localStorage!
            }
            if (sidebar === 'visible') {
                document.getElementById('sidebar-checkbox').checked = true;
            }
        } catch(e) {
            console.info(e);
        }
    </script>
    <header id="topbar-placeholder">
      <div id="topbar">
        <div id="titlebar">
          <div class="buttons">
            <label for="sidebar-checkbox" id="sidebar-button" role="button" tabindex="0" aria-controls="sphinxsidebar" accesskey="M">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M16 132h416c8.837 0 16-7.163 16-16V76c0-8.837-7.163-16-16-16H16C7.163 60 0 67.163 0 76v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16zm0 160h416c8.837 0 16-7.163 16-16v-40c0-8.837-7.163-16-16-16H16c-8.837 0-16 7.163-16 16v40c0 8.837 7.163 16 16 16z"/></svg>
            </label>
            <button id="search-button" type="button" title="Search" aria-label="Search" aria-expanded="false" aria-controls="search-form" accesskey="S">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"/></svg>
            </button>
          </div>
          <div class="title">
            <a class="parent" href="pypath.utils.homology.html" accesskey="U">pypath.utils.homology</a>
            <a class="top" href="#">pypath.utils.homology.ProteinHomology</a>
          </div>
          <div class="buttons">
            <button id="fullscreen-button" type="button" aria-hidden="true">
              <span class="enable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M212.686 315.314L120 408l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C10.697 480 0 469.255 0 456V344c0-21.382 25.803-32.09 40.922-16.971L72 360l92.686-92.686c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.249 6.248 6.249 16.378 0 22.627zm22.628-118.628L328 104l-32.922-31.029C279.958 57.851 290.666 32 312.048 32h112C437.303 32 448 42.745 448 56v112c0 21.382-25.803 32.09-40.922 16.971L376 152l-92.686 92.686c-6.248 6.248-16.379 6.248-22.627 0l-25.373-25.373c-6.249-6.248-6.249-16.378 0-22.627z"/></svg>
              </span>
              <span class="disable">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M4.686 427.314L104 328l-32.922-31.029C55.958 281.851 66.666 256 88.048 256h112C213.303 256 224 266.745 224 280v112c0 21.382-25.803 32.09-40.922 16.971L152 376l-99.314 99.314c-6.248 6.248-16.379 6.248-22.627 0L4.686 449.941c-6.248-6.248-6.248-16.379 0-22.627zM443.314 84.686L344 184l32.922 31.029c15.12 15.12 4.412 40.971-16.97 40.971h-112C234.697 256 224 245.255 224 232V120c0-21.382 25.803-32.09 40.922-16.971L296 136l99.314-99.314c6.248-6.248 16.379-6.248 22.627 0l25.373 25.373c6.248 6.248 6.248 16.379 0 22.627z"/></svg>
              </span>
            </button>
          </div>
        </div>
        <div id="searchbox" role="search">
          <form id="search-form" class="search" style="display: none" action="../search.html" method="get">
            <input type="search" name="q" placeholder="Search ..." autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" />
            <button>Go</button>
          </form>
        </div>
      </div>
    </header>
    <nav>
      <a href="pypath.utils.homology.Ortholog.html" class="nav-icon previous" title="previous:&#13;pypath.utils.homology.Ortholog" aria-label="Previous topic" accesskey="P" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 256 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"/></svg>
      </a>
      <a href="pypath.utils.homology.Proteomes.html" class="nav-icon next" title="next:&#13;pypath.utils.homology.Proteomes" aria-label="Next topic" accesskey="N" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 256 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"/></svg>
      </a>
    </nav>

    <nav class="relbar">
      <a class="previous" href="pypath.utils.homology.Ortholog.html" aria-label="Previous topic" >
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"/></svg>
        </div>
        <div class="title">
          <span class="text">
            <span class="direction">previous</span>
            pypath.utils.homology.Ortholog
          </span>
        </div>
      </a>
      <a class="next" href="pypath.utils.homology.Proteomes.html" aria-label="Next topic" >
        <div class="title">
          <span class="text">
            <span class="direction">next</span>
            pypath.utils.homology.Proteomes
          </span>
        </div>
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"/></svg>
        </div>
      </a>
    </nav>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
<section id="pypath-utils-homology-proteinhomology">
<h1>pypath.utils.homology.ProteinHomology<a class="headerlink" href="#pypath-utils-homology-proteinhomology" title="Permalink to this heading">§</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pypath.utils.homology.</span></span><span class="sig-name descname"><span class="pre">ProteinHomology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology" title="Permalink to this definition">§</a></dt>
<dd><p>Bases: <a class="reference internal" href="pypath.utils.homology.Proteomes.html#pypath.utils.homology.Proteomes" title="pypath.utils.homology.Proteomes"><code class="xref py py-class docutils literal notranslate"><span class="pre">Proteomes</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_swissprot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.__init__" title="Permalink to this definition">§</a></dt>
<dd><p>This class translates between homologous UniProt IDs of two organisms
based on NCBI HomoloGene and Ensembl data. In case of HomoloGene,
the UniProt-UniProt translation table is created by translating the
source organism UniProts to RefSeq and Entrez IDs, finding the
homologues (orthologues) for these IDs, and then translating them
to the target organism UniProt IDs. In case of Ensembl, we obtain
data with Ensembl protein identifiers and translate those to UniProt.</p>
<dl>
<dt>Args</dt><dd><p>target (int): NCBI Taxonomy ID of the target organism.
source (int): NCBI Taxonomy ID of the default source organism.</p>
<blockquote>
<div><p>Multiple source organisms can be used on the same instance.</p>
</div></blockquote>
<p>only_swissprot (bool): Use only SwissProt IDs.
homologene (bool): Use homology information from NCBI HomoloGene.
ensembl (bool): Use homology information from Ensembl.
ensembl_hc (bool): Use only the high confidence</p>
<blockquote>
<div><p>orthology relations from Ensembl.</p>
</div></blockquote>
<dl class="simple">
<dt>ensembl_types (list): Ensembl orthology relation types to use.</dt><dd><p>Possible values are <cite>one2one</cite>, <cite>one2many</cite> and <cite>many2many</cite>.
By default only <cite>one2one</cite> is used.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.__init__" title="pypath.utils.homology.ProteinHomology.__init__"><code class="xref py py-obj docutils literal notranslate"><span class="pre">__init__</span></code></a>(target[, source, only_swissprot, ...])</p></td>
<td><p>This class translates between homologous UniProt IDs of two organisms based on NCBI HomoloGene and Ensembl data.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.asdict" title="pypath.utils.homology.ProteinHomology.asdict"><code class="xref py py-obj docutils literal notranslate"><span class="pre">asdict</span></code></a>([source, homologene, ensembl, ...])</p></td>
<td><p>Create a dictionary for one source organism and ID type.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.df" title="pypath.utils.homology.ProteinHomology.df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">df</span></code></a>([source, homologene, ensembl, ...])</p></td>
<td><p>Create a data frame for one source organism and ID type.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">ensure_source_taxon</span></code>(source)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_source</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_taxon</span></code>(protein[, swissprot_only])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_taxon_trembl</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">has_protein</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">is_swissprot</span></code>(protein)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_ensembl</span></code>(source)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.load_homologene" title="pypath.utils.homology.ProteinHomology.load_homologene"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_homologene</span></code></a>(source)</p></td>
<td><p>Builds orthology translation table as dict from UniProt to Uniprot, obtained from NCBI HomoloGene data.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_proteome</span></code>(taxon[, swissprot_only])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reload</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_default_source</span></code>([source])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.translate" title="pypath.utils.homology.ProteinHomology.translate"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code></a>(protein[, source, homologene, ...])</p></td>
<td><p>For one UniProt ID of the source organism returns all orthologues from the target organism.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.translate_df" title="pypath.utils.homology.ProteinHomology.translate_df"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_df</span></code></a>(df[, cols, source, homologene, ...])</p></td>
<td><p>Translate columns in a data frame.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.asdict">
<span class="sig-name descname"><span class="pre">asdict</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uniprot'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.asdict"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.asdict" title="Permalink to this definition">§</a></dt>
<dd><p>Create a dictionary for one source organism and ID type.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>id_type:</dt><dd><p>Identifier type to translate from and to. If a tuple, the
first element will be used for keys (source organism) and
the second for the values (target organism). By default
UniProt IDs are assumed. For a complete list of ID types see
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().id_types()</span></code> and
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().mapping_tables()</span></code>.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A dict with identifiers of the source organism as keys, and
sets of their orthologs as values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.df">
<span class="sig-name descname"><span class="pre">df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uniprot'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.df" title="Permalink to this definition">§</a></dt>
<dd><p>Create a data frame for one source organism and ID type.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>id_type:</dt><dd><p>Identifier type to translate from and to. If a tuple, the
first element will be used for keys (source organism) and
the second for the values (target organism). By default
UniProt IDs are assumed. For a complete list of ID types see
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().id_types()</span></code> and
<code class="docutils literal notranslate"><span class="pre">pypath.utils.mapping.get_mapper().mapping_tables()</span></code>.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A data frame with pairs of orthologous identifiers,
in two columns: “source” and “target”.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.load_homologene">
<span class="sig-name descname"><span class="pre">load_homologene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">source</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.load_homologene"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.load_homologene" title="Permalink to this definition">§</a></dt>
<dd><p>Builds orthology translation table as dict from UniProt to Uniprot,
obtained from NCBI HomoloGene data. Uses RefSeq and Entrez IDs for
translation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.translate">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'uniprot'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.translate" title="Permalink to this definition">§</a></dt>
<dd><p>For one UniProt ID of the source organism returns all orthologues
from the target organism.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>protein:</dt><dd><p>UniProt ID of one or more protein in the source organism.</p>
</dd>
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>id_type:</dt><dd><p>Identifier type to translate from and to.</p>
</dd>
</dl>
</dd>
<dt>Returns</dt><dd><p>Set of UniProt IDs of homologous proteins in the target taxon.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pypath.utils.homology.ProteinHomology.translate_df">
<span class="sig-name descname"><span class="pre">translate_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cols</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">homologene</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_hc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_types</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'one2one'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'one2many'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'many2many'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pypath/utils/homology.html#ProteinHomology.translate_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pypath.utils.homology.ProteinHomology.translate_df" title="Permalink to this definition">§</a></dt>
<dd><p>Translate columns in a data frame.</p>
<dl class="simple">
<dt>Args</dt><dd><dl class="simple">
<dt>df:</dt><dd><p>A data frame.</p>
</dd>
<dt>cols:</dt><dd><p>One or more columns to be translated. It can be a single
column name, an iterable of column names or a dict where
keys are column names and values are ID types. Except this
last case, identifiers are assumed to be UniProt.</p>
</dd>
<dt>source:</dt><dd><p>The source organism.</p>
</dd>
<dt>homologene:</dt><dd><p>Use NCBI HomoloGene data for ortholog lookup.</p>
</dd>
<dt>ensembl:</dt><dd><p>Use Ensembl data for ortholog lookup.</p>
</dd>
<dt>ensembl_hc:</dt><dd><p>Use only high confidence orthology relations
from Ensembl. By default it is True. You can also set it
by the <cite>ensembl_hc</cite> attribute.</p>
</dd>
<dt>ensembl_types:</dt><dd><p>The Ensembl orthology relationship types
to use. Possible values are <cite>one2one</cite>, <cite>one2many</cite> and
<cite>many2many</cite>. By default only <cite>one2one</cite> is used. You can
also set this parameter by the <cite>ensembl_types</cite> attribute.</p>
</dd>
<dt>kwargs:</dt><dd><p>Same as providing a dict to <code class="docutils literal notranslate"><span class="pre">cols</span></code>, but beware, keys
(column names) can not match existing argument names of
this function.</p>
</dd>
</dl>
</dd>
<dt>Return</dt><dd><p>A data frame with the same column layout as the input, and the
identifiers translated as demanded. Rows that could not be
translated are omitted.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>

            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">

          <div class="sidebar-resize-handle"></div>

<h3><a href="../index.html">Table of Contents</a></h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../webservice.html">Webservice</a></li>
<li class="toctree-l1"><a class="reference internal" href="../releasehistory.html">Release History</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="pypath.html">pypath</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="pypath.disclaimer.html">pypath.disclaimer</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.info.html">pypath.info</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.log.html">pypath.log</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.core.html">pypath.core</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.data.html">pypath.data</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.inputs.html">pypath.inputs</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.internals.html">pypath.internals</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.legacy.html">pypath.legacy</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.obsolete.html">pypath.obsolete</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.omnipath.html">pypath.omnipath</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.resources.html">pypath.resources</a></li>
<li class="toctree-l2"><a class="reference internal" href="pypath.share.html">pypath.share</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="pypath.utils.html">pypath.utils</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.go.html">pypath.utils.go</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="pypath.utils.homology.html">pypath.utils.homology</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.get_df.html">pypath.utils.homology.get_df</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.get_dict.html">pypath.utils.homology.get_dict</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.get_manager.html">pypath.utils.homology.get_manager</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.init.html">pypath.utils.homology.init</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.translate.html">pypath.utils.homology.translate</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.translate_df.html">pypath.utils.homology.translate_df</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.HomologyManager.html">pypath.utils.homology.HomologyManager</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.Ortholog.html">pypath.utils.homology.Ortholog</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">pypath.utils.homology.ProteinHomology</a><ul>
<li class="toctree-l5"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology"><code class="docutils literal notranslate"><span class="pre">ProteinHomology</span></code></a><ul>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.__init__"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.__init__()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.asdict"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.asdict()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.df"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.df()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.load_homologene"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.load_homologene()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.translate"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.translate()</span></code></a></li>
<li class="toctree-l6"><a class="reference internal" href="#pypath.utils.homology.ProteinHomology.translate_df"><code class="docutils literal notranslate"><span class="pre">ProteinHomology.translate_df()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.Proteomes.html">pypath.utils.homology.Proteomes</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.PtmHomology.html">pypath.utils.homology.PtmHomology</a></li>
<li class="toctree-l4"><a class="reference internal" href="pypath.utils.homology.SequenceContainer.html">pypath.utils.homology.SequenceContainer</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.mapping.html">pypath.utils.mapping</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.pdb.html">pypath.utils.pdb</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.proteomicsdb.html">pypath.utils.proteomicsdb</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.pyreact.html">pypath.utils.pyreact</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.reflists.html">pypath.utils.reflists</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.residues.html">pypath.utils.residues</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.seq.html">pypath.utils.seq</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.taxonomy.html">pypath.utils.taxonomy</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.unichem.html">pypath.utils.unichem</a></li>
<li class="toctree-l3"><a class="reference internal" href="pypath.utils.uniprot.html">pypath.utils.uniprot</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pypath.visual.html">pypath.visual</a></li>
</ul>
</li>
</ul>

<hr class="docutils" />
<ul>
  <li class="toctree-l1"><a class="reference internal" href="../genindex.html" accesskey="I">General Index</a></li>
  <li class="toctree-l1"><a class="reference internal" href="../py-modindex.html">Python Module Index</a></li>
</ul>
<div id="ethical-ad-placement"></div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>

    <nav class="relbar">
      <a class="previous" href="pypath.utils.homology.Ortholog.html" aria-label="Previous topic" >
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"/></svg>
        </div>
        <div class="title">
          <span class="text">
            <span class="direction">previous</span>
            pypath.utils.homology.Ortholog
          </span>
        </div>
      </a>
      <a class="next" href="pypath.utils.homology.Proteomes.html" aria-label="Next topic" >
        <div class="title">
          <span class="text">
            <span class="direction">next</span>
            pypath.utils.homology.Proteomes
          </span>
        </div>
        <div class="icon">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!-- Font Awesome Free 5.15.4 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) --><path d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"/></svg>
        </div>
      </a>
    </nav>

    <footer role="contentinfo">
      &#169; Copyright 2022, Denes Turei &lt;turei.denes@gmail.com&gt;, Nicolàs Palacio, Sebastian Lobentanzer, Olga Ivanova, Erva Ulusoy, Ahmet Rifaioglu, Charles Tapley Hoyt.
      Created using <a class="reference external" href="https://www.sphinx-doc.org/">Sphinx</a> 5.3.0.
      <a class="reference external" href="https://insipid-sphinx-theme.readthedocs.io/">Insipid Theme</a>.
    </footer>
  </body>
</html>